---
title: "Nanopore iBAT browser tracks Adcy3/Cars2"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
    html_document:
        toc: true
        toc_float: true
        toc_depth: 2
        fig_caption: false
        dev: "svg"
editor_options:
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE,
    cache = FALSE, dpi = 600)
```

```{r library, include=FALSE}
source(".Rprofile")
library("GenomicFeatures")
library("Gviz")
library("dplyr")
library("AnnotationDbi")
```


```{r load txdb}
txdb <- loadDb(snakemake@input[["txdb"]])
txdb_dump <- as.list(txdb)

txdb_flair <- loadDb(snakemake@input[["txdb_flair"]])
txdb_flair_dump <- as.list(txdb_flair)
```


```{r functions}
subset.txdb <- function(txdb, txdb_dump, gene_id) {

    all_tx <- AnnotationDbi::select(txdb, gene_id,
            c("GENEID", "TXNAME"), "GENEID") %>%
        pull(TXNAME)

    all_tx_ids <- txdb_dump[[1]] %>%
        as_tibble() %>%
        filter(tx_name %in% all_tx) %>%
        pull(tx_id)

    new_txdb_dump <- vector("list", 4)
    new_txdb_dump[1:3] <- txdb_dump[1:3] %>%
        purrr::map(filter, tx_id %in% all_tx_ids)
    new_txdb_dump[4] <- txdb_dump[4]

    do.call(makeTxDb, new_txdb_dump)
}


plot.transcripts <- function(gene_id, max_cov_h3k4 = 5, max_cov_cdna = 300,
    max_cov_illumina = 2000, extend = 2500) {

    gene_info <- AnnotationDbi::select(txdb, keys = gene_id,
            keytype = "GENEID",
            columns = c("TXCHROM", "TXSTART", "TXEND")) %>%
        as_tibble() %>%
        group_by(GENEID) %>%
        summarise(TXCHROM = unique(TXCHROM),
            TXSTART = min(TXSTART),
            TXEND = max(TXEND))

    txdb_flair_subset <- subset.txdb(txdb_flair, txdb_flair_dump, gene_id)
    
    flairtrack <- GeneRegionTrack(txdb_flair_subset,
        chromosome = gene_info$TXCHROM,
        from = gene_info$TXSTART, to = gene_info$TXEND,
        name = "flair cDNA")

    
    # scale
    gtrack <- GenomeAxisTrack(from = min(gene_info$TXSTART),
        to = max(gene_info$TXEND),
        scale = 0.25)


    # cDNA
    cdna_warm <- AlignmentsTrack(snakemake@input[["cdna_warm"]],
        isPaired = FALSE,
        genome = "mm10",
        name = "cDNA 22째C",
        ylim = c(0, max_cov_cdna),
        type = c("pileup"),
        lwd.sashimiMax = 2,
        sashimiHeight = .33,
        min.height = 2, # minimum height of a read in px, controls how many read are plotted
        chromosome = gene_info$TXCHROM,
        start = gene_info$TXSTART,
        end = gene_info$TXEND)

    # cdna_cold <- AlignmentsTrack(snakemake@input[["cdna_cold"]],
    #     isPaired = FALSE,
    #     genome = "mm10",
    #     name = "cDNA 4째C",
    #     ylim = c(0, max_cov_cdna),
    #     type = type,
    #     lwd.sashimiMax = 1,
    #     sashimiHeight = .33,
    #     chromosome = gene_info$TXCHROM,
    #     start = gene_info$TXSTART,
    #     end = gene_info$TXEND)
    
    
    # Illumina
    illumina_warm <- AlignmentsTrack(snakemake@input[["illumina_warm"]],
        isPaired = TRUE,
        genome = "mm10",
        name = "Illu 22째C",
        ylim = c(0, max_cov_illumina),
        type = c("coverage", "sashimi"),
        lwd.sashimiMax = 1,
        sashimiHeight = .33,
        transformation = function(x) log1p(x),
        chromosome = gene_info$TXCHROM,
        start = gene_info$TXSTART,
        end = gene_info$TXEND)

    illumina_cold <- AlignmentsTrack(snakemake@input[["illumina_cold"]],
        isPaired = TRUE,
        genome = "mm10",
        name = "Illu 4째C",
        ylim = c(0, max_cov_illumina),
        type = c("coverage", "sashimi"),
        lwd.sashimiMax = 1,
        sashimiHeight = .33,
        transformation = function(x) log1p(x),
        chromosome = gene_info$TXCHROM,
        start = gene_info$TXSTART,
        end = gene_info$TXEND)

    
    # ChIP seq
    ncd_warm_h3k4me3 <- DataTrack(snakemake@input[["ncd_warm_h3k4me3"]],
        genome = "mm10",
        name = "H3K4me3 22 NCD",
        ylim = c(0, max_cov_h3k4),
        chromosome = unique(gene_info$TXCHROM),
        start = min(gene_info$TXSTART) - extend,
        end = max(gene_info$TXEND) + extend,
        type = "histogram")

    ncd_cold_h3k4me3 <- DataTrack(snakemake@input[["ncd_cold_h3k4me3"]],
        genome = "mm10",
        name = "H3K4me3 4 NCD",
        ylim = c(0, max_cov_h3k4),
        chromosome = unique(gene_info$TXCHROM),
        start = min(gene_info$TXSTART) - extend,
        end = max(gene_info$TXEND) + extend,
        type = "histogram")
    
    # Plotting
    options(ucscChromosomeNames = FALSE)

    tracks <- list(gtrack,
#            illumina_warm, illumina_cold,
            cdna_warm,# cdna_cold,
            ncd_warm_h3k4me3, ncd_cold_h3k4me3,
            flairtrack)
    
    plotTracks(
        tracks,
        from = min(gene_info$TXSTART) - extend,
        to = max(gene_info$TXEND) + extend,
        extend.left = 250,
        extend.right = 250,
        background.title = "darkblue",
        cex = .75,
        cex.axis = .6, 
        sashimiFilter = unique(unlist(intronsByTranscript(txdb_flair_subset))), #intronicParts(txdb_flair_subset),
#        sashimiFilterTolerance = 5L,
        lwd = .25,
        sizes = c(.1, 
              rep(.5, 2),
              1,
              rep(.5, 2),
              rep(.25, 1)))
}
```

# Adcy3

```{r adcy3}
plot.transcripts(gene_id = "ENSMUSG00000020654.15", max_cov_cdna = 450,
    max_cov_illumina = 2000, max_cov_h3k4 = 5)
```

# Cars2

```{r cars2}
plot.transcripts("ENSMUSG00000056228.10", max_cov_cdna = 250,
    max_cov_illumina = 800, max_cov_h3k4 = 10, extend = 400)
```

