---
title: "Nanopore iBAT, Coverage"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
    html_document:
        toc: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, cache = FALSE)
```

```{r loadPackages}
source("packrat/init.R")
library("dplyr")
library("readr")
library("purrr")
library("ggplot2")
library("ggthemes")
library("AnnotationDbi")
```

```{r}
sample_info <- read_csv(snakemake@input[["sample_info"]]) %>% 
    filter(!is.na(ont)) %>%
    mutate_at(vars(matches("condition")), as.factor) %>% 
    dplyr::select(sample_id, illumina, ont)
```

# Gene body coverage

```{r dataImport_geneBodyCoverage}
read_geneBodyCoverage <- function(file) {
    file %>% 
        read_tsv() %>%
        tidyr::gather() %>% 
        magrittr::set_colnames(.[1, ]) %>% 
        `[`(-1, ) %>% 
        mutate_all(as.integer)
}

prepare_dataset = function(files, name) {
    files %>% 
        map(read_geneBodyCoverage) %>% 
        purrr::reduce(left_join, by = "Percentile") %>% 
        tidyr::gather("sample", "n", -Percentile) %>% 
        mutate(dataset = name)
}

illumina <- snakemake@input[["geneBodyCoverage_illumina"]] %>% 
    prepare_dataset(name = "illumina") %>% 
    mutate(sample = substr(sample, 2, 9)) %>% 
    left_join(sample_info, by = c("sample" = "illumina")) %>% 
    dplyr::select(-sample, -ont)
    
teloprime <- snakemake@input[["geneBodyCoverage_teloprime"]] %>% 
    prepare_dataset(name = "teloprime") %>% 
    mutate(sample = substr(sample, 1, 9)) %>% 
    left_join(sample_info, by = c("sample" = "ont")) %>% 
    dplyr::select(-sample, -illumina)

```

```{r plot_geneBodyCoverage}
list(illumina, teloprime) %>% 
    bind_rows() %>%
    group_by(dataset, sample_id) %>% 
    mutate(p = n / max(n)) %>% 
    ungroup() %>% 
    ggplot(aes(Percentile, p)) +
        geom_line(aes(colour = dataset, lty = sample_id)) +
        theme_tufte() +
        geom_rangeframe(sides = "bl") +
        xlab("Gene body percentile (5' -> 3')") +
        ylab("Coverage") +
        scale_color_viridis_d()

rm(list = c("illumina", "teloprime"))
```

# Coverage

```{r import_txdb}
biomart <- snakemake@input[["biomaRt_tx"]] %>% 
    read_rds()
```

```{r import_coverage}
get_map_type <- function(flag) {
    if_else(flag %in% c(0, 16), "primary",
        if_else(flag %in% c(256, 275), "secondary", "supplementary"))
}

prepare_dataset2 = function(files, name) {
    files %>% 
        as.list() %>% 
        set_names(basename(files)) %>% 
        map(read_rds) %>% 
        map(mutate, type = get_map_type(flag)) %>% 
        map(dplyr::select, -qname, -flag, -qwidth) %>% 
        map(tidyr::separate, col = seqnames, into = "ensembl_transcript_id_version",
            sep = "\\|", extra = "drop") %>% 
        map(left_join, y = biomart, by = "ensembl_transcript_id_version") %>% 
        bind_rows(.id = "sample") %>% 
        mutate(coverage = coverage/ transcript_length)
}

teloprime <- snakemake@input[["coverage_teloprime"]] %>% 
    prepare_dataset2(name = "teloprime")
```

```{r}
df <- teloprime %>% 
    filter(type == "primary") %>% 
    tidyr::drop_na() %>% 
    mutate(grouped_transcript_length = cut(transcript_length, 
        breaks = c(0, 500, 1000, 1500, 2000, 2500, 3000, max(transcript_length))))

df %>% 
    ggplot(aes(grouped_transcript_length, coverage)) +
    geom_violin() +
    xlab(NULL) +
    ylab("Fraction of transcript covered by read") +
    theme_tufte() +
    geom_rangeframe(sides = 'l')

df %>% 
    ggplot(aes(transcript_length, coverage)) +
    geom_hex(aes(fill = ..density..), binwidth = .025) +
    scale_x_log10() +
    scale_fill_viridis_c(trans = "log") +
    theme_tufte() +
    geom_rangeframe(sides = "bl") +
    xlab("Transcript length") + 
    ylab("Fraction of transcript covered by read")

df %>% 
    mutate(is_fullLength = if_else(coverage > .9, TRUE, FALSE)) %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(n = n(),
        n_fullLength = sum(is_fullLength),
        transcript_length = unique(transcript_length)) %>% 
    mutate(rate = n_fullLength / n) %>% 
    ggplot(aes(transcript_length, rate)) +
    geom_point() +
    scale_x_log10() +
    scale_y_continuous(trans = "logit",
        breaks = c(0, .0001, .001, .01, .1, .25, .5, .75, .9, .99, 1), 
        labels = function(x) signif(x, 1)) +
    theme_tufte() +
    geom_rangeframe(sides = "bl") +
    xlab("Transcript length") + 
    ylab("Fraction of full length reads")
```

