---
title: "Comparison of DGE, DTE and DTU"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
    html_document:
        toc: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, cache = FALSE)
```

```{r}
save.image("dge.RData")
```

```{r library, include=FALSE}
source("packrat/init.R")
library("dplyr")
library("readr")
library("purrr")
library("ggplot2")
library("UpSetR")
    select <- dplyr::select
    rename <- dplyr::rename
```

# Data Import

I use the illumina data from only those samples also analysed on the nanopore.

```{r dataImport}
df <- c(snakemake@input[["illumina_tx"]], 
        snakemake@input[["illumina_gene"]], 
        snakemake@input[["illumina_dtu"]],
        snakemake@input[["teloprime_tx"]], 
        snakemake@input[["teloprime_gene"]]) %>% 
    map(read_csv)
df[[6]] <- read_tsv(snakemake@input[["teloprime_dtu"]]) %>% 
    rename(mgi_symbol = "GeneSymbol",
        ensembl_gene_id_version = "geneID",
        ensembl_transcript_id_version = "txID")

teloprime_gene_counts <- read_csv(snakemake@input[["teloprime_counts"]])
illumina_gene_counts <- read_csv(snakemake@input[["illumina_counts"]])

biomart_gene <- read_rds(snakemake@input[["biomaRt_gene"]])
```

```{r mergeDatasets}
my_select <- function(x) {
    select_at(x, vars(one_of("ensembl_gene_id_version", "svalue", "gene", "transcript")))
}

summarise_txlevelDeseq <- function(x) {
    x %>% 
        group_by(ensembl_gene_id_version) %>% 
        summarise(svalue = min(svalue))
}

summarise_dtu <- function(x) {
    x %>% 
        group_by(ensembl_gene_id_version) %>% 
        summarise(gene = unique(gene),
            transcript = min(transcript))
}

df <- df %>% 
    map(my_select) %>% 
    set_names(c("illumina_tx", "illumina_gene", "illumina_dtu",
        "teloprime_tx", "teloprime_gene", "teloprime_dtu")) %>% 
    map_at(vars(matches("tx")), summarise_txlevelDeseq) %>% 
    map_at(vars(matches("dtu")), summarise_dtu) %>% 
    reduce(full_join, by = "ensembl_gene_id_version")
colnames(df) <- c("ensembl_gene_id_version", 
    "illumina_tx", "illumina_gene", "illumina_dtu_gene", "illumina_dtu_tx",
    "teloprime_tx", "teloprime_gene", "teloprime_dtu_gene", "teloprime_dtu_tx") 
```    

# Intersections

Analysis on gene level. The txlevel pvalues are the smallest for any of the transcripts
of a gene.

The ont DTE and DGE values have been caluclated with the same settings as used for
the illumina data.

## All genes

```{r upset}
df %>%
    mutate_if(is.numeric, function(x) tidyr::replace_na(x, 1)) %>% 
    mutate_if(is.numeric, function(x) as.integer(x < .05)) %>%
    tidyr::drop_na(ensembl_gene_id_version) %>%
    as.data.frame() %>% 
    magrittr::set_rownames(.$ensembl_gene_id_version) %>% 
    `[`(, -1) %>% 
    upset(sets = c("illumina_tx", "illumina_gene", "illumina_dtu_gene", "illumina_dtu_tx",
        "teloprime_tx", "teloprime_gene", "teloprime_dtu_gene", "teloprime_dtu_tx"),
        nsets = 8, keep.order = TRUE,
        sets.bar.color = rep(c("red", "blue"), each = 4),
        nintersects = 20,
        order.by = "freq")
```

The take home message is: There is not much overlap between the genes detected 
by ont and illumina.

## Only genes, that were detected by both datasets

```{r}
get_detectedTrancripts <- function(df, cutoff = 1) {
    df %>% 
        select(-mgi_symbol, -gene_name) %>%
        as.data.frame() %>%
        magrittr::set_rownames(.$gene_id_ens) %>%
        `[`( ,-1) %>%
        data.matrix() %>% 
        rowSums() %>% 
        (function(x) x >= cutoff) %>% 
        tibble::enframe() %>% 
        filter(value == TRUE) %>% 
        pull(name)
}
```

```{r}
detected_by_teloprime <- get_detectedTrancripts(teloprime_gene_counts, cutoff = 10)
detected_by_illumina <- get_detectedTrancripts(illumina_gene_counts, cutoff = 10)
```

```{r}
df %>%
    filter(ensembl_gene_id_version %in% detected_by_teloprime & 
        ensembl_gene_id_version %in% detected_by_illumina) %>% 
    mutate_if(is.numeric, function(x) tidyr::replace_na(x, 1)) %>% 
    mutate_if(is.numeric, function(x) as.integer(x < .05)) %>%
    tidyr::drop_na(ensembl_gene_id_version) %>%
    as.data.frame() %>% 
    magrittr::set_rownames(.$ensembl_gene_id_version) %>% 
    `[`(, -1) %>% 
    upset(sets = c("illumina_tx", "illumina_gene", "illumina_dtu_gene", "illumina_dtu_tx",
        "teloprime_tx", "teloprime_gene", "teloprime_dtu_gene", "teloprime_dtu_tx"),
        nsets = 8, keep.order = TRUE,
        sets.bar.color = rep(c("red", "blue"), each = 4),
        nintersects = 20,
        order.by = "freq")
```

What is different between those genes significant in teloprime, illumina or both?

```{r}
df2 <- df %>%
    filter(ensembl_gene_id_version %in% detected_by_teloprime & 
        ensembl_gene_id_version %in% detected_by_illumina) %>% 
    mutate_if(is.numeric, function(x) tidyr::replace_na(x, 1)) %>% 
    mutate_if(is.numeric, function(x) x < .05) %>%
    tidyr::drop_na(ensembl_gene_id_version) %>%
    rowwise() %>% 
    mutate(group = if_else(any(c(illumina_tx, illumina_gene) & any(teloprime_gene, teloprime_tx)), "both",
        if_else(any(c(illumina_tx, illumina_gene)), "illumina", 
            if_else(any(c(teloprime_tx, teloprime_gene)), "teloprime",
                "not")))) %>% 
    ungroup() %>% 
    select(ensembl_gene_id_version, group) %>% 
    mutate(group = as.factor(group))

df2$group %>%
    table()
```

```{r}
df2 <- df2 %>% 
    left_join(biomart_gene, by = "ensembl_gene_id_version")

df2 %>% 
    filter(group != "not") %>% 
    ggplot(aes(group, fill = gene_biotype)) +
    geom_bar(position = "fill")
```

Teloprime is picking up unprocessed pseudogenes (minimap2 likes to align without
splicing. Error prone reads, ...)

counts/tpm?
