---
title: "nanopore iBAT, Comparison of Illumina and ONT reads on transcript level"
author: "Christoph Kiefer"
date: "20. September 2019"
output: 
    html_document:
        toc: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE, cache=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, cache = FALSE)
```

```{r library, include=FALSE, cache=FALSE}
library("dplyr")
library("readr")
library("purrr")
library("ggplot2")
library("ggthemes")
library("cowplot")
library("ggpmisc")
    rename <- dplyr::rename
```

```{r dataImport_tx}
import_data <- function(illumina, ont, tpm) {
    sample_info <- read_csv("../sample_info/sampleInfo.csv")

    illumina <- read_csv(illumina) %>% 
        select(-gene_name, -mgi_symbol) %>% 
        tidyr::gather(key = "sample", value = "counts_illumina", -transcript_id_ens)
    ont <- read_tsv(ont) %>% 
        tidyr::separate(transcript, "transcript_id_ens", sep = '\\|')
    tpm <- read_csv(tpm) %>% 
        select(-gene_name, -mgi_symbol) %>% 
        tidyr::gather(key = "sample", value = "tpm", -transcript_id_ens)
    
    lookup <- sample_info$illumina
    names(lookup) <- sample_info$ont
    
    colnames(ont) <- lookup[colnames(ont)]
    colnames(ont)[1] <- "transcript_id_ens"
        
    illumina %>% 
        full_join(ont %>% tidyr::gather(key = "sample", value = "counts_ont", -transcript_id_ens),
            by = c("transcript_id_ens", "sample")) %>% 
        left_join(tpm, by = c("transcript_id_ens", "sample")) %>% 
        rename(ensembl_transcript_id_version = "transcript_id_ens")
}

illumina <- "../res/txlevel_ont/txlevel_ont_cm_rld.csv.gz"
ont <- "../res/wien/6samples/ChrKiefer_6samples_reg_log_transf_counts.tsv"
tpm <- "../res/txlevel_ont/txlevel_ont_cm_tpm.csv.gz"

df <- import_data(illumina, ont, tpm)
```

```{r dataImport_gene}
  illumina <- "../res/genelevel_ont/genelevel_ont_cm_rld.csv.gz"
  tpm <- "../res/genelevel_ont/genelevel_ont_cm_tpm.csv.gz"
  ont <- "../res/wien/6samples/ChrKiefer_6samples_raw_gene_counts.tsv"

import_data_genes <- function(tpm, ont, illumina) {

    sample_info <- read_csv("../sample_info/sampleInfo.csv")

    tpm <- read_csv(tpm) %>% 
        select(-gene_name, -mgi_symbol) %>% 
            tidyr::gather(key = "sample", value = "tpm", -gene_id_ens)
    
    illumina <- read_csv(illumina) %>% 
        select(-gene_name, -mgi_symbol) %>% 
        tidyr::gather(key = "sample", value = "counts_illumina", -gene_id_ens)

    ont <- read_tsv(ont) %>% 
        tidyr::separate(transcript, c("transcript_id_ens", "gene_id_ens"), sep = '\\|')
    
    lookup <- sample_info$illumina
    names(lookup) <- sample_info$ont
    
    colnames(ont) <- lookup[colnames(ont)]
    colnames(ont)[c(1, 2)] <- c("transcript_id_ens", "gene_id_ens")
    
    ont <- ont %>%
        tidyr::gather(key = "sample", value = "counts_ont",
            -transcript_id_ens, -gene_id_ens) %>% 
        group_by(gene_id_ens, sample) %>% 
        summarise(counts_ont = sum(counts_ont)) %>% 
        tidyr::spread(key = sample, value = counts_ont) %>% 
        filter(!is.na(gene_id_ens)) %>% 
        magrittr::set_rownames(.$gene_id_ens) %>% 
        as.matrix() %>% 
        `[`( ,-1)
    
    class(ont) <- "numeric"
    gene_ids <- rownames(ont)
    
    ont <- ont %>% 
        DESeq2::rlog() 
    
    ont <- ont %>% 
        as_tibble() %>% 
        mutate(gene_id_ens = gene_ids) %>% 
        tidyr::gather(key = "sample", value = "counts_ont",
            -gene_id_ens)
        
    tpm %>% 
        left_join(illumina, by = c("gene_id_ens", "sample")) %>% 
        full_join(ont, by = c("gene_id_ens", "sample")) %>% 
        mutate(tpm = log2(tpm + 1)) %>% 
        rename(ensembl_gene_id_version = "gene_id_ens")
}

df_genes <- import_data_genes(tpm, ont, illumina)
```

# Filtering

## Transcripts

### Before filtering

```{r densityBeforeFilter, dependson="dataImport_tx"}
pl1 <- df %>% 
    ggplot(aes(counts_illumina)) +
    geom_density(aes(colour = sample)) +
    scale_colour_viridis_d()
pl2 <- df %>% 
    ggplot(aes(counts_ont)) +
    geom_density(aes(colour = sample)) +
    scale_colour_viridis_d()

ggdraw(plot_grid(plot_grid(pl1 + theme(legend.position="none"), pl2 + theme(legend.position="none"), ncol=1, align='v'),
    plot_grid(NULL, get_legend(pl1), ncol=1),
    rel_widths=c(1, 0.2)))
```

Actually most transcripts are not expressed. 

```{r filterNotExpressed, echo = TRUE, dependson="dataImport_tx"}
# filter not expressed transcripts
not_expressed <- df %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(counts_illumina = sum(counts_illumina),
        counts_ont = sum(counts_ont)) %>% 
    filter(counts_illumina < 1 & counts_ont < 1) %>% 
    pull(ensembl_transcript_id_version)

df <- df %>% 
    filter(!(ensembl_transcript_id_version %in% not_expressed))
```

### Transcripts expressed in one of the datasets

```{r densityAfterFilter, dependson="filterNotExpressed"}
pl1 <- df %>% 
    ggplot(aes(counts_illumina)) +
    geom_density(aes(colour = sample)) +
    scale_colour_viridis_d()
pl2 <- df %>% 
    ggplot(aes(counts_ont)) +
    geom_density(aes(colour = sample)) +
    scale_colour_viridis_d()

ggdraw(plot_grid(plot_grid(pl1 + theme(legend.position="none"), pl2 + theme(legend.position="none"), ncol=1, align='v'),
    plot_grid(NULL, get_legend(pl1), ncol=1),
    rel_widths=c(1, 0.2)))
```

Filtering for genes that are not expressed in either of the datasets reveals,
that many transcripts observed in the illumina dataset are not observed in the
ont dataset.

```{r annotation, dependson="filterNotExpressed"}
# get annotation from biomart
ensembl = biomaRt::useMart("ensembl",dataset="mmusculus_gene_ensembl")
annotation <- biomaRt::getBM(
    attributes = c("ensembl_transcript_id_version",
        "ensembl_gene_id_version", "mgi_symbol", "description",
        "gene_biotype", "transcript_biotype", "transcript_length"), 
    filters = "ensembl_transcript_id_version", 
    values = unique(df$ensembl_transcript_id_version), 
    mart = ensembl)

biotype_groups <- read_csv("../data/biotype_groups.csv") %>% 
    rename(gene_biotype = transcript_biotype)

df <- df %>% 
    left_join(annotation, by = "ensembl_transcript_id_version") %>% 
    left_join(biotype_groups, by = "gene_biotype") %>% 
    rename(gene_biotype_group = "grouped_biotype") %>% 
    left_join(biotype_groups %>% rename(transcript_biotype = "gene_biotype"), by = "transcript_biotype") %>% 
    rename(transcript_biotype_group = "grouped_biotype")
    

dir.create("../res/comparison", showWarnings = FALSE)
df %>% 
    write_csv("../res/comparison/countMatrix_transcripts.csv.gz")
```

### Transcripts expressed in both datasets

```{r filterNotExpressed2, dependson="annotation"}
not_expressed <- df %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(counts_illumina = sum(counts_illumina),
        counts_ont = sum(counts_ont)) %>% 
    filter(counts_illumina < 1 | counts_ont < 1) %>% 
    pull(ensembl_transcript_id_version)

df_both <- df %>% 
    filter(!(ensembl_transcript_id_version %in% not_expressed))

n_tx_in_one <- length(unique(df$ensembl_transcript_id_version))
n_tx_in_both <- length(unique(df_both$ensembl_transcript_id_version))

cat(sprintf("%d transcripts are exressed in one dataset.
%d transcripts are expressed in both datasets.
Cutoff: sum of log2 counts over all replicates > 1.", n_tx_in_one, n_tx_in_both))

df_both %>% 
    distinct(ensembl_transcript_id_version, .keep_all = TRUE) %>% 
    pull(transcript_biotype_group) %>% 
    table()

pl1 <- df_both %>% 
    ggplot(aes(counts_illumina)) +
    geom_density(aes(colour = sample)) +
    scale_colour_viridis_d()
pl3 <- df_both %>% 
    ggplot(aes(log10(tpm))) +
    geom_density(aes(colour = sample)) +
    scale_colour_viridis_d()
pl2 <- df_both %>% 
    ggplot(aes(counts_ont)) +
    geom_density(aes(colour = sample)) +
    scale_colour_viridis_d()

ggdraw(plot_grid(plot_grid(pl1 + theme(legend.position="none"),
                           pl3 + theme(legend.position="none"),
                           pl2 + theme(legend.position="none"),
                           ncol=1, align='v'),
    plot_grid(NULL, get_legend(pl1), ncol=1),
    rel_widths=c(1, 0.2)))
```

## Genes

all values on log2 scale

### Before filtering

```{r densityBeforeFilter_genes, dependson="dataImport_gene"}
pl1 <- df_genes %>% 
    ggplot(aes(tpm)) +
    geom_density(aes(colour = sample)) +
    scale_colour_viridis_d()
pl2 <- df_genes %>% 
    ggplot(aes(counts_ont)) +
    geom_density(aes(colour = sample)) +
    scale_colour_viridis_d()

ggdraw(plot_grid(plot_grid(pl1 + theme(legend.position="none"), pl2 + theme(legend.position="none"),
            ncol=1, align = "v", axis = "none"),
        plot_grid(NULL, get_legend(pl1), ncol=1),
        rel_widths=c(1, 0.2)))
```

### Genes expressed in one dataset only

```{r filterNotExpressed_gene, echo = TRUE, dependson="dataImport_gene"}
# filter not expressed transcripts
not_expressed <- df_genes %>% 
    group_by(ensembl_gene_id_version) %>% 
    summarise(counts_illumina = sum(counts_illumina),
        counts_ont = sum(counts_ont)) %>% 
    filter(counts_illumina < 1 & counts_ont < 1) %>% 
    pull(ensembl_gene_id_version)

df_genes <- df_genes %>% 
    filter(!(ensembl_gene_id_version %in% not_expressed))
```

```{r densityFilter_genes, dependson="dataImport_gene"}
pl1 <- df_genes %>% 
    ggplot(aes(tpm)) +
    geom_density(aes(colour = sample)) +
    scale_colour_viridis_d()
pl2 <- df_genes %>% 
    ggplot(aes(counts_ont)) +
    geom_density(aes(colour = sample)) +
    scale_colour_viridis_d()

ggdraw(plot_grid(plot_grid(pl1 + theme(legend.position="none"), pl2 + theme(legend.position="none"),
            ncol=1, align = "v", axis = "none"),
        plot_grid(NULL, get_legend(pl1), ncol=1),
        rel_widths=c(1, 0.2)))
```

### Genes expressed in both datasets

```{r annotateGenes}
annotation_genes <- biomaRt::getBM(
    attributes = c("ensembl_gene_id_version", "mgi_symbol", "description", 
        "gene_biotype"), 
    filters = "ensembl_gene_id_version", 
    values = unique(df_genes$ensembl_gene_id_version), 
    mart = ensembl)

biotype_groups <- read_csv("../data/biotype_groups.csv") %>% 
    rename(gene_biotype = "transcript_biotype")

df_genes <- df_genes %>% 
    left_join(annotation_genes, by = "ensembl_gene_id_version") %>% 
    left_join(biotype_groups, by = "gene_biotype") %>% 
    rename(gene_biotype_group = "grouped_biotype")

df_genes %>% 
    write_csv("../res/comparison/countMatrix_genes.csv.gz")
```


```{r filterNotExpressed2_genes, dependson="annotateGenes"}
not_expressed <- df_genes %>% 
    group_by(ensembl_gene_id_version) %>% 
    summarise(counts_illumina = sum(counts_illumina),
        counts_ont = sum(counts_ont)) %>% 
    filter(counts_illumina < 1 | counts_ont < 1) %>% 
    pull(ensembl_gene_id_version)

df_genes_both <- df_genes %>% 
    filter(!(ensembl_gene_id_version %in% not_expressed))

n_genes_in_one <- length(unique(df_genes$ensembl_gene_id_version))
n_genes_in_both <- length(unique(df_genes_both$ensembl_gene_id_version))

cat(sprintf("%d genes are exressed in one dataset.
%d genes are expressed in both datasets.
Cutoff: sum of log2 counts over all replicates > 1.", n_genes_in_one, n_genes_in_both))

df_genes_both %>% 
    distinct(ensembl_gene_id_version, .keep_all = TRUE) %>% 
    pull(gene_biotype_group) %>% 
    table()

pl1 <- df_genes_both %>% 
    ggplot(aes(tpm)) +
    geom_density(aes(colour = sample)) +
    scale_colour_viridis_d()
pl2 <- df_genes_both %>% 
    ggplot(aes(counts_ont)) +
    geom_density(aes(colour = sample)) +
    scale_colour_viridis_d()

ggdraw(plot_grid(plot_grid(pl1 + theme(legend.position="none"),
                           pl2 + theme(legend.position="none"),
                           ncol=1, align='v'),
    plot_grid(NULL, get_legend(pl1), ncol=1),
    rel_widths=c(1, 0.2)))
```

# Correlation

## Transcripts expressed in one of the datasets

### Illumina counts (rlog)

```{r correlationWithoutSamples_counts, dependson="annotation"}
df %>% 
    distinct(ensembl_transcript_id_version, .keep_all = TRUE) %>% 
    pull(transcript_biotype_group) %>% 
    table()

my_breaks <- function (n = 5, ...) {

    function(x) {
        dings <- .045 * (max(x) - min(x))
        extended_range_breaks_(min(x) + dings , max(x) - dings, n, ...)
    }
}

pl1 <- df %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(counts_illumina = sum(counts_illumina),
        counts_ont = sum(counts_ont),
        gene_biotype_group = unique(gene_biotype_group)) %>% 
    filter(complete.cases(.)) %>% 
    ggplot(aes(counts_illumina, counts_ont)) +
        geom_point(aes(colour = gene_biotype_group)) +
        geom_smooth(colour = "red") +
        geom_abline(slope = 1) +
        theme_tufte() +
        geom_rangeframe() +
        scale_y_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_x_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_colour_viridis_d()

pl2 <- df %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(counts_illumina = sum(counts_illumina),
        counts_ont = sum(counts_ont),
        gene_biotype_group = unique(gene_biotype_group)) %>% 
    filter(complete.cases(.)) %>% 
    ggplot(aes(counts_illumina, counts_ont, colour = gene_biotype_group)) +
        geom_point() +
        geom_smooth(colour = "red") +
        geom_abline(slope = 1) +
        theme_tufte() +
        geom_rangeframe(colour = "black") +
        facet_wrap(~gene_biotype_group, ncol = 3) +
        scale_y_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_x_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_color_viridis_d()

plot_grid(pl1 + theme(legend.position="none"),
        pl2 + theme(legend.position="none"), 
        get_legend(pl1),
    ncol = 3, rel_widths = c(1,1,.3))
   
pl1 <- df %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(counts_illumina = sum(counts_illumina),
        counts_ont = sum(counts_ont),
        transcript_biotype_group = unique(transcript_biotype_group)) %>% 
    filter(complete.cases(.)) %>% 
    ggplot(aes(counts_illumina, counts_ont)) +
        geom_point(aes(colour = transcript_biotype_group)) +
        geom_smooth(colour = "red") +
        geom_abline(slope = 1) +
        theme_tufte() +
        geom_rangeframe() +
        scale_y_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_x_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_colour_viridis_d()

pl2 <- df %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(counts_illumina = sum(counts_illumina),
        counts_ont = sum(counts_ont),
        transcript_biotype_group = unique(transcript_biotype_group)) %>% 
    filter(complete.cases(.)) %>% 
    ggplot(aes(counts_illumina, counts_ont, colour = transcript_biotype_group)) +
        geom_point() +
        geom_smooth(colour = "red") +
        geom_abline(slope = 1) +
        theme_tufte() +
        geom_rangeframe(colour = "black") +
        facet_wrap(~transcript_biotype_group, ncol = 3) +
        scale_y_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_x_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_color_viridis_d()

plot_grid(pl1 + theme(legend.position="none"),
        pl2 + theme(legend.position="none"), 
        get_legend(pl1),
    ncol = 3, rel_widths = c(1,1,.3))
```

### TPM

```{r correlationWithoutSamples_tpm, dependson="annotation"}
pl1 <- df %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(tpm = sum(tpm),
        counts_ont = sum(counts_ont),
        gene_biotype_group = unique(gene_biotype_group)) %>% 
    filter(complete.cases(.)) %>% 
    ggplot(aes(log10(tpm), counts_ont)) +
        geom_point(aes(colour = gene_biotype_group)) +
        geom_smooth(colour = "red") +
        geom_abline(slope = 20) +
        theme_tufte() +
        geom_rangeframe() +
        scale_y_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_x_continuous(breaks = log10(c(1,100,10000,max(df$tpm))),
            labels = function(x) round(10^x),
            name = "tpm") +
        scale_colour_viridis_d()

pl2 <- df %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(tpm = sum(tpm),
        counts_ont = sum(counts_ont),
        gene_biotype_group = unique(gene_biotype_group)) %>% 
    filter(complete.cases(.)) %>% 
    ggplot(aes(log10(tpm), counts_ont, colour = gene_biotype_group)) +
        geom_point() +
        geom_smooth(colour = "red") +
        geom_abline(slope = 20) +
        theme_tufte() +
        geom_rangeframe(colour = "black") +
        facet_wrap(~gene_biotype_group, ncol = 3) +
        scale_y_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_x_continuous(breaks = log10(c(1,100,10000,max(df$tpm))),
            labels = function(x) round(10^x),
            name = "tpm") +
        scale_color_viridis_d()

plot_grid(pl1 + theme(legend.position="none"),
        pl2 + theme(legend.position="none"), 
        get_legend(pl1),
    ncol = 3, rel_widths = c(1,1,.3))

pl1 <- df %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(tpm = sum(tpm),
        counts_ont = sum(counts_ont),
        transcript_biotype_group = unique(transcript_biotype_group)) %>% 
    filter(complete.cases(.)) %>% 
    ggplot(aes(log10(tpm), counts_ont)) +
        geom_point(aes(colour = transcript_biotype_group)) +
        geom_smooth(colour = "red") +
        geom_abline(slope = 20) +
        theme_tufte() +
        geom_rangeframe() +
        scale_y_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_x_continuous(breaks = log10(c(1,100,10000,max(df$tpm))),
            labels = function(x) round(10^x),
            name = "tpm") +
        scale_colour_viridis_d()

pl2 <- df %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(tpm = sum(tpm),
        counts_ont = sum(counts_ont),
        transcript_biotype_group = unique(transcript_biotype_group)) %>% 
    filter(complete.cases(.)) %>% 
    ggplot(aes(log10(tpm), counts_ont, colour = transcript_biotype_group)) +
        geom_point() +
        geom_smooth(colour = "red") +
        geom_abline(slope = 20) +
        theme_tufte() +
        geom_rangeframe(colour = "black") +
        facet_wrap(~transcript_biotype_group, ncol = 3) +
        scale_y_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_x_continuous(breaks = log10(c(1,100,10000,max(df$tpm))),
            labels = function(x) round(10^x),
            name = "tpm") +
        scale_color_viridis_d()

plot_grid(pl1 + theme(legend.position="none"),
        pl2 + theme(legend.position="none"), 
        get_legend(pl1),
    ncol = 3, rel_widths = c(1,1,.3))
```

The red line does not really represent the data, but I think it shows some feature
of the data: Transcripts with low expression in the illumina dataset are not
picked up in the ont dataset. Of course stricter filtering might change this plot.

All the "counts" values so far are on log2 level! TPM makes sense.

## Transcripts expressed in both datasets

### Counts

```{r correlationWithoutSamples_counts2, dependson="filterNotExpressed2"}
# counts
pl1 <- df_both %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(counts_illumina = sum(counts_illumina),
        counts_ont = sum(counts_ont),
        gene_biotype_group = unique(gene_biotype_group)) %>% 
    filter(complete.cases(.)) %>% 
    ggplot(aes(counts_illumina, counts_ont)) +
        geom_point(aes(colour = gene_biotype_group)) +
        geom_smooth(colour = "red") +
        geom_abline(slope = 1) +
        theme_tufte() +
        geom_rangeframe() +
        scale_y_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_x_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_color_viridis_d()


pl2 <- df_both %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(counts_illumina = sum(counts_illumina),
        counts_ont = sum(counts_ont),
        gene_biotype_group = unique(gene_biotype_group)) %>% 
    filter(complete.cases(.)) %>% 
    ggplot(aes(counts_illumina, counts_ont, colour = gene_biotype_group)) +
        geom_point() +
        geom_smooth(colour = "red") +
        geom_abline(slope = 1) +
        theme_tufte() +
        geom_rangeframe(colour = "black") +
        facet_wrap(~gene_biotype_group, ncol = 3) +
        scale_y_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_x_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_color_viridis_d()

plot_grid(pl1 + theme(legend.position="none"),
        pl2 + theme(legend.position="none"), 
        get_legend(pl1),
    ncol = 3, rel_widths = c(1,1,.3))

pl1 <- df_both %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(counts_illumina = sum(counts_illumina),
        counts_ont = sum(counts_ont),
        transcript_biotype_group = unique(transcript_biotype_group)) %>% 
    filter(complete.cases(.)) %>% 
    ggplot(aes(counts_illumina, counts_ont)) +
        geom_point(aes(colour = transcript_biotype_group)) +
        geom_smooth(colour = "red") +
        geom_abline(slope = 1) +
        theme_tufte() +
        geom_rangeframe() +
        scale_y_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_x_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_colour_viridis_d()

pl2 <- df_both %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(counts_illumina = sum(counts_illumina),
        counts_ont = sum(counts_ont),
        transcript_biotype_group = unique(transcript_biotype_group)) %>% 
    filter(complete.cases(.)) %>% 
    ggplot(aes(counts_illumina, counts_ont, colour = transcript_biotype_group)) +
        geom_point() +
        geom_smooth(colour = "red") +
        geom_abline(slope = 1) +
        theme_tufte() +
        geom_rangeframe(colour = "black") +
        facet_wrap(~transcript_biotype_group, ncol = 3) +
        scale_y_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_x_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_color_viridis_d()

plot_grid(pl1 + theme(legend.position="none"),
        pl2 + theme(legend.position="none"), 
        get_legend(pl1),
    ncol = 3, rel_widths = c(1,1,.3))
```

### TPM

```{r correlationWithoutSamples_tpm2, dependson="filterNotExpressed2"}
pl1 <- df_both %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(tpm = sum(tpm),
        counts_ont = sum(counts_ont),
        gene_biotype_group = unique(gene_biotype_group)) %>% 
    filter(complete.cases(.)) %>% 
    ggplot(aes(log10(tpm), counts_ont)) +
        geom_point(aes(colour = gene_biotype_group)) +
        geom_smooth(colour = "red", method = "lm") +
        geom_abline(slope = 20) +
        theme_tufte() +
        geom_rangeframe() +
        scale_y_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_x_continuous(breaks = log10(c(1,100,10000,max(df$tpm))),
            labels = function(x) round(10^x),
            name = "tpm") +
        scale_color_viridis_d()

pl2 <- df_both %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(tpm = sum(tpm),
        counts_ont = sum(counts_ont),
        gene_biotype_group = unique(gene_biotype_group)) %>% 
    filter(complete.cases(.)) %>% 
    ggplot(aes(log10(tpm), counts_ont, colour = gene_biotype_group)) +
        geom_point() +
        geom_smooth(colour = "red", method = "lm") +
        geom_abline(slope = 20) +
        theme_tufte() +
        geom_rangeframe(colour = "black") +
        facet_wrap(~gene_biotype_group, ncol = 3) +
        scale_y_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_x_continuous(breaks = log10(c(1,100,10000,max(df$tpm))),
            labels = function(x) round(10^x),
            name = "tpm") +
        scale_color_viridis_d()

plot_grid(pl1 + theme(legend.position="none"),
        pl2 + theme(legend.position="none"), 
        get_legend(pl1),
    ncol = 3, rel_widths = c(1,1,.3))

pl1 <- df_both %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(tpm = sum(tpm),
        counts_ont = sum(counts_ont),
        transcript_biotype_group = unique(transcript_biotype_group)) %>% 
    filter(complete.cases(.)) %>% 
    ggplot(aes(log10(tpm), counts_ont)) +
        geom_point(aes(colour = transcript_biotype_group)) +
        geom_smooth(colour = "red", method = "lm") +
        geom_abline(slope = 20) +
        theme_tufte() +
        geom_rangeframe() +
        scale_y_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_x_continuous(breaks = log10(c(1,100,10000,max(df$tpm))),
            labels = function(x) round(10^x),
            name = "tpm") +
        scale_color_viridis_d()

pl2 <- df_both %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(tpm = sum(tpm),
        counts_ont = sum(counts_ont),
        transcript_biotype_group = unique(transcript_biotype_group)) %>% 
    filter(complete.cases(.)) %>% 
    ggplot(aes(log10(tpm), counts_ont, colour = transcript_biotype_group)) +
        geom_point() +
        geom_smooth(colour = "red", method = "lm") +
        geom_abline(slope = 20) +
        theme_tufte() +
        geom_rangeframe(colour = "black") +
        facet_wrap(~transcript_biotype_group, ncol = 3) +
        scale_y_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_x_continuous(breaks = log10(c(1,100,10000,max(df$tpm))),
            labels = function(x) round(10^x),
            name = "tpm") +
        scale_color_viridis_d()

plot_grid(pl1 + theme(legend.position="none"),
        pl2 + theme(legend.position="none"), 
        get_legend(pl1),
    ncol = 3, rel_widths = c(1,1,.3))
```

It seems to me, that many transcripts below log2(counts_illumina) = 30 do not
show any expression in the ont data set. These also seem to have a relatively high 
part of "processed others" transcripts.

Of course, small ncRNAs are not picked up in the ont dataset because of the 
missing polyA tail. The same is probably true for the "others" category. Pseudogenes
might theoretically catch some reads from related genes in the illumina dataset,
which should not be the case in the ont dataset. long ncRNAs are also not really 
picked up in the ont dataset.

```{r correlationSampleWise, dependson="filterNotExpressed2"}
df_both %>% 
    filter(complete.cases(.)) %>%
    ggplot(aes(counts_illumina, counts_ont)) +
        geom_hex() +
        geom_smooth(colour = "red") +
        geom_abline(slope = 1) +
        theme_tufte() +
        geom_rangeframe(colour = "black") +
        facet_wrap(~sample, ncol = 3) +
        scale_y_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_x_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_fill_viridis_c()

df_both %>% 
    filter(complete.cases(.)) %>%
    ggplot(aes(log10(tpm), counts_ont)) +
        geom_hex() +
        geom_smooth(colour = "red") +
        geom_abline(slope = 20) +
        theme_tufte() +
        geom_rangeframe(colour = "black") +
        facet_wrap(~sample, ncol = 3) +
        scale_y_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_x_continuous(breaks = log10(c(1,100,10000,max(df$tpm))),
            labels = function(x) round(10^x),
            name = "tpm") +
        scale_fill_viridis_c()
```

All the samples show the same behaviour.

## Genes expressed in one dataset

```{r}
df_genes %>% 
    filter(complete.cases(.)) %>%
    ggplot(aes(tpm, counts_ont)) +
        geom_point(aes(colour = gene_biotype_group)) +
        stat_poly_eq(aes(label = ..adj.rr.label..), formula = y ~ x, 
               parse = TRUE) +
        scale_y_continuous(breaks = log2(c(1, 10, 100, 1000, 10000, 100000)),
            labels = c(1, 10, 100, 1000, 10000, 100000)) +
        scale_x_continuous(breaks = log2(c(1, 10, 100, 1000, 10000, 100000)),
            labels = c(1, 10, 100, 1000, 10000, 100000)) +
        theme_tufte() +
        geom_rangeframe(colour = "black") +
        scale_colour_viridis_d()
```

## Genes expressed in both datasets

```{r}
df_genes_both %>% 
    filter(complete.cases(.)) %>%
    ggplot(aes(tpm, counts_ont)) +
        geom_point(aes(colour = gene_biotype_group)) +
        stat_poly_eq(aes(label = ..adj.rr.label..), formula = y ~ x, 
               parse = TRUE) +
        scale_y_continuous(breaks = log2(c(1, 10, 100, 1000, 10000, 100000)),
            labels = c(1, 10, 100, 1000, 10000, 100000)) +
        scale_x_continuous(breaks = log2(c(1, 10, 100, 1000, 10000, 100000)),
            labels = c(1, 10, 100, 1000, 10000, 100000)) +
        theme_tufte() +
        geom_rangeframe(colour = "black") +
        scale_colour_viridis_d()
```

what are these small ncRNAs?

```{r}
df_genes_both %>% 
    filter(gene_biotype_group == "small ncRNA") %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling() %>%
    kableExtra::scroll_box(width = "100%", height = "600px")
```

what about the pseudogenes on the right side?

```{r}
df_genes_both %>% 
    filter(gene_biotype_group == "pseudogenes") %>% 
    filter(tpm > log2(100)) %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling() %>%
    kableExtra::scroll_box(width = "100%", height = "600px")
```

Mainly mitochondrial genes. How do they get there? PolyA?

# Counts by length

Coding transcripts only

```{r countsByLength}
pl1 <- df_both %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(counts_illumina = sum(counts_illumina),
        counts_ont = sum(counts_ont),
        transcript_biotype_group = unique(transcript_biotype_group),
        transcript_length = unique(transcript_length)) %>% 
    filter(complete.cases(.)) %>% 
    filter(transcript_biotype_group == "coding") %>% 
    ggplot(aes(transcript_length, counts_illumina)) +
        geom_point() +
        scale_x_log10() +
        geom_smooth(method = "lm", colour = "red") +
        stat_poly_eq(formula = y ~ x,
            aes(label = paste(..eq.label.., ..rr.label.., sep = "~~~")),
            parse = TRUE, family = "serif", size = 3, colour = "grey30",
            label.x.npc = "left", label.y.npc = "top") +
        theme_tufte() +
        geom_rangeframe(colour = "black")

pl2 <- df_both %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(tpm = sum(tpm),
        counts_ont = sum(counts_ont),
        transcript_biotype_group = unique(transcript_biotype_group),
        transcript_length = unique(transcript_length)) %>% 
    filter(complete.cases(.)) %>% 
    filter(transcript_biotype_group == "coding") %>% 
    ggplot(aes(log10(transcript_length), log10(tpm))) +
        geom_point() +
        scale_x_continuous(breaks = log10(c(1000,10000,max(df_both$transcript_length))),
            labels = function(x) round(10^x),
            name = "length") +
        scale_y_continuous(breaks = log10(c(1, 10, 100, 1000 ,10000,max(df_both$tpm, na.rm = TRUE))),
            labels = function(x) round(10^x),
            name = "tpm") +
        geom_smooth(method = "lm", colour = "red") +
        stat_poly_eq(formula = y ~ x,
            aes(label = paste(..eq.label.., ..rr.label.., sep = "~~~")),
            parse = TRUE, family = "serif", size = 3, colour = "grey30",
            label.x.npc = "left", label.y.npc = "top") +
        theme_tufte() +
        geom_rangeframe(colour = "black")

pl3 <- df_both %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(counts_illumina = sum(counts_illumina),
        counts_ont = sum(counts_ont),
        transcript_biotype_group = unique(transcript_biotype_group),
        transcript_length = unique(transcript_length)) %>% 
    filter(complete.cases(.)) %>% 
    filter(transcript_biotype_group == "coding") %>% 
    ggplot(aes(transcript_length, counts_ont)) +
        geom_point() +
        scale_x_log10() +
        geom_smooth(method = "lm", colour = "red") +
        stat_poly_eq(formula = y ~ x,
            aes(label = paste(..eq.label.., ..rr.label.., sep = "~~~")),
            parse = TRUE, family = "serif", size = 3, colour = "grey30",
            label.x.npc = "left", label.y.npc = "top") +
        theme_tufte() +
        geom_rangeframe(colour = "black")

plot_grid(pl1, pl2, pl3,
    ncol = 2)

pl1 <- df_both %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(counts_illumina = sum(counts_illumina),
        counts_ont = sum(counts_ont),
        transcript_biotype_group = unique(transcript_biotype_group),
        transcript_length = unique(transcript_length)) %>% 
    filter(complete.cases(.)) %>% 
    filter(transcript_biotype_group == "coding") %>% 
    ggplot(aes(counts_illumina, counts_ont, colour = log2(transcript_length))) +
        geom_point() +
        geom_smooth(colour = "red") +
        geom_abline(slope = 1) +
        theme_tufte() +
        geom_rangeframe(colour = "black") +
        facet_wrap(~transcript_biotype_group, ncol = 3) +
        scale_y_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_x_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_color_viridis_c()

pl2 <- df_both %>% 
    group_by(ensembl_transcript_id_version) %>% 
    summarise(tpm = sum(tpm),
        counts_ont = sum(counts_ont),
        transcript_biotype_group = unique(transcript_biotype_group),
        transcript_length = unique(transcript_length)) %>% 
    filter(complete.cases(.)) %>% 
    filter(transcript_biotype_group == "coding") %>% 
    ggplot(aes(log10(tpm), counts_ont, colour = log2(transcript_length))) +
        geom_point() +
        #geom_smooth(colour = "red") +
        geom_quantile() +
        geom_abline(slope = 20) +
        theme_tufte() +
        geom_rangeframe(colour = "black") +
        facet_wrap(~transcript_biotype_group, ncol = 3) +
        scale_y_continuous(breaks = my_breaks(),
            labels = function(x) round(x, 1)) +
        scale_x_continuous(breaks = log10(c(1, 10, 100, 1000 ,10000,max(df_both$tpm, na.rm = TRUE))),
            labels = function(x) round(10^x),
            name = "tpm") +
        scale_color_viridis_c()

plot_grid(pl1 + theme(legend.position="none"),
        pl2 + theme(legend.position="none"), 
        get_legend(pl1),
    ncol = 3, rel_widths = c(1,1,.3))
```

