---
title: "Read lengths bam"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
    html_document:
        toc: true
        toc_float: true
        toc_depth: 2
        fig_caption: false
        dev: "svg"
editor_options:
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE,
    cache = FALSE, out.width = 800)
```


```{r loadPackages}
source(".Rprofile")
library("dplyr")
library("readr")
library("purrr")
library("ggplot2")
library("ggthemes")
library("scales")
```


```{r import_general}
sample_info <- read_csv(snakemake@input[["sample_info"]]) %>% 
    filter(!is.na(ont)) %>%
    mutate_at(vars(matches("condition")), as.factor)

sample_info2 <- sample_info %>% 
    dplyr::select(sample_id, illumina, ont, cdna) %>% 
    tidyr::gather("method", "barcode", -sample_id) %>% 
    dplyr::select(-method) 

biomart <- snakemake@input[["biomaRt_tx"]] %>%
    read_rds()
```

```{r dataImport_tx_bam}
df_tx <- read_rds(snakemake@input$collapsed_transcripts)

tpm <- snakemake@input[["avg_counts"]] %>% 
    read_csv() %>% 
    select(transcript_length, illumina) %>% 
    rename(qwidth = "transcript_length") %>%
    tidyr::drop_na() %>% 
    mutate(illumina = 2^illumina - 1) %>% 
    group_by(qwidth) %>% 
    summarise(n = sum(illumina, na.rm = TRUE)) %>% 
    ungroup() %>% 
    mutate(library = "illumina")
```

# Read categories (transcriptome mapping)

```{r, fig.width = 6.85 / 2, fig.asp = .65}
df_tx %>% 
    group_by(library, category) %>% 
    summarise(n = n()) %>% 
    ungroup() %>% 
    tidyr::pivot_wider(names_from = "category", values_from = "n") %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling()

df_tx %>% 
    filter(category != "supplementary") %>% 
    group_by(library, barcode, category) %>% 
    summarise(n = n()) %>% 
    left_join(sample_info2) %>%  
    mutate(n = n / sum(n)) %>% 
    ungroup() %>% 
    mutate(sample_id = if_else(library == "rna", barcode, sample_id)) %>% 
    ggplot(aes(x = sample_id, y = n,  fill = category)) +
    geom_bar(stat = "identity") +
    facet_wrap(~library, scales = "free_x") +
    scale_y_continuous(name = "Fraction of reads", labels = label_percent()) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "l") +
    scale_x_discrete(labels = NULL, breaks = NULL, name = NULL) +
    scale_fill_brewer(palette = "Dark2", name = NULL,
        labels = c("primary_with_supplementary" = "pri. + sup.",
                   "primary_wo_supplementary" = "pri. - sup.")) +
    theme(legend.key.size = unit(.33, "cm"))
```


# Length of primary and supplementary reads (against transcriptome)

```{r}
df_tx <- df_tx %>% 
    filter(type != "unmapped")
```


## Grouped read lengths

The illumina values are estimated based on the tpm values from salmon. They are
somewhat a reference, but not actual read lengths.

```{r, fig.width = 6.85 / 2, fig.asp = .65}
df_tx %>% 
    filter(type == "primary") %>% 
    group_by(library, qwidth) %>% 
    summarise(n = n()) %>% 
    bind_rows(tpm) %>% 
    mutate(weight = n / sum(n)) %>% 
    ungroup() %>% 
    mutate(library = factor(library, levels = c("illumina", "teloprime", "cdna", "rna"))) %>% 
    ggplot(aes(library, qwidth, weight = weight)) +
    geom_violin(lwd = .25) +
    scale_y_log10(labels = trans_format("log10", math_format(10^.x))) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "bl") +
    xlab(NULL) + 
    ylab("Read/transcript length") +
    theme(legend.key.size = unit(.33, "cm"))
```


```{r, fig.width = 6.85 / 2, fig.asp = .65}
df_tx %>% 
    ggplot(aes(category, qwidth)) + 
    geom_violin(aes(fill = library), lwd = .25) +
    scale_y_log10(labels = trans_format("log10", math_format(10^.x))) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "bl") +
    xlab(NULL) + 
    ylab("Read length") +
    scale_fill_brewer(palette = "Dark2", name = NULL,
        labels = c("primary_with_supplementary" = "pri. + sup.",
            "primary_wo_supplementary" = "pri. - sup.")) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5)) +
    theme(legend.key.size = unit(.33, "cm"))
```


```{r, fig.width = 6.85 / 2, fig.asp = .65}
df_tx %>% 
    ggplot(aes(category, aligned)) + 
    geom_violin(aes(fill = library), lwd = .25) +
    scale_y_log10(labels = trans_format("log10", math_format(10^.x))) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "bl") +
    xlab(NULL) + 
    ylab("Aligned read length") +
    scale_fill_brewer(palette = "Dark2", name = NULL,
        labels = c("primary_with_supplementary" = "pri. + sup.",
            "primary_wo_supplementary" = "pri. - sup.")) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5)) +
    theme(legend.key.size = unit(.33, "cm"))
```


# Mapping of supplementary reads

```{r, fig.width = 6.85 / 2, fig.asp = .65}
df_supplementary <- df_tx %>% 
    filter(category %in% c("primary_with_supplementary", "supplementary")) %>% 
    tidyr::separate(rname, c("ensembl_transcript_id_version", "ensembl_gene_id_version"),
        extra = "drop", sep = '\\|') %>% 
    group_by(library, barcode, qname) %>% 
    summarise(
        supplementary_type = 
            if_else(all(ensembl_transcript_id_version == ensembl_transcript_id_version[1]),
                "same_transcript", 
            if_else(all(ensembl_gene_id_version == ensembl_gene_id_version), "same_gene",
                "not_same_gene")),
        n = n()) %>% 
    ungroup() %>% 
    mutate(supplementary_type = as.factor(supplementary_type))

df_supplementary %>% 
    ggplot(aes(n - 1)) +
    geom_histogram(aes(fill = supplementary_type), binwidth = 1) +
    facet_wrap(~library, ncol = 1) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "bl") +
    scale_y_log10(name = "Count", labels = trans_format("log10", math_format(10^.x))) +
    xlab("Supplementary reads per primary read") +
    scale_fill_brewer(palette = "Dark2", name = NULL) +
    theme(legend.key.size = unit(.33, "cm"))
```


# Correlations

```{r}
df_tx <- df_tx %>% 
    filter(type == "primary") %>% 
    tidyr::separate(rname, "ensembl_transcript_id_version", sep = '\\|', extra = "drop") %>% 
    left_join(biomart %>% 
                  select(ensembl_transcript_id_version, transcript_length),
        by = "ensembl_transcript_id_version")
```


## Aligned length vs read length

Primary hits only

```{r plotTranscriptome, fig.width = 6.85 / 2, fig.asp = .65}
df_tx %>% 
    dplyr::select(library, qwidth, aligned) %>% 
    ggplot(aes(qwidth, aligned)) +
    geom_hex(aes(fill = ..density..), binwidth = .025) +
    facet_grid(~library) +
    scale_x_log10(labels = trans_format("log10", math_format(10^.x))) +
    scale_y_log10(labels = trans_format("log10", math_format(10^.x))) +
    scale_fill_viridis_c(trans = "log10", labels = trans_format("log10", math_format(10^.x))) +
    geom_abline(slope = 1, lwd = .25) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "bl") +
    xlab("Read Length") + 
    ylab("Aligned Length") +
    theme(legend.key.size = unit(.33, "cm"))
```

## Aligned length vs expected length

```{r, fig.width = 6.85 / 2, fig.asp = .65}
df_tx %>%
    dplyr::select(library, ensembl_transcript_id_version, aligne, transcript_length) %>%
    filter(!is.na(transcript_length)) %>%
    ggplot(aes(transcript_length, aligned)) +
    geom_hex(aes(fill = ..density..), binwidth = .025) +
    facet_grid(~library) +
    scale_x_log10(labels = trans_format("log10", math_format(10^.x))) +
    scale_y_log10(labels = trans_format("log10", math_format(10^.x))) +
    scale_fill_viridis_c(trans = "log10", labels = trans_format("log10", math_format(10^.x))) +
    geom_abline(slope = 1, lwd = .25) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "bl") +
    xlab("Transcript length") +
    ylab("Aligned length") +
    theme(legend.key.size = unit(.33, "cm"))
```

## Read length vs expected length

```{r, fig.width = 6.85 / 2, fig.asp = .65}
df_tx %>%
    dplyr::select(library, ensembl_transcript_id_version, qwidth, transcript_length) %>%
    filter(!is.na(transcript_length)) %>%
    ggplot(aes(transcript_length, qwidth)) +
    geom_hex(aes(fill = ..density..), binwidth = .025) +
    facet_grid(~library) +
    scale_x_log10(labels = trans_format("log10", math_format(10^.x))) +
    scale_y_log10(labels = trans_format("log10", math_format(10^.x))) +
    scale_fill_viridis_c(trans = "log10", labels = trans_format("log10", math_format(10^.x))) +
    geom_abline(slope = 1, lwd = .25) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "bl") +
    xlab("Transcript length") +
    ylab("Read length") +
    theme(legend.key.size = unit(.33, "cm"))
```
