---
title: "Nanopore iBAT, Read lengths etc"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
    html_document:
        toc: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, cache = FALSE)
```

```{r loadPackages}
source("packrat/init.R")
library("dplyr")
library("readr")
library("purrr")
library("ggplot2")
library("ggthemes")
library("AnnotationDbi")
```

```{r dataImport_tx_bam}
get_map_type <- function(flag) {
    if_else(flag %in% c(0, 16), "primary",
        if_else(flag %in% c(256, 275), "secondary", "supplementary"))
}

df_tx <- snakemake@input[["teloprime_bam_tx"]] %>%
    as.list() %>% 
    set_names(basename(snakemake@input[["teloprime_bam_tx"]])) %>% 
    map(read_rds) %>%     
    map(mutate, type = get_map_type(flag)) %>% 
    map(tidyr::separate, col = seqnames,
        into = c("ensembl_transcript_id_version", "ensembl_gene_id_version"),
        sep = '\\|', extra = "drop") %>% 
    bind_rows(.id = "sample")
```

```{r dataImport_fastqReadLengths}
# histogram of read lengths from fasta file
df <- read_csv(snakemake@input[["teloprime_readLengths"]]) %>%
    tidyr::gather("sample", "counts", -read_length) %>% 
    tidyr::separate(sample, c("waste1", "waste2", "flowcell", "waste3", "barcode"), 
        extra = "drop") %>% 
    select_at(vars(-matches("waste"))) 
```

```{r dataImport_referenceTxLengths}
referenceTxLengths <- read_csv(snakemake@input[["annotation_txLengths"]]) %>% 
    mutate(group = "gencode")
```

```{r dataImport_genome}
df_genome <- snakemake@input[["teloprime_bam_genome"]] %>%
    as.list() %>% 
    set_names(snakemake@input[["teloprime_bam_genome"]]) %>% 
    map(read_rds) %>%     
    map(mutate, type = get_map_type(flag)) %>% 
    map(dplyr::select, -seqnames) %>% 
    bind_rows(.id = "sample")

mapping_df_genome <- df_genome %>% 
    dplyr::select(qname, type) %>% 
    group_by(qname) %>% 
    summarise(has_primary = any(type == "primary"),
        has_supplementary = any(type == "supplementary"))

rm(df_genome)
```

# Length of reads per sample and flowcell

Read lengths from fastq files

```{r}
df %>% 
    group_by(flowcell, barcode) %>% 
    mutate(weight = counts / sum(counts)) %>% 
    ungroup() %>% 
    ggplot(aes(barcode, read_length, weight = weight, fill = flowcell)) +
    geom_violin() +
    scale_y_log10() +
    scale_fill_viridis_d() +
    theme_tufte() +
    geom_rangeframe(sides = "l") +
    xlab(NULL) + 
    ylab("Read length")
```

#### Compared with tx length from GENCODE annotation

```{r}
df %>% 
    group_by(read_length) %>% 
    summarise(n = sum(counts),
        group = "teloprime") %>%
    bind_rows(referenceTxLengths) %>%
    group_by(group) %>% 
    mutate(weight = n / sum(n)) %>% 
    ungroup() %>% 
    ggplot(aes(group, read_length, weight = weight)) +
    geom_violin() +
    scale_y_log10() +
    scale_fill_viridis_d() +
    theme_tufte() +
    geom_rangeframe(sides = "l") +
    xlab(NULL) + 
    ylab("Read/transcript length")
```

# Reads and mapping

```{r}
mapping_df_tx <- df_tx %>% 
    dplyr::select(qname, type) %>% 
    group_by(qname) %>% 
    summarise(has_primary = any(type == "primary"),
        has_supplementary = any(type == "supplementary"))

mapping_df2 <- mapping_df_tx %>% 
    left_join(mapping_df_genome, by = "qname", suffix = c("_tx", "_genome")) %>% 
    summarise(n_primary_genome = sum(has_primary_genome, na.rm = TRUE),
        n_supplementary_genome = sum(has_supplementary_genome, na.rm = TRUE),
        n_primary_transcriptome = sum(has_primary_tx, na.rm = TRUE),
        n_supplementary_transcriptome = sum(has_supplementary_tx, na.rm = TRUE)) %>% 
    tidyr::gather("type", "count") %>% 
    add_row(type = "n_total", count = sum(df$counts)) %>% 
    mutate(method = "teloprime") %>% 
    group_by(method) %>% 
    mutate(p = count/count[type == "n_total"]) %>% 
    ungroup()

mapping_df2 %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling()

mapping_df2 %>% 
    mutate(type = factor(type, 
        levels = c("n_total", "n_primary_genome", "n_supplementary_genome",
            "n_primary_transcriptome", "n_supplementary_transcriptome"))) %>% 
    ggplot(aes(x = method, y = count,  fill = type)) +
    geom_bar(stat = "identity", position=position_dodge()) +
    theme_tufte() +
    geom_rangeframe(sides = "l") +
    xlab(NULL) + 
    ylab("Reads") +
    scale_fill_viridis_d()

rm(mapping_df_genome)
```

## Length of primary and supplementary reads

From bam file.

```{r}
df_tx %>% 
    dplyr::select(-sample, -flag) %>% 
    left_join(mapping_df_tx, by = "qname") %>% 
    mutate(category = if_else(type == "supplementary", "supplementary",
        if_else(has_supplementary, "primary_with_supplementary", "primary_wo_supplementary"))) %>% 
    ggplot(aes(category, qwidth)) + 
    geom_violin() +
    scale_y_log10() +
    theme_tufte() +
    geom_rangeframe(sides = "l") +
    xlab(NULL) + 
    ylab("Read length")

df_tx %>% 
    dplyr::select(-sample, -flag) %>% 
    left_join(mapping_df_tx, by = "qname") %>% 
    mutate(category = if_else(type == "supplementary", "supplementary",
        if_else(has_supplementary, "primary_with_supplementary", "primary_wo_supplementary"))) %>% 
    ggplot(aes(category, aligned)) + 
    geom_violin() +
    scale_y_log10() +
    theme_tufte() +
    geom_rangeframe(sides = "l") +
    xlab(NULL) + 
    ylab("Aligned read length")
```

# Transcriptome

## Aligned length vs read length

Primary hits only

From bam file

```{r plotTranscriptome}
df_tx %>% 
    filter(type == "primary") %>% 
    ggplot(aes(qwidth, aligned)) +
    geom_hex(aes(fill = ..density..), binwidth = .025) +
    scale_x_log10() +
    scale_y_log10() +
    scale_fill_viridis_c(trans = "log") +
    geom_abline(slope = 1) +
    theme_tufte() +
    geom_rangeframe(sides = "bl") +
    xlab("Read Length") + 
    ylab("Aligned Length") 
```

## Aligned length vs expected length ()

```{r import_txdb}
biomart <- snakemake@input[["biomaRt_tx"]] %>% 
    read_rds()
```

```{r}
df_tx %>% 
    filter(type == "primary") %>% 
    dplyr::select(ensembl_transcript_id_version, aligned) %>% 
    left_join(biomart, by = "ensembl_transcript_id_version") %>% 
    filter(!is.na(transcript_length)) %>% 
    ggplot(aes(transcript_length, aligned)) +
    geom_hex(aes(fill = ..density..), binwidth = .025) +
    scale_x_log10() +
    scale_y_log10() +
    scale_fill_viridis_c(trans = "log") +
    geom_abline(slope = 1) +
    theme_tufte() +
    geom_rangeframe(sides = "bl") +
    xlab("Transcript length") + 
    ylab("Aligned length")
```

## Read length vs expected length ()

```{r}
df_tx %>% 
    filter(type == "primary") %>% 
    dplyr::select(ensembl_transcript_id_version, qwidth) %>% 
    left_join(biomart, by = "ensembl_transcript_id_version") %>% 
    filter(!is.na(transcript_length)) %>% 
    ggplot(aes(transcript_length, qwidth)) +
    geom_hex(aes(fill = ..density..), binwidth = .025) +
    scale_x_log10() +
    scale_y_log10() +
    scale_fill_viridis_c(trans = "log") +
    geom_abline(slope = 1) +
    theme_tufte() +
    geom_rangeframe(sides = "bl") +
    xlab("Transcript length") + 
    ylab("Read length")
```
