---
title: "Nanopore iBAT, Coverage"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
    html_document:
        toc: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, cache = FALSE)
```

```{r}
dir.create(snakemake@params[["fig_folder"]],
    recursive = TRUE, showWarnings = FALSE)
```

```{r loadPackages}
source(".Rprofile")
library("dplyr")
library("readr")
library("purrr")
library("ggplot2")
library("ggthemes")
library("ggpmisc")
library("quantreg")
library("AnnotationDbi")
```

```{r}
sample_info <- read_csv(snakemake@input[["sample_info"]]) %>%
    filter(!is.na(ont)) %>%
    mutate_at(vars(matches("condition")), as.factor) %>%
    dplyr::select(sample_id, illumina, ont, cdna)
```

# Gene body coverage

```{r dataImport_geneBodyCoverage}
read_geneBodyCoverage <- function(file) {
    file %>%
        read_tsv() %>%
        tidyr::gather() %>%
        magrittr::set_colnames(.[1, ]) %>%
        `[`(-1, ) %>%
        mutate_all(as.integer)
}

prepare_dataset = function(files, name) {
    files %>%
        map(read_geneBodyCoverage) %>%
        purrr::reduce(left_join, by = "Percentile") %>%
        tidyr::gather("sample", "n", -Percentile) %>%
        mutate(dataset = name)
}

illumina <- snakemake@input[["geneBodyCoverage_illumina"]] %>%
    prepare_dataset(name = "illumina") %>%
    mutate(sample = substr(sample, 2, 9)) %>%
    left_join(sample_info, by = c("sample" = "illumina")) %>%
    dplyr::select(-sample, -ont)

teloprime <- snakemake@input[["geneBodyCoverage_teloprime"]] %>%
    prepare_dataset(name = "teloprime") %>%
    mutate(sample = substr(sample, 1, 9)) %>%
    left_join(sample_info, by = c("sample" = "ont")) %>%
    dplyr::select(-sample, -illumina)

cdna <- snakemake@input[["geneBodyCoverage_cdna"]] %>%
    prepare_dataset(name = "cdna") %>%
    mutate(sample = substr(sample, 1, 9)) %>%
    left_join(sample_info, by = c("sample" = "cdna")) %>%
    dplyr::select(-sample, -illumina)
```

```{r plot_geneBodyCoverage}
pl <- list(illumina, teloprime, cdna) %>%
    bind_rows() %>%
    group_by(dataset, sample_id) %>%
    mutate(p = n / max(n)) %>%
    ungroup() %>%
    ggplot(aes(Percentile, p)) +
        geom_line(aes(colour = dataset, lty = sample_id)) +
        theme_tufte() +
        geom_rangeframe(sides = "bl") +
        xlab("Gene body percentile (5' -> 3')") +
        ylab("Coverage") +
        scale_colour_colorblind()
pl
saveRDS(pl,
    file.path(snakemake@params[["fig_folder"]], "geneBodyCoverage.rds"))
rm(list = c("illumina", "teloprime", "cdna"))
```

# Coverage

```{r import_txdb}
biomart <- snakemake@input[["biomaRt_tx"]] %>%
    read_rds()
```

```{r import_coverage}
get_map_type <- function(flag) {
    if_else(flag %in% c(0, 16), "primary",
        if_else(flag %in% c(256, 275), "secondary", "supplementary"))
}

prepare_dataset2 = function(files) {
    files %>%
        as.list() %>%
        set_names(basename(files)) %>%
        map(read_rds) %>%
        map(mutate, type = get_map_type(flag)) %>%
        map(dplyr::select, -qname, -flag, -qwidth) %>%
        map(tidyr::separate, col = seqnames, into = "ensembl_transcript_id_version",
            sep = "\\|", extra = "drop") %>%
        map(left_join, y = biomart, by = "ensembl_transcript_id_version") %>%
        bind_rows(.id = "sample") %>%
        mutate(coverage = coverage / transcript_length)
}

df <- list(snakemake@input[["coverage_teloprime"]],
        snakemake@input[["coverage_cdna"]]) %>%
    map(prepare_dataset2) %>%
    map(dplyr::select, coverage, transcript_length, type, ensembl_transcript_id_version) %>%
    map(filter, type == "primary") %>%
    set_names(c("teloprime", "cdna")) %>%
    bind_rows(.id = "dataset") %>%
    tidyr::drop_na()
```

```{r}
pl <- df %>%
    mutate(grouped_transcript_length = cut(transcript_length,
        breaks = c(0, 500, 1000, 1500, 2000, 2500, 3000, max(transcript_length)))) %>%
    ggplot(aes(grouped_transcript_length, coverage)) +
    geom_violin(aes(fill = dataset)) +
    xlab(NULL) +
    ylab("Fraction of transcript covered by read") +
    theme_tufte() +
    geom_rangeframe(sides = 'l') +
    scale_fill_colorblind() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1))

pl
saveRDS(pl,
    file.path(snakemake@params[["fig_folder"]], "coverage_by_transcriptLength.rds"))
```


```{r}
df %>%
    mutate(is_fullLength = if_else(coverage > .9, TRUE, FALSE)) %>%
    group_by(dataset, ensembl_transcript_id_version) %>%
    summarise(n = n(),
        n_fullLength = sum(is_fullLength),
        transcript_length = unique(transcript_length)) %>%
    mutate(rate = n_fullLength / n) %>%
    ggplot(aes(transcript_length, rate)) +
    geom_hex(aes(fill = ..density..), bins = 50) +
    geom_smooth(method = "lm", colour = "black", size = .5, se = FALSE) +
    geom_quantile(colour = "red", size = .5) +
    stat_poly_eq(formula = y ~ x,
        aes(label = paste(..rr.label.., sep = "~~~")),
        parse = TRUE, family = "serif", size = 3, colour = "grey30",
        label.x.npc = "left", label.y.npc = "top") +
    facet_wrap(~dataset) +
    scale_x_log10() +
    scale_y_continuous(trans = "logit",
        breaks = c(0, .0001, .001, .01, .1, .25, .5, .75, .9, .99, 1),
        labels = function(x) signif(x, 1)) +
    theme_tufte() +
    geom_rangeframe(sides = "bl") +
    xlab("Transcript length") +
    ylab("Fraction of full length reads") +
    scale_fill_viridis_c()
```
