---
title: "Nanopore iBAT, Feature detection"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
    html_document:
        toc: true
        toc_float: true
        toc_depth: 2
        fig_caption: false
        dev: "svg"
editor_options:
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE,
    cache = FALSE, out.width = 800)
```

```{r loadPackages}
source(".Rprofile")
library("dplyr")
library("readr")
library("ggplot2")
library("ggthemes")
library("purrr")
library("UpSetR")
library("scales")
    rename <- dplyr::rename
    select <- dplyr::select
```

```{r importData}
# biomart
biomaRt_tx <- read_rds(snakemake@input[["biomaRt_tx"]])
biomaRt_gene <- read_rds(snakemake@input[["biomaRt_gene"]])
biotype_groups <- read_csv(snakemake@input[["biotype_groups"]])

# sample info
sample_info <- read_csv(snakemake@input[["sample_info"]]) %>%
    mutate_at(vars(matches("condition")), as.factor)

# gene level
df_gene <- snakemake@input[["gene_counts"]] %>%
    set_names(basename(.)) %>%
    map(read_csv, show_col_types = FALSE) %>%
    map(select, -mgi_symbol, -description, -gene_biotype) %>%
    map(tidyr::gather, key = "sample_id", value = "counts", -ensembl_gene_id_version) %>%
    bind_rows(.id = "dataset") %>%
    tidyr::separate(dataset, "dataset", extra = "drop")

df_gene_sum <- df_gene %>%
    group_by(dataset, ensembl_gene_id_version) %>%
    summarise(sum_counts = sum(counts)) %>%
    left_join(biomaRt_gene, by = "ensembl_gene_id_version") %>%
    left_join(biotype_groups, by = c("gene_biotype" = "transcript_biotype"))

gene_tpm <- read_csv(snakemake@input[["gene_tpm"]], show_col_types = FALSE) %>%
    select(-mgi_symbol, -description, - gene_biotype) %>%
    tidyr::gather(key = "sample_id", value = "counts", -ensembl_gene_id_version) %>%
    group_by(ensembl_gene_id_version) %>%
    summarise(average_tpm = mean(counts))

# tx level
df_tx <- snakemake@input[["tx_counts"]] %>%
    set_names(basename(.)) %>%
    map(read_csv, show_col_types = FALSE) %>%
    map(select, -mgi_symbol, -description, -gene_biotype, -transcript_biotype, -transcript_length) %>%
    map(tidyr::gather, key = "sample_id", value = "counts",
        -ensembl_gene_id_version, -ensembl_transcript_id_version) %>%
    bind_rows(.id = "dataset") %>%
    tidyr::separate(dataset, "dataset", extra = "drop")

df_tx_sum <- df_tx %>%
    group_by(dataset, ensembl_transcript_id_version) %>%
    summarise(sum_counts = sum(counts)) %>%
    left_join(biomaRt_tx, by = "ensembl_transcript_id_version") %>%
    left_join(biotype_groups, by = "transcript_biotype")
```

# Number of counts

The counts are taken from the DESeq raw count files.

```{r mappedReads, fig.width = 6.85 / 2, fig.asp = .7}
df_gene %>%
    group_by(dataset, sample_id) %>%
    summarise(tot_counts = sum(counts)) %>%
    left_join(sample_info %>% select(sample_id, condition_temp),
        by = "sample_id") %>%
    mutate(dataset = factor(dataset, levels = c("illumina", "teloprime", "cdna", "rna"))) %>% 
    mutate(condition_temp = relevel(as.factor(condition_temp), "22")) %>% 
    ggplot(aes(dataset, tot_counts)) +
    geom_point(aes(colour = condition_temp),
        size = 1,
        alpha = .5,
        position = position_dodge2(width = .75)) +
    stat_summary(aes(colour = condition_temp),
        fun = mean, geom = "crossbar",
        position = position_dodge(width = .75),
        width = 1,
        fatten = 1, lwd = .5) +
    stat_summary(aes(colour = condition_temp),
        geom = "errorbar", fun.data = mean_se,
        width = .75,
        lwd = .25,
        position = position_dodge(width = .75)) +
    scale_y_log10(labels = label_scientific()) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    xlab(NULL) +
    ylab("Total counts") +
    scale_colour_brewer(palette = "Dark2", name = NULL,
        labels = c("22" = "22°C", "4" = "4°C")) +
    theme(legend.key.size = unit(.33, "cm"))

df_gene %>%
    group_by(dataset, sample_id) %>%
    summarise(tot_counts = sum(counts - 1)) %>%
    left_join(sample_info %>% select(sample_id, condition_temp),
        by = "sample_id") %>%
    tidyr::spread(dataset, tot_counts) %>%
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling()
```

Sophia reported the difference in the library size between the two conditions in
teloprime dataset before.

Comparing to the read numbers from teloprime, it seems, there are more counts
then reads. This is probably caused by supplementary mappings.

# Saturation curve

```{r}
get_numberDetectedFeatures <- function(cm, prop, cutoff = 1) {
    cm %>%
        DropletUtils::downsampleMatrix(prop = prop, bycol = TRUE) %>%
        Matrix::rowSums() %>%
        (function(x) x >= cutoff) %>%
        sum()
}
```

Cutoff is 1 read in any of the librarys per method. (Maybe it should be redone
taking replicates, condition etc. into account)

```{r saturationCurve, fig.width = 6.85 / 2, fig.asp = .7}
fractions <- 10^seq(-4, 0, length.out = 25)

cm_gene <- snakemake@input[["gene_counts"]] %>%
    set_names(basename(.)) %>%
    map(read_csv, show_col_types = FALSE) %>%
    map(select, -mgi_symbol, -description, -gene_biotype) %>%
    map(tibble::column_to_rownames, "ensembl_gene_id_version") %>%
    map(data.matrix)

cm_tx <- snakemake@input[["tx_counts"]] %>%
    set_names(basename(.)) %>%
    map(read_csv, show_col_types = FALSE) %>%
    map(select, -mgi_symbol, -description, -gene_biotype,
        -ensembl_gene_id_version, -transcript_biotype, -transcript_length) %>%
    map(tibble::column_to_rownames, "ensembl_transcript_id_version") %>%
    map(data.matrix)

df_features <- tibble(fraction = fractions) %>%
    rowwise() %>%
    mutate(teloprime_tx = get_numberDetectedFeatures(cm_tx$teloprime_txlevel_cm_cts.csv.gz, prop = fraction),
        illumina_tx = get_numberDetectedFeatures(cm_tx$illumina_txlevel_cm_cts.csv.gz, prop = fraction),
        cdna_tx = get_numberDetectedFeatures(cm_tx$cdna_txlevel_cm_cts.csv.gz, prop = fraction),
        rna_tx = get_numberDetectedFeatures(cm_tx$rna_txlevel_cm_cts.csv.gz, prop = fraction),
        teloprime_gene = get_numberDetectedFeatures(cm_gene$teloprime_genelevel_cm_cts.csv.gz, prop = fraction),
        illumina_gene = get_numberDetectedFeatures(cm_gene$illumina_genelevel_cm_cts.csv.gz, prop = fraction),
        cdna_gene = get_numberDetectedFeatures(cm_gene$cdna_genelevel_cm_cts.csv.gz, prop = fraction),
        rna_gene = get_numberDetectedFeatures(cm_gene$rna_genelevel_cm_cts.csv.gz, prop = fraction))

tot_counts <- list(df_gene, df_tx) %>%
    set_names("gene", "tx") %>%
    bind_rows(.id = "level") %>%
    group_by(level, dataset) %>%
    summarise(tot_counts = sum(counts)) %>%
    tidyr::unite(library, dataset, level)

df_features <- df_features %>%
    tidyr::gather(key = "library", value = "features", -fraction) %>%
    left_join(tot_counts, by = "library") %>%
    mutate(reads = fraction * tot_counts) %>%
    tidyr::separate(library, c("method", "feature_type"))

df_features %>%
    mutate(method = factor(method, levels = c("illumina", "teloprime", "cdna", "rna"))) %>% 
    mutate(feature_type = if_else(feature_type == "tx", "transcripts", "genes")) %>% 
    filter(reads >= 10^4) %>% 
    ggplot(aes(reads, features)) +
    # geom_point(aes(colour = method),
    #     size = 1,
    #     alpha = .5,) +
    geom_path(aes(colour = method), lwd = .25) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe() +
    facet_wrap(~feature_type, scale = "free_y") +
    scale_y_log10(labels = label_scientific()) +
    scale_x_log10(labels = trans_format("log10", math_format(10^.x))) +
    xlab("Counts") +
    ylab("Detected features") +
    scale_colour_brewer(palette = "Dark2", name = NULL) +
    theme(legend.key.size = unit(.33, "cm"))
```

While the illumina dataset has reached saturation, the ont datasets have not
yet. In addition, the teloprime curves seem to plateau out at smaller values.

# Transcript detection rate by transcript length

## Overview

```{r, fig.width = 6.85 / 2, fig.asp = .7}
df <- df_tx_sum %>%
    tidyr::drop_na() %>%
    mutate(grouped_transcript_length = cut(transcript_length,
        breaks = c(0, 500, 1000, 1500, 2000, 2500, 3000,
            3500, 4000, 4500, 5000, 7500, 123179))) %>%
    mutate(detected = if_else(sum_counts > 1, TRUE, FALSE)) %>%
    group_by(dataset, grouped_transcript_length) %>%
    summarise(n = n(),
        detected = sum(detected)) %>%
    mutate(rate = detected / n)

df %>%
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling() %>% 
    kableExtra::scroll_box(width = "100%", height = "600px")

df %>%
    ggplot(aes(grouped_transcript_length, rate)) +
    geom_bar(stat = "identity") +
    facet_wrap(~dataset) +
    xlab("Transcript length") +
    ylab("Detected transcripts") +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(
        data = tibble(rate = c(0, .8), grouped_transcript_length = c(0,1)),
        sides = 'l') +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    scale_y_continuous(labels = label_percent())
```

## By transcript biotype

```{r, fig.width = 6.85 / 2, fig.asp = .7}
df_tx_sum %>%
    tidyr::drop_na() %>%
    mutate(grouped_transcript_length = cut(transcript_length,
        breaks = c(0, 1000, 2000, 3000, 4000, 5000, 7500, 123179),
        labels = c("1000", "2000", "3000", "4000",
            "5000", "7500", "123179"))) %>%
    mutate(detected = if_else(sum_counts > 1, TRUE, FALSE)) %>%
    group_by(dataset, grouped_transcript_length, grouped_biotype) %>%
    summarise(n = n(),
        detected = sum(detected)) %>%
    mutate(rate = detected / n) %>%
    filter(grouped_biotype != "small ncRNA") %>%
    mutate(dataset = factor(dataset, levels = c("illumina", "teloprime", "cdna", "rna"))) %>% 
    ggplot(aes(grouped_transcript_length, rate)) +
    geom_bar(aes(fill = dataset),
        stat = "identity",
        position = position_dodge()) +
    facet_wrap(~grouped_biotype) +
    xlab("Transcript length") +
    scale_y_continuous(labels = label_percent()) +
    ylab("Detected transcripts") +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(
        data = tibble(rate = c(0, 1), grouped_transcript_length = c(0,1)),
        sides = 'l') +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5)) +
    scale_fill_brewer(palette = "Dark2", name = NULL) +
    theme(legend.key.size = unit(.33, "cm")) +
    theme(legend.position = "top",
        legend.margin = margin(t = 0, r = 0, b = 0, l = 0, unit = "cm"))
```

# Features not detected by one but by the other method

## Genes

```{r, fig.asp = .7}
df_gene_sum %>%
    mutate(detected = if_else(sum_counts > 1, TRUE, FALSE)) %>%
    select(dataset, ensembl_gene_id_version, detected) %>%
    mutate(detected = as.integer(detected)) %>% 
    tidyr::spread(dataset, detected) %>%
    tidyr::drop_na() %>%
    as.data.frame() %>%
    tibble::column_to_rownames("ensembl_gene_id_version") %>%
    upset(nsets = 4, keep.order = TRUE,
        order.by = "freq",
        sets = c("illumina", "teloprime", "cdna", "rna"),
        sets.x.label = "Detected genes",
        mainbar.y.label = "Shared detected genes",
        text.scale = 1.75)
```

## Transcripts

```{r, fig.asp = .7}
df_tx_sum %>%
    mutate(detected = if_else(sum_counts > 1, TRUE, FALSE)) %>%
    select(dataset, ensembl_transcript_id_version, detected) %>%
    mutate(detected = as.integer(detected)) %>%
    tidyr::spread(dataset, detected) %>%
    tidyr::drop_na() %>%
    as.data.frame() %>%
    tibble::column_to_rownames("ensembl_transcript_id_version") %>%
    upset(nsets = 4, keep.order = TRUE,
        order.by = "freq",
        sets = c("illumina", "teloprime", "cdna", "rna"),
        sets.x.label = "Detected transcripts",
        mainbar.y.label = "Shared detected transcripts",
        text.scale = 1.75)
```

## Genes detected by illumina but not by teloprime

```{r, fig.width = 6.85 / 2, fig.asp = .7}
df_gene_sum %>%
    mutate(detected = if_else(sum_counts > 1, TRUE, FALSE)) %>%
    select(dataset, ensembl_gene_id_version, detected) %>%
    tidyr::spread(dataset, detected) %>%
    filter(illumina) %>%
    left_join(gene_tpm, by = "ensembl_gene_id_version") %>%
    left_join(biomaRt_gene, by = "ensembl_gene_id_version") %>%
    left_join(biotype_groups, by = c("gene_biotype" = "transcript_biotype")) %>%
    filter(!grouped_biotype == c("small ncRNA")) %>%
    tidyr::drop_na() %>%
    mutate(teloprime = factor(teloprime, levels = c(TRUE, FALSE))) %>%
    ggplot(aes(grouped_biotype, average_tpm)) +
    geom_violin(aes(fill = teloprime), lwd = .25) +
    scale_y_continuous(trans = "log1p",
        breaks = c(0.3, 1, 3, 10)) +
    xlab(NULL) +
    ylab("Average TPM") +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    scale_fill_brewer(palette = "Dark2", name = "in teloprime") +
    theme(legend.key.size = unit(.33, "cm")) 
```

In general, the transcripts not detected by teloprime show a lower expression in the
illumina libraries compared to the transcripts detected by both mehtods.

