---
title: "Nanopore iBAT, Read lengths etc"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
    html_document:
        toc: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, cache = FALSE)
```

```{r loadPackages}
save.image("readLengths.RData")
source("packrat/init.R")
library("dplyr")
library("readr")
library("purrr")
library("ggplot2")
library("ggthemes")
library("AnnotationDbi")
```

```{r}
dir.create(snakemake@params[["fig_folder"]],
    recursive = TRUE, showWarnings = FALSE)
```

```{r}
sample_info <- read_csv(snakemake@input[["sample_info"]]) %>% 
    filter(!is.na(ont)) %>%
    mutate_at(vars(matches("condition")), as.factor) %>% 
    dplyr::select(sample_id, illumina, ont, cdna) %>% 
    tidyr::gather("method", "barcode", -sample_id) %>% 
    dplyr::select(-method) 
```

# Reads and mapping

```{r import_flagstats}
flagstats <- snakemake@input[["flagstats"]] %>% 
    map(read_csv) %>% 
    bind_rows() %>%  
    tidyr::separate(file, c("waste", "library", "flowcell", "filename"),
        sep = '/') %>% 
    tidyr::separate(filename, c("waste1", "waste2", "type", "barcode"),
        extra = "drop") %>% 
    left_join(sample_info, by = "barcode") %>% 
    dplyr::select(-barcode) %>% 
    dplyr::select_at(vars(-matches("waste"))) %>% 
    mutate(flowcell = if_else(flowcell %in% c("X1_flowcell", "pool1"),
        "cell_1", "cell_2")) %>% 
    dplyr::select(sample_id, everything())

flagstats %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling()
```

```{r plot_flagstats}
add_total <- function(df) {
    df %>% add_row(library = unique(.$library), 
        type = unique(.$type),
        mapping_type = "total",
        count = .$count[.$mapping_type == "primary"] +
            .$count[.$mapping_type == "unmapped"]) 
}

grouped_flagstats <- flagstats %>% 
    group_by(library, type) %>%
    summarise_if(is.numeric, sum) %>%
    ungroup() %>% 
    tidyr::gather("mapping_type", "count", -library, -type) %>% 
    arrange(type, library) %>% 
    split(list(.$library, .$type)) %>% 
    map(add_total) %>% 
    bind_rows()

grouped_flagstats %>% 
    tidyr::spread(mapping_type, count) %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling()

grouped_flagstats %>% 
    filter(mapping_type != "secondary") %>%
    mutate(mapping_type = factor(mapping_type,
        levels = c("total", "primary", "unmapped", "supplementary"))) %>% 
    ggplot(aes(x = library, y = count,  fill = mapping_type)) +
    geom_bar(stat = "identity", position = position_dodge()) +
    facet_wrap(~type) +
    scale_y_continuous(labels = function(x) x / 10^6) +
    theme_tufte() +
    geom_rangeframe(sides = "l") +
    xlab(NULL) + 
    ylab("Mio. Reads") +
    scale_fill_colorblind()
```

# Length of primary and supplementary reads

Based on the mapping against the transcriptome. From bam files.

```{r }
get_map_type <- function(flag) {
    if_else(flag == 4, "unmapped",
        if_else(flag %in% c(0, 16), "primary",
            if_else(flag %in% c(256, 275), "secondary", "supplementary")))
}
```

```{r dataImport_tx_bam}
df_tx <- c(snakemake@input[["teloprime_bam_tx"]], snakemake@input[["cdna_bam_tx"]]) %>%
    set_names(  
        tools::file_path_sans_ext(basename(snakemake@input[["teloprime_bam_tx"]])),
        tools::file_path_sans_ext(basename(snakemake@input[["cdna_bam_tx"]])))

names(df_tx) <- strsplit(names(df_tx), '_') %>% 
    map(`[`, 1:2) %>% 
    map(paste, collapse = '_')

df_tx <- df_tx[c(1,7)] %>% 
    map(read_rds) %>%  
    map(mutate, type = get_map_type(flag)) %>% 
    map(dplyr::select, -flag) %>% 
    map(tidyr::separate, col = seqnames,
        into = "ensembl_transcript_id_version",
        sep = '\\|', extra = "drop") %>% 
    map(as_tibble) %>% 
    map(group_by, qname) %>% 
    map(mutate, has_supplementary = any(type == "supplementary")) %>% 
    map(ungroup) %>% 
    bind_rows(.id = "sample") %>% 
    tidyr::separate(sample, c("library", "barcode"), sep = '_')
```

```{r}
df_tx %>% 
    dplyr::select(-barcode) %>% 
    left_join(mapping_df_tx %>% dplyr::select(-library), by = "qname") %>% 
    mutate(category = if_else(type == "supplementary", "supplementary",
        if_else(has_supplementary, "primary_with_supplementary", "primary_wo_supplementary"))) %>% 
    ggplot(aes(category, qwidth)) + 
    geom_violin(aes(fill = library)) +
    scale_y_log10() +
    scale_fill_colorblind() +
    theme_tufte() +
    geom_rangeframe(sides = "l") +
    xlab(NULL) + 
    ylab("Read length")
```

```{r}
df_tx %>% 
    dplyr::select(-sample, -flag) %>% 
    left_join(mapping_df_tx %>% dplyr::select(-library), by = "qname") %>% 
    mutate(category = if_else(type == "supplementary", "supplementary",
        if_else(has_supplementary, "primary_with_supplementary", "primary_wo_supplementary"))) %>% 
    ggplot(aes(category, aligned)) + 
    geom_violin(aes(fill = library)) +
    scale_y_log10() +
    scale_fill_colorblind() +
    theme_tufte() +
    geom_rangeframe(sides = "l") +
    xlab(NULL) + 
    ylab("Aligned read length")
```


# Transcriptome

## Aligned length vs read length

Primary hits only

```{r plotTranscriptome}
df_tx %>% 
    filter(type == "primary") %>% 
    ggplot(aes(qwidth, aligned)) +
    geom_hex(aes(fill = ..density..), binwidth = .025) +
    facet_grid(~library) +
    scale_x_log10() +
    scale_y_log10() +
    scale_fill_viridis_c(trans = "log") +
    geom_abline(slope = 1) +
    theme_tufte() +
    geom_rangeframe(sides = "bl") +
    xlab("Read Length") + 
    ylab("Aligned Length")
```

## Aligned length vs expected length

```{r import_txdb}
biomart <- snakemake@input[["biomaRt_tx"]] %>% 
    read_rds()
```

```{r}
df_tx %>% 
    filter(type == "primary") %>% 
    dplyr::select(ensembl_transcript_id_version, aligned) %>% 
    left_join(biomart, by = "ensembl_transcript_id_version") %>% 
    filter(!is.na(transcript_length)) %>% 
    ggplot(aes(transcript_length, aligned)) +
    geom_hex(aes(fill = ..density..), binwidth = .025) +
    facet_grid(~library) +
    scale_x_log10() +
    scale_y_log10() +
    scale_fill_viridis_c(trans = "log") +
    geom_abline(slope = 1) +
    theme_tufte() +
    geom_rangeframe(sides = "bl") +
    xlab("Transcript length") + 
    ylab("Aligned length")
```

## Read length vs expected length

```{r}
df_tx %>% 
    filter(type == "primary") %>% 
    dplyr::select(ensembl_transcript_id_version, qwidth) %>% 
    left_join(biomart, by = "ensembl_transcript_id_version") %>% 
    filter(!is.na(transcript_length)) %>% 
    ggplot(aes(transcript_length, qwidth)) +
    geom_hex(aes(fill = ..density..), binwidth = .025) +
    facet_grid(~library) +
    scale_x_log10() +
    scale_y_log10() +
    scale_fill_viridis_c(trans = "log") +
    geom_abline(slope = 1) +
    theme_tufte() +
    geom_rangeframe(sides = "bl") +
    xlab("Transcript length") + 
    ylab("Read length")
```
