---
title: "Comparison of reannotation by flair and stringtie"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
    html_document:
        toc: true
        toc_float: true
        toc_depth: 2
        fig_caption: false
        dev: "svg"
editor_options:
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE,
    cache = FALSE, out.width = 800)
```

```{r library, include=FALSE}
source(".Rprofile")
suppressPackageStartupMessages({
    library("dplyr")
    library("readr")
    library("purrr")
    library("ggplot2")
    library("ggthemes")
    library("UpSetR")
    library("scales")
})
```

```{r dataImport}
# load gffcompare
df <- snakemake@input[["gffcompare"]] %>% 
    set_names(., sub("^.+/gffcmp.", "", .) %>%
                     sub("\\.isoforms.gtf.tmap$", "", .) %>% 
                     sub("\\.gtf.+$", "", .))  %>% 
    map(read_tsv, show_col_types = FALSE) %>%  
    bind_rows(.id = "dataset") %>% 
    mutate(dataset = factor(dataset,
        levels = c("illumina_stringtie", "teloprime_stringtie", "cdna_stringtie",
            "rna_stringtie", "teloprime_flair", "cdna_flair", "rna_flair")))


# load sqanti
df2 <- snakemake@input[["sqanti"]] %>% 
    set_names(., sub("^.+stringtie/", "", .) %>% 
                     sub("^.+flair/", "", .) %>% 
                     sub("_noU.+$", "", .) %>% 
                     sub("\\.iso.+$", "", .) %>% 
                     sub("^.+/", "", .)) %>% 
    map(read_tsv, show_col_types = FALSE) %>% 
    bind_rows(.id = "dataset") %>% 
    mutate(
        dataset = factor(dataset,
            levels = c("illumina_stringtie", "teloprime_stringtie", "cdna_stringtie",
                "rna_stringtie", "teloprime_flair", "cdna_flair", "rna_flair")),
        structural_category = factor(structural_category,
            levels = c("full-splice_match", "novel_in_catalog", "incomplete-splice_match",
                "novel_not_in_catalog", "fusion", "antisense",
                "genic", "genic_intron", "intergenic")))
```


# Gffcompare

```{r}
knitr::include_graphics("http://ccb.jhu.edu/software/stringtie/gffcompare_codes.png")
```

http://ccb.jhu.edu/software/stringtie/gffcompare_codes.png

```{r plot_gffcompare, fig.width = 6.85, fig.asp = .65 / 2}
df %>%
    mutate(grouped_num_exons = cut(num_exons, 
        breaks = c(0, 1, 2, 5, max(num_exons)),
        labels = c("1", "2", "3-5", ">5"))) %>% 
    ggplot(aes(class_code)) +
    geom_bar(aes(fill = grouped_num_exons), position = "fill") +
    facet_wrap(~dataset, nrow = 1) +
    scale_fill_brewer(palette = "Dark2", name = "exons") +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    ylab(NULL) +
    scale_y_continuous(labels = scales::percent) +
    theme(legend.key.size = unit(.33, "cm"))


df %>% 
    ggplot(aes(class_code, len)) +
    geom_boxplot() +
    facet_wrap(~dataset) +
    scale_y_log10() +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    ylab("Transcript length [nt]")
```


```{r plot_gffcompare2, fig.width = 6.85 / 2, fig.asp = .65}
df %>% 
    ggplot(aes(dataset)) +
    geom_bar(aes(fill = class_code), position = "fill") +
    scale_fill_viridis_d() +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    ylab(NULL) +
    xlab(NULL) +
    scale_y_continuous(labels = scales::percent) +
    theme(legend.key.size = unit(.33, "cm")) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))

df %>% 
    ggplot(aes(dataset)) +
    geom_bar(aes(fill = class_code)) +
    scale_fill_viridis_d() +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    ylab(NULL) +
    xlab(NULL) +
    theme(legend.key.size = unit(.33, "cm")) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))
```

# Sqanti

```{r}
knitr::include_graphics("https://raw.githubusercontent.com/Magdoll/images_public/master/SQANTI2_figures/sqanti2_classification.png")
```

https://github.com/Magdoll/SQANTI2

```{r plot_sqanti, fig.width = 6.85, fig.asp = .65}
df2 %>% 
    mutate(grouped_num_exons = cut(exons, 
        breaks = c(0, 1, 2, 5, max(exons)),
        labels = c("1", "2", "3-5", ">5"))) %>% 
    ggplot(aes(structural_category)) +
    geom_bar(aes(fill = grouped_num_exons), position = "fill") +
    facet_wrap(~dataset, nrow = 2) +
    scale_fill_brewer(palette = "Dark2", name = "exons") +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    ylab(NULL) +
    xlab(NULL) +
    scale_y_continuous(labels = scales::percent) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5)) +
    theme(legend.key.size = unit(.33, "cm"))


df2 %>% 
    ggplot(aes(structural_category, length)) +
    geom_boxplot() +
    facet_wrap(~dataset, nrow = 2) +
    scale_y_log10() +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    ylab("Transcript length [nt]") +
    xlab(NULL) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))
```


```{r plot_sqanti2, fig.width = 6.85 / 2, fig.asp = .65}
df2 %>% 
    mutate(structural_category = factor(structural_category, rev(levels(structural_category)))) %>% 
    ggplot(aes(dataset)) +
    geom_bar(aes(fill = structural_category), position = "fill") +
    scale_fill_brewer(palette = "Paired", name = NULL, limits = levels(df2$structural_category)) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    ylab("Reassembled transcripts") +
    xlab(NULL) +
    scale_y_continuous(labels = scales::percent) +
    theme(legend.key.size = unit(.33, "cm")) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))

df2 %>% 
    mutate(structural_category = factor(structural_category, rev(levels(structural_category)))) %>% 
    ggplot(aes(dataset)) +
    geom_bar(aes(fill = structural_category)) +
    scale_fill_brewer(palette = "Paired", name = NULL, limits = levels(df2$structural_category)) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    ylab("Reassembled transcripts") +
    xlab(NULL) +
    theme(legend.key.size = unit(.33, "cm")) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))
```

# Overlap

Based on sqanti. Comparison between the different methods ability to identify 
transcripts from the reference annotation.

```{r upset, fig.asp = .65}
df2 %>% 
    filter(grepl("^ENS", associated_transcript)) %>% 
    group_by(dataset, associated_transcript) %>% 
    summarise(n = n()) %>% 
    ungroup() %>% 
    tidyr::pivot_wider(names_from = dataset, values_from = n, values_fill = 0) %>% 
    mutate_if(is.numeric, ~ as.integer(. > 0)) %>% 
    as.data.frame() %>%
    tibble::column_to_rownames("associated_transcript") %>% 
    upset(nsets = 7, keep.order = TRUE,
        nintersects = 15,
        order.by = "freq",
        sets = c("illumina_stringtie", "teloprime_stringtie", "cdna_stringtie",
            "rna_stringtie", "teloprime_flair", "cdna_flair", "rna_flair"),
        sets.x.label = "Annotated transcripts",
        mainbar.y.label = "Shared annotated transcripts",
        text.scale = 1.5)
```

# Sensitivity and specificity

```{r, fig.width = 6.85 / 2, fig.asp = .65}
tibble(
    dataset = c("illumina_stringtie", "teloprime_stringtie", "cdna_stringtie",
        "rna_stringtie", "teloprime_flair", "cdna_flair", "rna_flair"),
    sensitivity = c(19.7, 23.1, 19.3, 18.1, 19.0, 15.1, 15.8) / 100,
    precision = c(72.0, 76.5, 70.0, 74.7, 43.9, 45.8, 65.2) / 100) %>% 
    ggplot(aes(precision, sensitivity)) +
    geom_point(aes(colour = grepl("flair", dataset)), size = 1, alpha = .5) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    scale_color_brewer(palette = "Dark2", guide = NULL) + 
    scale_x_continuous(name = "Precision", labels = label_percent(accuracy = 1)) +
    scale_y_continuous(name = "Sensitivity", labels = label_percent()) +
    ggrepel::geom_text_repel(aes(label = dataset), size = 2)
```
