---
title: "Read lengths bam"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
    html_document:
        toc: true
        toc_float: true
        toc_depth: 2
        fig_caption: false
        dev: "svg"
editor_options:
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE,
    cache = FALSE, out.width = 800)
```

```{r loadPackages}
source(".Rprofile")
library("dplyr")
library("readr")
library("purrr")
library("ggplot2")
library("ggthemes")
library("scales")
```


```{r import_general}
sample_info <- read_csv(snakemake@input[["sample_info"]]) %>% 
    filter(!is.na(ont)) %>%
    mutate_at(vars(matches("condition")), as.factor)

sample_info2 <- sample_info %>% 
    dplyr::select(sample_id, illumina, ont, cdna) %>% 
    tidyr::gather("method", "barcode", -sample_id) %>% 
    dplyr::select(-method) 

biomart <- snakemake@input[["biomaRt_tx"]] %>%
    read_rds()
```


# Read categories based on flagstat

```{r import_flagstats}
separate_variables <- function(df, method) {
    if (method %in% c("cdna", "teloprime")) {
        df <- df %>%
            tidyr::separate(file, c("waste", "library", "flowcell", "filename"),
                sep = '/') %>% 
            tidyr::separate(filename, c("waste1", "waste2", "type", "barcode"),
                extra = "drop") %>%  
            select_at(vars(-matches("waste")))
    } else if (method == "rna") {
        df <- df %>%
            tidyr::separate(file, c("waste1", "library", "filename"), sep = '/') %>% 
            tidyr::separate(filename, c("type", "barcode"), extra = "drop") %>% 
            select_at(vars(-matches("waste"))) 
    }
    return(df)
}

flagstats <- snakemake@input[["flagstats"]] %>% 
    map(read_csv) %>% 
    map2(c("cdna", "teloprime", "rna"), separate_variables) %>% 
    bind_rows() %>% 
    left_join(sample_info2, by = "barcode") %>% 
    mutate(sample_id = if_else(library == "rna", barcode, sample_id)) %>% 
    dplyr::select(-barcode) %>% 
    dplyr::select_at(vars(-matches("waste"))) %>% 
    mutate(flowcell = if_else(flowcell %in% c("X1_flowcell", "pool1"),
        "cell_1", "cell_2")) %>% 
    mutate(flowcell = if_else(library == "rna", NA_character_, flowcell)) %>% 
    dplyr::select(sample_id, everything())
```

```{r plot_flagstats, fig.width = 6.85 / 2, fig.asp = .65}
add_total <- function(df) {
    df %>% add_row(library = unique(.$library), 
        type = unique(.$type),
        mapping_type = "total",
        count = .$count[.$mapping_type == "primary"] +
            .$count[.$mapping_type == "unmapped"]) 
}

grouped_flagstats <- flagstats %>% 
    group_by(library, type) %>%
    summarise_if(is.numeric, sum) %>%
    ungroup() %>% 
    tidyr::gather("mapping_type", "count", -library, -type) %>% 
    arrange(type, library) %>% 
    split(list(.$library, .$type)) %>% 
    map(add_total) %>% 
    bind_rows()

grouped_flagstats %>% 
    tidyr::spread(mapping_type, count) %>% 
    mutate(primary_pc = primary / total,
           supplementary_pc = supplementary / total,
           unmapped_pc = unmapped / total) %>% 
    select(library, type, total, primary, primary_pc, supplementary,
        supplementary_pc, unmapped, unmapped_pc) %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling()

grouped_flagstats %>% 
    filter(mapping_type != "secondary") %>%
    mutate(mapping_type = factor(mapping_type,
        levels = c("total", "primary", "unmapped", "supplementary"))) %>% 
    mutate(library = factor(library, levels = c("illumina", "teloprime", "cdna", "rna"))) %>% 
    ggplot(aes(x = library, y = count,  fill = mapping_type)) +
    geom_bar(stat = "identity", position = position_dodge()) +
    facet_wrap(~type) +
    scale_y_continuous(labels = label_scientific()) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "lb") +
    xlab(NULL) + 
    ylab("Reads") +
    scale_fill_brewer(palette = "Dark2", name = NULL) +
    theme(legend.key.size = unit(.33, "cm"))

grouped_flagstats %>% 
    filter(mapping_type != "secondary") %>%
    mutate(mapping_type = factor(mapping_type,
        levels = c("total", "primary", "unmapped", "supplementary"))) %>%
    group_by(type, library) %>% 
    mutate(count = count / count[mapping_type == "total"]) %>% 
    ungroup() %>% 
    mutate(library = factor(library, levels = c("illumina", "teloprime", "cdna", "rna"))) %>% 
    ggplot(aes(x = library, y = count,  fill = mapping_type)) +
    geom_bar(stat = "identity", position = position_dodge()) +
    facet_wrap(~type) +
    scale_y_continuous(labels = label_percent()) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "lb") +
    xlab(NULL) + 
    ylab("Fraction of reads") +
    scale_fill_brewer(palette = "Dark2", name = NULL) +
    theme(legend.key.size = unit(.33, "cm"))
```

# Length of primary and supplementary reads (against transcriptome)

Based on the mapping against the transcriptome. From bam files.

```{r dataImport_tx_bam}
df_tx <- read_rds(snakemake@input$collapsed_transcripts) %>% 
    mutate(category = if_else(type == "supplementary", "supplementary",
        if_else(has_supplementary, "primary_with_supplementary", "primary_wo_supplementary")))

tpm <- snakemake@input[["avg_counts"]] %>% 
    read_csv() %>% 
    select(transcript_length, illumina) %>% 
    rename(qwidth = "transcript_length") %>%
    tidyr::drop_na() %>% 
    mutate(illumina = 2^illumina - 1) %>% 
    group_by(qwidth) %>% 
    summarise(n = sum(illumina, na.rm = TRUE)) %>% 
    ungroup() %>% 
    mutate(library = "illumina")
```

## Read categories

```{r, fig.width = 6.85 / 2, fig.asp = .65}
df_tx %>% 
    group_by(library, barcode, category) %>% 
    summarise(n = n()) %>% 
    left_join(sample_info2) %>%
    ungroup() %>% 
    mutate(sample_id = if_else(library == "rna", barcode, sample_id)) %>% 
    ggplot(aes(x = sample_id, y = n,  fill = category)) +
    geom_bar(stat = "identity") +
    facet_wrap(~library, scales = "free_x") +
    scale_y_continuous(labels = label_scientific()) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "l") +
    scale_x_discrete(labels = NULL, breaks = NULL, name = NULL) +
    ylab("Reads") +
    scale_fill_brewer(palette = "Dark2", name = NULL) +
    theme(legend.key.size = unit(.33, "cm"))

df_tx %>% 
    group_by(library, barcode, category) %>% 
    summarise(n = n()) %>% 
    left_join(sample_info2) %>%  
    mutate(n = n / sum(n)) %>% 
    ungroup() %>% 
    mutate(sample_id = if_else(library == "rna", barcode, sample_id)) %>% 
    ggplot(aes(x = sample_id, y = n,  fill = category)) +
    geom_bar(stat = "identity") +
    facet_wrap(~library, scales = "free_x") +
    scale_y_continuous(name = "Fraction of reads", labels = label_percent()) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "l") +
    scale_x_discrete(labels = NULL, breaks = NULL, name = NULL) +
    scale_fill_brewer(palette = "Dark2", name = NULL) +
    theme(legend.key.size = unit(.33, "cm"))
```

## Grouped read lengths

The illumina values are estimated based on the tpm values from salmon. They are
somewhat a reference, but not actual read lengths.

```{r, fig.width = 6.85 / 2, fig.asp = .65}
df_tx %>% 
    filter(type == "primary") %>% 
    group_by(library, qwidth) %>% 
    summarise(n = n()) %>% 
    bind_rows(tpm) %>% 
    mutate(weight = n / sum(n)) %>% 
    ungroup() %>% 
    mutate(library = factor(library, levels = c("illumina", "teloprime", "cdna", "rna"))) %>% 
    ggplot(aes(library, qwidth, weight = weight)) +
    geom_violin(lwd = .25) +
    scale_y_log10(labels = trans_format("log10", math_format(10^.x))) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "bl") +
    xlab(NULL) + 
    ylab("Read/transcript length") +
    theme(legend.key.size = unit(.33, "cm"))
```

# Mapping of supplementary reads

```{r}
df_tx %>% 
    filter(category != "primary_wo_supplementary") %>% 
    left_join(biomart %>% 
            select(ensembl_transcript_id_version, ensembl_gene_id_version),
        by = "ensembl_transcript_id_version") %>% 
    group_by(library, qname) %>% 
    summarise(supplementary_type = if_else(all(ensembl_transcript_id_version == ensembl_transcript_id_version), "same_transcript", 
            if_else(all(ensembl_gene_id_version == ensembl_gene_id_version), "same_gene", "not_same_gene")),
        n = n()) %>% 
    mutate(supplementary_type = as.factor(supplementary_type)) %>% 
    filter(n > 1) %>% 
    summary()
```


```{r, fig.width = 6.85 / 2, fig.asp = .65}
df_tx %>% 
    ggplot(aes(category, qwidth)) + 
    geom_violin(aes(fill = library), lwd = .25) +
    scale_y_log10(labels = trans_format("log10", math_format(10^.x))) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "bl") +
    xlab(NULL) + 
    ylab("Read length") +
    scale_fill_brewer(palette = "Dark2", name = NULL) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5)) +
    theme(legend.key.size = unit(.33, "cm"))
```

```{r, fig.width = 6.85 / 2, fig.asp = .65}
df_tx %>% 
    ggplot(aes(category, aligned)) + 
    geom_violin(aes(fill = library), lwd = .25) +
    scale_y_log10(labels = trans_format("log10", math_format(10^.x))) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "bl") +
    xlab(NULL) + 
    ylab("Aligned read length") +
    scale_fill_brewer(palette = "Dark2", name = NULL) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5)) +
    theme(legend.key.size = unit(.33, "cm"))
```

## Correlations

```{r}
df_tx <- df_tx %>% 
    filter(type == "primary")
```


### Aligned length vs read length

Primary hits only

```{r plotTranscriptome, fig.width = 6.85 / 2, fig.asp = .65}
df_tx %>% 
    dplyr::select(library, qwidth, aligned) %>% 
    ggplot(aes(qwidth, aligned)) +
    geom_hex(aes(fill = ..density..), binwidth = .025) +
    facet_grid(~library) +
    scale_x_log10(labels = trans_format("log10", math_format(10^.x))) +
    scale_y_log10(labels = trans_format("log10", math_format(10^.x))) +
    scale_fill_viridis_c(trans = "log10", labels = trans_format("log10", math_format(10^.x))) +
    geom_abline(slope = 1, lwd = .25) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "bl") +
    xlab("Read Length") + 
    ylab("Aligned Length") +
    theme(legend.key.size = unit(.33, "cm"))
```

## Aligned length vs expected length

```{r, fig.width = 6.85 / 2, fig.asp = .65}
df_tx %>%
    dplyr::select(library, ensembl_transcript_id_version, aligned) %>%
    left_join(biomart, by = "ensembl_transcript_id_version") %>%
    filter(!is.na(transcript_length)) %>%
    ggplot(aes(transcript_length, aligned)) +
    geom_hex(aes(fill = ..density..), binwidth = .025) +
    facet_grid(~library) +
    scale_x_log10(labels = trans_format("log10", math_format(10^.x))) +
    scale_y_log10(labels = trans_format("log10", math_format(10^.x))) +
    scale_fill_viridis_c(trans = "log10", labels = trans_format("log10", math_format(10^.x))) +
    geom_abline(slope = 1, lwd = .25) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "bl") +
    xlab("Transcript length") +
    ylab("Aligned length") +
    theme(legend.key.size = unit(.33, "cm"))
```

## Read length vs expected length

```{r, fig.width = 6.85 / 2, fig.asp = .65}
df_tx %>%
    dplyr::select(library, ensembl_transcript_id_version, qwidth) %>%
    left_join(biomart, by = "ensembl_transcript_id_version") %>%
    filter(!is.na(transcript_length)) %>%
    ggplot(aes(transcript_length, qwidth)) +
    geom_hex(aes(fill = ..density..), binwidth = .025) +
    facet_grid(~library) +
    scale_x_log10(labels = trans_format("log10", math_format(10^.x))) +
    scale_y_log10(labels = trans_format("log10", math_format(10^.x))) +
    scale_fill_viridis_c(trans = "log10", labels = trans_format("log10", math_format(10^.x))) +
    geom_abline(slope = 1, lwd = .25) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "bl") +
    xlab("Transcript length") +
    ylab("Read length") +
    theme(legend.key.size = unit(.33, "cm"))
```
