---
title: "Nanopore iBAT, Coverage"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
    html_document:
        toc: true
        toc_float: true
        toc_depth: 2
        fig_caption: false
        dev: "svg"
editor_options:
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE,
    cache = FALSE, out.width = 800)
```

```{r loadPackages}
source(".Rprofile")
suppressPackageStartupMessages({
    library("dplyr")
    library("readr")
    library("purrr")
    library("ggplot2")
    library("ggthemes")
    library("ggpmisc")
    library("scales")
})
```


```{r}
sample_info <- read_csv(snakemake@input[["sample_info"]], show_col_types = FALSE) %>%
    filter(!is.na(ont)) %>%
    mutate_at(vars(matches("condition")), as.factor) %>%
    dplyr::select(sample_id, illumina, ont, cdna)
```

# Gene body coverage

All plots are primary mappings only

```{r dataImport_geneBodyCoverage}
read_geneBodyCoverage <- function(file) {
    file %>%
        read_tsv(show_col_types = FALSE) %>%
        tidyr::gather() %>%
        magrittr::set_colnames(.[1, ]) %>%
        `[`(-1, ) %>%
        mutate_all(as.integer)
}

prepare_dataset = function(files, name) {
    files %>%
        map(read_geneBodyCoverage) %>%
        purrr::reduce(left_join, by = "Percentile") %>%
        tidyr::gather("sample", "n", -Percentile) %>%
        mutate(dataset = name)
}

illumina <- snakemake@input[["geneBodyCoverage_illumina"]] %>%
    prepare_dataset(name = "illumina") %>%
    mutate(sample = substr(sample, 2, 9)) %>%
    left_join(sample_info, by = c("sample" = "illumina")) %>%
    dplyr::select(-sample, -ont)

teloprime <- snakemake@input[["geneBodyCoverage_teloprime"]] %>%
    prepare_dataset(name = "teloprime") %>% 
    tidyr::separate(sample, c("library", "sample"), extra = "drop") %>%
    left_join(sample_info, by = c("sample" = "ont")) %>%
    dplyr::select(-sample, -illumina)

cdna <- snakemake@input[["geneBodyCoverage_cdna"]] %>%
    prepare_dataset(name = "cdna") %>%
    tidyr::separate(sample, c("library", "sample"), extra = "drop") %>%
    left_join(sample_info, by = c("sample" = "cdna")) %>%
    dplyr::select(-sample, -illumina)

rna <- snakemake@input[["geneBodyCoverage_rna"]] %>%
    prepare_dataset(name = "rna") %>% 
    tidyr::separate(sample, c("waste", "sample_id"), extra = "drop") %>% 
    dplyr::select(-waste)
```


```{r plot_geneBodyCoverage, fig.width = 6.85 / 2, fig.asp = .65}
list(illumina, teloprime, cdna, rna) %>%
    bind_rows() %>%
    group_by(dataset, sample_id) %>%
    mutate(p = n / max(n)) %>%
    ungroup() %>%
    mutate(dataset = factor(dataset, levels = c("illumina", "teloprime", "cdna", "rna"))) %>% 
    ggplot(aes(Percentile, p)) +
        geom_line(aes(colour = dataset, lty = sample_id), lwd = .25) +
        theme_tufte(base_size = 8, base_family = "Helvetica") +
        geom_rangeframe(sides = "bl") +
        scale_y_continuous(labels = scales::percent) +
        xlab("Gene body percentile (5' -> 3')") +
        ylab("Coverage") +
        scale_colour_brewer(palette = "Dark2", name = NULL) +
        guides(linetype = "none") +
        theme(legend.key.size = unit(.33, "cm"))
```

# Read length distribution compared to salmon tpm

From here on it is only primary reads

```{r import_coverage}
df <- snakemake@input[["collapsed_coverage"]] %>% 
    read_rds()

biomart <- read_rds(snakemake@input$biomaRt_tx) %>% 
    dplyr::select(ensembl_transcript_id_version, transcript_length, transcript_biotype)

tpm <- snakemake@input$tpm %>% 
    read_csv(show_col_types = FALSE) %>% 
    mutate(tpm = expm1((rowMeans(log1p(dplyr::select(., starts_with("190220"))))))) %>% 
    dplyr::select(id, tpm) %>% 
    filter(tpm != 0) %>% 
    arrange(desc(tpm)) %>% 
    tidyr::drop_na() %>% 
    left_join(biomart, by = c("id" = "ensembl_transcript_id_version")) %>% 
    group_by(transcript_length) %>% 
    summarise(n = sum(tpm, na.rm = TRUE)) %>% 
    ungroup() %>% 
    dplyr::rename(qwidth = "transcript_length") %>% 
    mutate(library = "illumina")
```

The illumina values are estimated based on the tpm values from salmon. They are
somewhat a reference, but not actual read lengths.

Primary mappings only!

```{r, fig.width = 6.85 / 2, fig.asp = .65}
df %>% 
    filter(type == "primary") %>% 
    filter(!is.na(transcript_length)) %>% 
    mutate(qwidth = as.integer(transcript_length * coverage)) %>% 
    group_by(library, qwidth) %>% 
    summarise(n = n()) %>% 
    bind_rows(tpm) %>% 
    mutate(weight = n / sum(n)) %>% 
    ungroup() %>% 
    mutate(library = factor(library, levels = c("illumina", "teloprime", "cdna", "rna"))) %>% 
    ggplot(aes(library, qwidth, weight = weight)) +
    geom_violin(lwd = .25) +
    scale_y_log10(labels = trans_format("log10", math_format(10^.x))) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "bl") +
    xlab(NULL) + 
    ylab("Read/transcript length") +
    theme(legend.key.size = unit(.33, "cm"))
```


# Coverage

## grouped coverage distribution

```{r, fig.width = 6.85, fig.asp = .65 / 2}
df %>%
    filter(!is.na(transcript_length)) %>% 
    mutate(grouped_transcript_length = cut(transcript_length,
        breaks = c(0, 1000, 2000, 3000, 4000, 5000, 7500, max(transcript_length)),
        labels = as.character(c(1000, 2000, 3000, 4000, 5000, 7500, max(transcript_length))))) %>%
    mutate(library = factor(library, levels = c("illumina", "teloprime", "cdna", "rna"))) %>%
    ggplot(aes(grouped_transcript_length, coverage)) +
    geom_violin(aes(fill = library), lwd = .25) +
    facet_wrap(~type, nrow = 2) +
    xlab(NULL) +
    ylab("Fraction of transcript covered by read") +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'lb') +
    scale_fill_brewer(palette = "Dark2", name = NULL) +
    scale_y_continuous(labels = scales::percent) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5)) +
    theme(legend.key.size = unit(.33, "cm"))
```


```{r, fig.width = 6.85 / 2, fig.asp = .65}
df %>% 
    filter(!is.na(transcript_length)) %>% 
    filter(transcript_length > 350) %>% 
    group_by(library, barcode, type, transcript_length) %>% 
    summarise(coverage = mean(coverage)) %>% 
    mutate(library = factor(library, levels = c("illumina", "teloprime", "cdna", "rna"))) %>%
    ggplot(aes(transcript_length, coverage)) +
    geom_smooth(aes(colour = library, lty = type, group = paste(barcode, type, library)),
        lwd = .25, se = FALSE, method = "loess", span = 0.1) +
    scale_x_log10(name = "Transcript Length",
        labels = trans_format("log10", math_format(10^.x))) +
    scale_y_continuous(name = "Coverage", labels = label_percent()) +
    scale_colour_brewer(palette = "Dark2", name = NULL) +
    scale_linetype(name = NULL) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "bl") +
    theme(legend.key.size = unit(.33, "cm"))
```


# Full length

mehr als 90 % 

Primary reads only

```{r, fig.width = 6.85 / 2, fig.asp = .65}
df %>%
    filter(!is.na(transcript_length)) %>% 
    filter(transcript_length > 350) %>% 
    mutate(is_fullLength = if_else(coverage > .9, TRUE, FALSE)) %>%
    group_by(library, barcode, type, transcript_length) %>% 
    summarise(
        n = n(),
        n_fullLength = sum(is_fullLength)) %>% 
    ungroup() %>% 
    mutate(rate = n_fullLength / n) %>%
    mutate(library = factor(library, levels = c("illumina", "teloprime", "cdna", "rna"))) %>%
    ggplot(aes(transcript_length, rate)) +
    geom_smooth(aes(colour = library, lty = type, group = paste(barcode, type, library)),
        lwd = .25, se = FALSE, method = "loess", span = 0.1) +
    scale_x_log10(name = "Transcript Length",
        labels = trans_format("log10", math_format(10^.x))) +
    scale_y_continuous(name = "Fraction of full length reads", labels = label_percent()) +
    scale_colour_brewer(palette = "Dark2", name = NULL) +
    scale_linetype(name = NULL) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "bl") +
    theme(legend.key.size = unit(.33, "cm"))
```

# Abundance of supplementary mappings

```{r, fig.width = 6.85 / 2, fig.asp = .65}
df %>% 
    filter(!is.na(transcript_length)) %>% 
    group_by(library, transcript_length, rname, type) %>% 
    summarise(counts = n()) %>% 
    summarise(ratio = counts[type == "supplementary"] / counts[type == "primary"]) %>% 
    ungroup() %>% 
    mutate(library = factor(library, levels = c("illumina", "teloprime", "cdna", "rna"))) %>%
    ggplot(aes(transcript_length, ratio)) +
    geom_hex(aes(fill = ..density..), bins = 50) +
    geom_hline(yintercept = 1, lwd = .25, lty = 2) +
    geom_smooth(method = "lm", colour = "red", lwd = .25, se = FALSE) +
    facet_wrap(~library) +
    scale_x_log10(name = "Transcript Length",
        labels = trans_format("log10", math_format(10^.x))) +
    scale_y_log10(name = "Supplementary / primary alignments",
        labels = trans_format("log10", math_format(10^.x))) +
    scale_fill_viridis_c() +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = "bl") +
    theme(legend.key.size = unit(.33, "cm"))
```

