---
title: "DEXSeq heatmap"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
    html_document:
        toc: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, cache = FALSE)
```

```{r loadPackages}
source(".Rprofile")
library("dplyr")
library("DEXSeq")
library("readr")
library("ggplot2")
library("ggiraph")
library("ggthemes")
    select <- dplyr::select
```

```{r import Data}
dxd <- read_rds(snakemake@input[["dxd"]])
dxr <- DEXSeqResults(dxd)
```

```{r prepare_df}
# list of significant genes
significant_genes <- dxr %>% 
    as_tibble() %>% 
    filter(padj < snakemake@params[["pvalue_cutoff"]]) %>% 
    pull(groupID) %>% 
    unique()

# log2 expression per exon and condition
df <- featureCounts(dxd, normalized = TRUE) %>% 
    as_tibble(rownames = "rownames") %>% 
    tidyr::separate(rownames,
        c("ensembl_gene_id_version", "exon_number"), sep = ':') %>% 
    filter(ensembl_gene_id_version %in% significant_genes) %>% 
    mutate_if(is.numeric, function(x) log2(x + 1)) %>% 
    rowwise() %>% 
    mutate(warm = mean(c(`5034_S33`, `5039_S38`, `5043_S42`)),
        cold = mean(c(`5035_S34`, `5041_S40`, `5044_S43`))) %>% 
    select(-`5034_S33`, -`5035_S34`, -`5039_S38`,
        -`5041_S40`, -`5043_S42`, -`5044_S43`) %>% 
    ungroup()

# normalise within condition and gene, get differences
df <- df %>% 
    group_by(ensembl_gene_id_version) %>% 
    mutate(warm = warm - mean(warm),
        cold = cold - mean(cold)) %>% 
    ungroup() %>% 
    mutate(difference = cold - warm)

# annotate
biomart <- read_rds(snakemake@input[["biomaRt_gene"]])
df <- df %>% 
    left_join(as_tibble(dxr),
        by = c("ensembl_gene_id_version" = "groupID",
            "exon_number" = "featureID")) %>% 
    left_join(biomart, by = "ensembl_gene_id_version")

# make tooltip
df <- df %>%
    mutate(mgi_symbol =if_else(is.na(mgi_symbol), ensembl_gene_id_version, mgi_symbol)) %>% 
    mutate(padj = tidyr::replace_na(padj, 1)) %>% 
    mutate(signif = if_else(padj < snakemake@params[["pvalue_cutoff"]], TRUE, FALSE)) %>% 
    mutate(tooltip = sprintf("%s \n %s \n %s \n %s \n padj = %s",
        mgi_symbol, ensembl_gene_id_version, description, exon_number, padj)) %>%
    mutate(tooltip = gsub("'", "", tooltip)) %>% 
    mutate(click = sprintf("window.open(\"%s%s\")",
        "http://www.ensembl.org/Mus_musculus/Gene/Summary?db=core;g=",
        ensembl_gene_id_version)) 

# get strand information
ensembl <- biomaRt::useMart("ensembl", dataset = "mmusculus_gene_ensembl")
annotation <- biomaRt::getBM(
        attributes = c("ensembl_gene_id_version", "strand"),
        filters = "ensembl_gene_id_version",
        values = unique(df$ensembl_gene_id_version),
        mart = ensembl)

df_list <- df %>% 
    left_join(annotation, by = "ensembl_gene_id_version") %>%
    split(., .$strand)

df_list$`-1` <- df_list$`-1` %>% 
    group_by(ensembl_gene_id_version) %>% 
    arrange(exon_number, .by_group = TRUE) %>% 
    mutate(exon_number = rev(exon_number)) %>% 
    ungroup()

df <- df_list %>% 
    bind_rows()
```

The fill colour represents the difference in exon usage between the two groups 
(warm and cold). The formula is as follows:

$$exonDifference = (exonExpression_{cold} - mean(exonExpression_{cold})) - (exonExpression_{warm} - mean(exonExpression_{warm}))$$

Click on the rectangles to go to ENSEMBL.

Contrary to the DEXSeq output, E001 is allways the first 5' exon, not the first 
exon according to genomic coordinates.


```{r heatmap}
p_order <- df %>% 
    select(mgi_symbol, padj) %>% 
    group_by(mgi_symbol) %>% 
    summarise(padj = min(padj)) %>% 
    arrange(desc(padj)) %>% 
    pull(mgi_symbol)

pl <- df %>% 
    mutate(mgi_symbol = factor(mgi_symbol, 
        levels = p_order)) %>% 
    ggplot(aes(exon_number, mgi_symbol)) +
    geom_tile_interactive(
        aes(fill = difference,
            colour = signif,
            tooltip = tooltip,
            onclick = click),
        size = .25) +
    scale_fill_viridis_c() +
    scale_colour_colorblind() +
    theme_tufte(base_size = 7) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    xlab(NULL) +
    ylab(NULL)
girafe(ggobj = pl)
```

