---
title: "DEXSeq analysis of illumina reads, iBAT, cold"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
    html_document:
        toc: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, cache = FALSE)
```

```{r loadPackages}
source("packrat/init.R")
library("dplyr")
library("DEXSeq")
library("readr")
BPPARAM = BiocParallel::MulticoreParam(snakemake@threads[[1]])
```

```{r importData}
dxd <- readRDS(snakemake@input[[1]])
```

```{r normalisation}
dxd <- estimateSizeFactors(dxd)
```

# Dispersion estimation

```{r dispersionEstimation}
dxd <- estimateDispersions(dxd, BPPARAM = BPPARAM)
plotDispEsts(dxd)
```

# Differential exon usage

```{r deu}
dxd <- testForDEU(dxd, BPPARAM = BPPARAM)
dxd <- estimateExonFoldChanges(dxd, fitExpToVar = "condition_temp",
    BPPARAM = BPPARAM)
```

```{r results}
dxr1 <- DEXSeqResults(dxd)
```

Genes showing differential exon usage:

```{r}
table(tapply(dxr1$padj < 0.05, dxr1$groupID, any))
```

Exons showing differential exon usage:

```{r}
table(dxr1$padj < 0.05)
```

```{r}
plotMA( dxr1, cex = 0.8, alpha = .05)
```

```{r}
res <- dxr1 %>% 
    as_tibble() %>% 
    select_at(vars(-matches("genomic"))) %>% 
    select_at(vars(-matches("countData"))) %>% 
    select(-transcripts) %>% 
    rename(ensembl_gene_id_version = "groupID")
```

```{r annotation}
ensembl = biomaRt::useMart("ensembl",dataset="mmusculus_gene_ensembl")
annotation <- biomaRt::getBM(
    attributes = c("ensembl_gene_id_version", "mgi_symbol", "description",
        "gene_biotype"), 
    filters = "ensembl_gene_id_version", 
    values = unique(res$ensembl_gene_id_version), 
    mart = ensembl) %>% 
    as_tibble() %>% 
    distinct(ensembl_gene_id_version, .keep_all = TRUE)

res <- res %>% 
    left_join(annotation, by = "ensembl_gene_id_version")
```


```{r resultsTable}
res %>% 
    filter(padj < .05) %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling() %>%
    kableExtra::scroll_box(width = "100%", height = "600px")

dir.create(snakemake@params[["out_folder"]], showWarnings = FALSE)
res %>% 
    arrange(padj) %>% 
    write_csv(file.path(snakemake@params[["out_folder"]], "all_deu.csv.gz"))
```


