---
title: "DEXSeq analysis of illumina reads, iBAT, cold"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
    html_document:
        toc: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, cache = FALSE)
```

```{r loadPackages}
source("packrat/init.R")
library("dplyr")
library("DEXSeq")
library("readr")
library("ggplot2")
library("ggthemes")
    select <- dplyr::select
BPPARAM = BiocParallel::MulticoreParam(snakemake@threads[[1]])
```

```{r}
#save.image("dexseq_illumina.RData")
```

```{r importData}
dxd <- readRDS(snakemake@input[["dxd"]])
```

```{r normalisation}
dxd <- estimateSizeFactors(dxd)
```

# Dispersion estimation

```{r dispersionEstimation}
dxd <- estimateDispersions(dxd, BPPARAM = BPPARAM)
plotDispEsts(dxd)
```

# Differential exon usage

```{r deu}
dxd <- testForDEU(dxd, BPPARAM = BPPARAM)
dxd <- estimateExonFoldChanges(dxd, fitExpToVar = "condition_temp",
    BPPARAM = BPPARAM)
```

```{r results}
dxr1 <- DEXSeqResults(dxd)
```

Genes showing differential exon usage:

```{r}
table(tapply(dxr1$padj < 0.05, dxr1$groupID, any))
```

Exons showing differential exon usage:

```{r}
table(dxr1$padj < 0.05)
```

```{r}
plotMA(dxr1, cex = 0.8, alpha = .05)
```

```{r}
res <- dxr1 %>% 
    as_tibble() %>% 
    select_at(vars(-matches("genomic"))) %>% 
    select_at(vars(-matches("countData"))) %>% 
    select(-X4, -X22) %>% 
    dplyr::rename(ensembl_gene_id_version = "groupID")
```

```{r annotation}
biomaRt_gene <- read_rds(snakemake@input[["biomaRt_gene"]])

res <- res %>% 
    left_join(biomaRt_gene, by = "ensembl_gene_id_version") %>% 
    select(ensembl_gene_id_version, mgi_symbol, description, gene_biotype, everything())
```

## Resultstable

```{r resultsTable}
res %>% 
    filter(padj < .05) %>%
    arrange(-abs(log2fold_22_4)) %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling() %>%
    kableExtra::scroll_box(width = "100%", height = "600px")

dir.create(snakemake@params[["out_folder"]], showWarnings = FALSE)
res %>% 
    write_csv(file.path(snakemake@params[["out_folder"]], "all_deu.csv.gz"))
```

## Gene Plots

```{r}
plotDEXSeq(dxr1, "ENSMUSG00000020654.15", .05, "condition_temp",
    norCounts = TRUE, legend = TRUE)

dxr1 %>% 
    as_tibble() %>% 
    filter(groupID == "ENSMUSG00000020654.15") %>%
    select_at(vars(matches("count|featureID"))) %>% 
    tidyr::gather(key = sample, value = counts, -featureID) %>% 
    mutate(sample = substr(sample, 11, 20)) %>% 
    left_join(as_tibble(colData(dxd)) %>% filter(exon == "this"), by = "sample") %>% 
    ggplot(aes(featureID, counts)) +
        geom_point(aes(colour = condition_temp)) +
        scale_y_log10() +
        scale_color_viridis_d() +
        theme_tufte() +
        theme(axis.text.x = element_text(angle = 90, hjust = 1))
```


