---
title: "DRIMSeq stageR analysis"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
    html_document:
        toc: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, cache = FALSE)
```

```{r library, include=FALSE}
source(".Rprofile")
library("dplyr")
library("readr")
library("DRIMSeq")
library("ggplot2")
library("ggthemes")
library("cowplot")
library("Gviz")
library("AnnotationDbi")
select <- dplyr::select
```

```{r importData}
dmds <- read_rds(snakemake@input[["dmds"]])
biomart_gene <- read_rds(snakemake@input[["biomaRt_gene"]])
biomart_tx <- read_rds(snakemake@input[["biomaRt_tx"]])
res_stageR <- read_csv(snakemake@input[["res"]])
```

# Table of transcripts per gene after filtering.

```{r}
table(table(counts(dmds)$gene_id))
```

# Precision

```{r plotPrecision}
plotPrecision(dmds) +
    geom_point(size = 2) +
    scale_color_viridis_c()
```

```{r buildResultsTables}
res_gene <- DRIMSeq::results(dmds)
res_tx <- DRIMSeq::results(dmds, level = "feature")
```

# Results tables

## Gene Level

```{r kableDrimSeqResults_gene}
res_gene %>% 
    as_tibble() %>%
    arrange(adj_pvalue) %>% 
    filter(adj_pvalue < .05) %>% 
    dplyr::rename(ensembl_gene_id_version = "gene_id") %>% 
    left_join(biomart_gene, by = "ensembl_gene_id_version") %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling() %>%
    kableExtra::scroll_box(width = "100%", height = "600px")
```

## Transcript level

```{r kableDrimSeqResults_tx}
res_tx %>% 
    as_tibble() %>%
    arrange(adj_pvalue) %>% 
    filter(adj_pvalue < .05) %>% 
    dplyr::rename(ensembl_gene_id_version = "gene_id",
        ensembl_transcript_id_version = "feature_id") %>% 
    left_join(biomart_tx, 
        by = c("ensembl_transcript_id_version", "ensembl_gene_id_version")) %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling() %>%
    kableExtra::scroll_box(width = "100%", height = "600px")
```

## StageR

StageR looks at the pvalues for both genes and transcripts thereby allowing to 
control the overall false discovery rate. (While in the two tables aboth, the 
false discovery rate is controlled withing each table, it is not controlled over
both.)

```{r stageRresultsTables}
res_stageR %>%
    filter(gene < .05 | transcript < .05) %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling() %>%
    kableExtra::scroll_box(width = "100%", height = "600px")
```


# Plots

The cross is the proportion estimated by DRIMSeq. The red transcripts are significantly changed.
Note that sometimes there is an indication of significant DTU at the genelevel, but no 
single transcript changes significantly.

```{r}
gene_ids <- res_stageR %>%
    filter(gene < .05 | transcript < .05) %>%
    pull(ensembl_gene_id_version) %>% 
    unique()

txdb <- loadDb(snakemake@input[["txdb"]])
txdb_dump <- as.list(txdb)

plot.transcripts <- function(gene_id) {
    gene_info <- AnnotationDbi::select(txdb, keys = gene_id,
            keytype = "GENEID",
            columns = c("TXCHROM", "TXSTART", "TXEND")) %>%
        as_tibble() %>%
        group_by(GENEID) %>%
        summarise(TXCHROM = unique(TXCHROM),
            TXSTART = min(TXSTART),
            TXEND = max(TXEND))
    
    # filter significant transcripts
    sig_tx <- res_stageR %>%
        filter(ensembl_gene_id_version == gene_id) %>%
        filter(transcript < .05) %>%
        pull(ensembl_transcript_id_version)
    
    # subset txdb
    all_tx <- res_stageR %>%
        filter(ensembl_gene_id_version == gene_id) %>%
        pull(ensembl_transcript_id_version)
    
    all_tx_ids <- txdb_dump[[1]] %>%
        as_tibble() %>% 
        filter(tx_name %in% all_tx) %>% 
        pull(tx_id)

    new_txdb_dump <- vector("list", 4)
    new_txdb_dump[1:3] <- txdb_dump[1:3] %>%
        purrr::map(filter, tx_id %in% all_tx_ids)
    new_txdb_dump[4] <- txdb_dump[4]
    
    new_txdb <- do.call(makeTxDb, new_txdb_dump)

    # annotation track
    atrack <- GeneRegionTrack(new_txdb, chromosome = gene_info$TXCHROM,
        from = gene_info$TXSTART, to = gene_info$TXEND,
        name = "GENCODE M23")
    symbol(atrack) <- symbol(atrack) %>%
        tibble::enframe() %>%
        dplyr::select(ensembl_transcript_id_version = "value") %>%
        left_join(biomart_tx, by = "ensembl_transcript_id_version") %>%
        pull(mgi_symbol)
    
    # set colour of significant transcripts
    feature(atrack) <- if_else(transcript(atrack) %in% sig_tx, "bad", "good")
    interestcolor <- list("bad" = "red", "good" = "#FFD58A")
    displayPars(atrack) <- interestcolor
    
    gtrack <- GenomeAxisTrack()
    
    options(ucscChromosomeNames = FALSE)
    
    plotTracks(list(gtrack, atrack),
        just.group = "above",
        transcriptAnnotation = "transcript",
        from = gene_info$TXSTART,
        to = gene_info$TXEND,
        extend.left = 250,
        extend.right = 250,
        background.title = "darkblue",
        fontsize = 15, cex.axis = .4,
        min.height = 0, coverageHeight = 0.08, minCoverageHeight = 0)
}

for (ens_gene_id in gene_ids) {
    
    print(ens_gene_id)

    plot.transcripts(ens_gene_id)

    df <- proportions(dmds) %>% 
        filter(gene_id == ens_gene_id) %>%
        dplyr::select(-gene_id) %>% 
        tidyr::gather("sample_id", "estimated_proportion", -feature_id) %>% 
        mutate(sample_id = substr(sample_id, 2, nchar(sample_id)))
    
    df2 <- counts(dmds) %>% 
        filter(gene_id == ens_gene_id) %>%
        select(-gene_id) %>% 
        mutate_if(is.numeric, function(x) x / sum(x)) %>% 
        tidyr::gather("sample_id", "proportion", -feature_id) %>% 
        mutate(sample_id = substr(sample_id, 2, nchar(sample_id))) %>% 
        left_join(df, by = c("feature_id", "sample_id")) %>% 
        left_join(DRIMSeq::samples(dmds), by = "sample_id")
    
    pl1 <- df2 %>% 
        ggplot(aes(condition_temp, proportion)) +
        geom_point(aes(colour = feature_id), position = position_dodge(width = .5)) +
        stat_summary(aes(colour = feature_id), geom = "crossbar",
            width = .5, lwd = .25,
            position = position_dodge(width = .5)) +
        geom_point(aes(y = estimated_proportion, colour = feature_id),
            position = position_dodge(width = .5),
            shape = 3, size = 3) +
        scale_color_colorblind() +
        theme_tufte() +
        geom_rangeframe(sides = "l") +
        xlab("T [°C]")
    
    if (snakemake@params[["axis"]] == "illumina") {
        pl1 <- pl1 +
            ylab("Proportion of scaled TPM")
    } else {
        pl1<- pl1 +
            ylab("Proportion of counts")
    }
   
    pl2 <- counts(dmds) %>% 
        filter(gene_id == ens_gene_id) %>%
        select(-gene_id) %>% 
        tidyr::gather("sample_id", "count", -feature_id) %>%
        mutate(sample_id = substr(sample_id, 2, nchar(sample_id))) %>% 
        left_join(DRIMSeq::samples(dmds), by = "sample_id") %>% 
        ggplot(aes(condition_temp, count)) +
        geom_point(aes(colour = feature_id), position = position_dodge(width = .5)) +
        stat_summary(aes(colour = feature_id), geom = "crossbar",
            width = .5, lwd = .25,
            position = position_dodge(width = .5)) +
        scale_color_colorblind() +
        theme_tufte() +
        geom_rangeframe(sides = "l") +
        xlab("T [°C]")
    
    if (snakemake@params[["axis"]] == "illumina") {
        pl2 <- pl2 +
            ylab("scaled TPM")
    } else {
        pl2 <- pl2 +
            ylab("Counts")
    }
    
    pl_gr <- plot_grid(pl1 + theme(legend.position="none"),
        pl2 + theme(legend.position="none"),
        get_legend(pl1),
        nrow = 1, rel_widths = c(1, 1, .55))
    
    print(pl_gr)
}
```


# Session info

```{r sessionInfo}
sessionInfo()
```
