---
title: "DTU, Illumina reads transcript level, all samples"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
    html_document:
        toc: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, cache = FALSE)
source("packrat/init.R")
```

```{r library, include=FALSE}
library("dplyr")
library("readr")
library("DRIMSeq")
library("ggplot2")
library("stageR")
BPPARAM = BiocParallel::MulticoreParam(snakemake@threads[[1]])
```

# Data import

```{r createTxdb}
txdb <- GenomicFeatures::makeTxDbFromGFF(snakemake@input[["annotation"]],
        format = "gtf", circ_seqs = character())

txdf <- txdb %>%
    AnnotationDbi::select(.,
        keys =  AnnotationDbi::keys(., keytype = "GENEID"),
        keytype = "GENEID",
        columns = "TXNAME") %>%
    dplyr::select(TXNAME, GENEID) %>% 
    group_by(GENEID) %>% 
    mutate(ntx = row_number()) %>% 
    ungroup()
```


```{r importScaledTpm}
cts <- readRDS(snakemake@input[["tpm"]])
txdf <- txdf[match(rownames(cts),txdf$TXNAME),]

counts <- cts %>% 
    as_tibble(rownames = "TXNAME") %>% 
    left_join(txdf, by = "TXNAME") %>% 
    dplyr::rename(gene_id = "GENEID",
        feature_id = "TXNAME") %>% 
    as.data.frame()
```


```{r importSampleInfo}
sample_info <- read_csv(snakemake@input[["sample_info"]]) %>%
    filter(!is.na(illumina)) %>%
    mutate_at(vars(matches("condition")), as.factor) %>%
    mutate(path = file.path("salmon", illumina, "quant.sf"),
        condition_temp = relevel(condition_temp, ref = "22")) %>%
    as.data.frame()
```

```{r createDrimSeqDataSet}
dmds <- dmDSdata(counts = counts, samples = sample_info)
```

```{r filterLowlyExpressed, echo = TRUE}
# total number of samples
n <- 12
# number of samples in the smallest group
n.small <- 6

dmds <- dmFilter(dmds,
              min_samps_feature_expr = n.small, min_feature_expr = 10,
              min_samps_feature_prop = n.small, min_feature_prop = 0.1,
              min_samps_gene_expr = n, min_gene_expr = 10)
```

Table of transcripts per gene after filtering.

```{r}
table(table(counts(dmds)$gene_id))
```

# Run DrimSeq

```{r fitAndTestModels}
design_full <- model.matrix(~condition_temp, data = DRIMSeq::samples(dmds))

dmds <- dmPrecision(dmds, design = design_full, 
    BPPARAM = BPPARAM)
dmds <- dmFit(dmds, design = design_full)
dmds <- dmTest(dmds, coef = "condition_temp4")
```

```{r plotPrecision}
plotPrecision(dmds) +
    geom_point(size = 2) +
    scale_color_viridis_c()
```

```{r buildResultsTables}
res_gene <- DRIMSeq::results(dmds)
res_tx <- DRIMSeq::results(dmds, level = "feature")
```

```{r annotation}
# get annotation from biomart
ensembl = biomaRt::useMart("ensembl", dataset = "mmusculus_gene_ensembl")
annotation_gene <- biomaRt::getBM(
    attributes = c("ensembl_gene_id_version", "mgi_symbol", "description",
        "gene_biotype"), 
    filters = "ensembl_gene_id_version", 
    values = res_gene$gene_id, 
    mart = ensembl)

annotation_tx <- biomaRt::getBM(
    attributes = c("ensembl_transcript_id_version", "ensembl_gene_id_version",
        "mgi_symbol", "description",
        "gene_biotype",
        "transcript_biotype", "transcript_length"), 
    filters = "ensembl_transcript_id_version", 
    values = res_tx$feature, 
    mart = ensembl)
```

```{r kableDrimSeqResults_gene}
res_gene %>% 
    as_tibble() %>%
    arrange(adj_pvalue) %>% 
    filter(adj_pvalue < .05) %>% 
    dplyr::rename(ensembl_gene_id_version = "gene_id") %>% 
    left_join(annotation_gene, by = "ensembl_gene_id_version") %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling() %>%
    kableExtra::scroll_box(width = "100%", height = "600px")
```

#### Transcript level

```{r kableDrimSeqResults_tx}
res_tx %>% 
    as_tibble() %>%
    arrange(adj_pvalue) %>% 
    filter(adj_pvalue < .05) %>% 
    dplyr::rename(ensembl_gene_id_version = "gene_id",
        ensembl_transcript_id_version = "feature_id") %>% 
    left_join(annotation_tx, by = c("ensembl_transcript_id_version",
        "ensembl_gene_id_version")) %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling() %>%
    kableExtra::scroll_box(width = "100%", height = "600px")
```

```{r exportDrimSeqResults}
dir.create(snakemake@params[["out_folder"]], showWarnings = FALSE)

res_gene %>% 
    as_tibble() %>%
    arrange(adj_pvalue) %>% 
    dplyr::rename(ensembl_gene_id_version = "gene_id") %>% 
    left_join(annotation_gene, by = "ensembl_gene_id_version") %>% 
    write_csv(file.path(snakemake@params[["out_folder"]], "genelevel_drimseq_dtu.csv.gz"))

res_tx %>% 
    as_tibble() %>%
    arrange(adj_pvalue) %>% 
    dplyr::rename(ensembl_gene_id_version = "gene_id",
        ensembl_transcript_id_version = "feature_id") %>% 
    left_join(annotation_tx, by = c("ensembl_transcript_id_version",
        "ensembl_gene_id_version")) %>%
    as_tibble() %>% 
    arrange(adj_pvalue) %>% 
    write_csv(file.path(snakemake@params[["out_folder"]], "txlevel_drimseq_dtu.csv.gz"))
```

```{r exchangeNAfor1}
no.na <- function(x) ifelse(is.na(x), 1, x)

res_gene$pvalue <- no.na(res_gene$pvalue)
res_tx$pvalue <- no.na(res_tx$pvalue)
```

# StageR

```{r stageRdataset}
pScreen <- res_gene$pvalue
names(pScreen) <- res_gene$gene_id

pConfirmation <- matrix(res_tx$pvalue, ncol = 1)
rownames(pConfirmation) <- res_tx$feature_id

tx2gene <- res_tx[ , c("feature_id", "gene_id")]

stageRObj <- stageRTx(pScreen = pScreen, pConfirmation = pConfirmation,
        pScreenAdjusted = FALSE, tx2gene = tx2gene)
stageRObj <- stageWiseAdjustment(stageRObj, method = "dtu", alpha = 0.05)
drim.padj <- getAdjustedPValues(stageRObj, order = TRUE,
    onlySignificantGenes = FALSE)
```

StageR looks at the pvalues for both genes and transcripts thereby allowing to 
control the overall false discovery rate. (While in the two tables aboth, the 
false discovery rate is controlled withing each table, it is not controlled over
both.)

```{r stageRresultsTables}
res_stageR <- drim.padj %>% 
    as_tibble() %>%
    dplyr::rename(ensembl_gene_id_version = "geneID",
        ensembl_transcript_id_version = "txID") %>% 
    left_join(annotation_tx, by = c("ensembl_transcript_id_version", "ensembl_gene_id_version"))

res_stageR %>% 
    write_csv(file.path(snakemake@params[["out_folder"]], "stageR_drimseq_dtu.csv.gz"))

res_stageR %>% 
    filter(gene < .05 | transcript < .05) %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling() %>%
    kableExtra::scroll_box(width = "100%", height = "600px")
```

# Session info

```{r sessionInfo}
sessionInfo()
```
