---
title: "DRIMSeq stageR analysis"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
    html_document:
        toc: true
        toc_float: true
        toc_depth: 2
        fig_caption: false
        dev: "svg"
editor_options:
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE,
    cache = FALSE, fig.width = 5.35538 * 2/3, fig.asp = 0.7, out.width = 800)
```

```{r library, include=FALSE}
source(".Rprofile")
library("dplyr")
library("readr")
library("ggplot2")
library("ggthemes")
library("patchwork")
library("DRIMSeq")
library("IsoformSwitchAnalyzeR")
library("topGO")
select <- dplyr::select
```

```{r importData, include=FALSE}
dmds <- read_rds(snakemake@input[["dmds"]])
biomart_gene <- read_rds(snakemake@input[["biomaRt_gene"]])
biomart_tx <- read_rds(snakemake@input[["biomaRt_tx"]])
res_stageR <- read_csv(snakemake@input[["res"]])
switchList <- read_rds(snakemake@input[["switchList"]])
switchList <- analyzeSignalP(switchList,
    pathToSignalPresultFile = snakemake@input[["signalP"]])
switchList <- analyzePFAM(switchList,
    pathToPFAMresultFile = snakemake@input[["pfam"]],
    showProgress = FALSE)
targetP <- read_tsv(snakemake@input[["targetP"]], skip = 1) %>% 
    select("# ID", "Prediction")
CPAT <- read_tsv(snakemake@input[["cpat"]]) %>% 
    select(seq_ID, Coding_prob, ORF) %>% 
    tidyr::separate(seq_ID,
        c("ensembl_transcript_id_version", "ensembl_gene_id_version"),
        sep = '_') %>% 
    select(-ensembl_gene_id_version) %>% 
    mutate(ensembl_transcript_id_version =  if_else(substr(ensembl_transcript_id_version, 1, 7) == "ENSMUST",
        ensembl_transcript_id_version, tolower(ensembl_transcript_id_version)))

switchList <- analyzeSwitchConsequences(switchList,
    c("intron_retention", "ORF_seq_similarity", "NMD_status", 
      "domains_identified", "signal_peptide_identified"))
```

# Overview

## Table of transcripts per gene after filtering.

```{r}
table(table(counts(dmds)$gene_id))
```

## Precision

```{r plotPrecision}
plotPrecision(dmds) +
    geom_point(size = 2) +
    scale_color_viridis_c()
```

```{r buildResultsTables}
res_gene <- DRIMSeq::results(dmds)
res_tx <- DRIMSeq::results(dmds, level = "feature")
```

# Results tables

## Gene Level

```{r kableDrimSeqResults_gene}
res_gene %>% 
    as_tibble() %>%
    arrange(adj_pvalue) %>% 
    filter(adj_pvalue < .05) %>% 
    dplyr::rename(ensembl_gene_id_version = "gene_id") %>% 
    left_join(biomart_gene, by = "ensembl_gene_id_version") %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling() %>%
    kableExtra::scroll_box(width = "100%", height = "600px")
```

## Transcript level

```{r kableDrimSeqResults_tx}
res_tx %>% 
    as_tibble() %>%
    arrange(adj_pvalue) %>% 
    filter(adj_pvalue < .05) %>% 
    dplyr::rename(ensembl_gene_id_version = "gene_id",
        ensembl_transcript_id_version = "feature_id") %>% 
    left_join(biomart_tx, 
        by = c("ensembl_transcript_id_version", "ensembl_gene_id_version")) %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling() %>%
    kableExtra::scroll_box(width = "100%", height = "600px")
```

## StageR

StageR looks at the pvalues for both genes and transcripts thereby allowing to 
control the overall false discovery rate. (While in the two tables aboth, the 
false discovery rate is controlled withing each table, it is not controlled over
both.)

```{r stageRresultsTables}
n_sig <- res_stageR %>% 
    filter(gene < .05) %>% 
    pull(ensembl_gene_id_version) %>% 
    unique() %>% 
    length()

sprintf("%s genes with significant DTU detected.", n_sig)    

res_stageR %>%
    filter(gene < .05 | transcript < .05) %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling() %>%
    kableExtra::scroll_box(width = "100%", height = "600px")
```

## GO

```{r}
all_genes <- res_stageR$ensembl_gene_id_version %>%
    unique

# function to generate genelist
get.geneList <- function(genes) {
    gl <- as.factor(as.integer(all_genes %in% genes))
    names(gl) <- all_genes
    # remove version
    names(gl) <- unlist(lapply(stringr::str_split(names(gl), "[.]"), "[[",1))
    gl
}

# function to actually create topGO object
make.topGO <- function(geneList, description) {
    new("topGOdata",
        description = description,
        ontology = "BP",
        allGenes = geneList,
        nodeSize = 10,
        annot = annFUN.org,
        ID = "ensembl",
        mapping = "org.Mm.eg")
}

# conveniance function to get results from topGO
get.results <- function(topGO) {
    topGOres <- runTest(topGO, algorithm = "parentchild", statistic = "fisher")
    GenTable(topGO, p = topGOres, orderBy = "p", ranksOf = "p", topNodes = 100) %>%
        as_tibble() %>%
        mutate(p = as.double(p))
}

# run the analysis
res_go <- res_stageR %>%
    filter(gene < .05) %>%
    pull(ensembl_gene_id_version) %>%
    unique() %>%
    get.geneList() %>%
    make.topGO(description = "allRegulatedGenes") %>%
    get.results()

res_go %>%
    filter(p < .05) %>%
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling() %>%
    kableExtra::scroll_box(width = "100%", height = "600px")
```

# Isoform switch analyser

## Legend

* ES: Exon Skipping. Compared to the hypothetical pre-RNA a single exon was skipped in the
isoform analyzed (for every ES event annotated).
* MEE: Mutually exclusive exon. Special case were two isoforms form the same gene contains
two mutually exclusive exons and which are not found in any of the other isoforms from that
gene.
* MES: Multiple Exon Skipping. Compared to the hypothetical pre-RNA multiple consecutive
exon was skipped in the isoform analyzed (for every MES event annotated).
* IR: Intron Retention. Compared to the hypothetical pre-RNA an intron was retained in the
isoform analyzed.
* A5: Alternative 5’end donor site. Compared to the hypothetical pre-RNA an alternative 5’end
donor site was used. Since it is compared to the pre-RNA, the donor site used is per definition
more upstream than the the pre-RNA (the upstream exon is shorter).
* A3: Alternative 3’end acceptor site. Compared to the hypothetical pre-RNA an alternative
3’end acceptor site was used. Since it is compared to the pre-RNA, the donor site used is per
definition more downstream than the the pre-RNA (the downstream exon is shorter).
* ATSS: Alternative Transcription Start Sites. Compared to the hypothetical pre-RNA an alter-
native transcription start site was used. Since it is compared to the pre-RNA, the ATSS site
used is per definition more downstream than the pre-RNA.
* ATTS: Alternative Transcription Termination Sites. Compared to the hypothetical pre-RNA an
alternative transcription Termination sites was used. Since it is compared to the pre-RNA, the
ATTS site used is per definition more upstream than the the pre-RNA.


## Splice analysis (significant genes)

```{r}
df <- extractSplicingSummary(switchList,
        onlySigIsoforms = TRUE,
        asFractionTotal = TRUE,
        returnResult = TRUE,
        plotGenes = TRUE,
        dIFcutoff = 0,
        plot = FALSE) %>% 
    select(AStype, splicingResult, nrGenesWithConsequences, geneFraction) %>% 
    mutate(splicingResult = if_else(grepl("more", splicingResult), "cold", "room temperature")) %>% 
    group_by(AStype) %>% 
    mutate(tot = sum(nrGenesWithConsequences)) %>% 
    ungroup() %>% 
    filter(tot != 0)

df %>% 
    ggplot(aes(AStype, geneFraction)) +
    geom_bar(aes(fill = splicingResult), stat = "identity") +
    theme_tufte() +
    geom_rangeframe(sides = "l", 
        data = df %>%
            group_by(AStype) %>%
            summarise (geneFraction = sum(geneFraction)) %>%
            add_row(AStype = "ATSS", geneFraction = 0)) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    scale_fill_colorblind(name = "isoform used more at") +
    scale_y_continuous(labels = scales::percent) +
    ylab("Fraction of genes") +
    xlab("Alternative splicing category")


extractSplicingEnrichment(switchList,
        returnResult = TRUE,
        returnSummary = TRUE, 
        onlySigIsoforms = TRUE,
        dIFcutoff = 0,
        plot = FALSE) %>% 
    as_tibble() %>% 
    select(-Comparison, -condition_1, -condition_2) %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling()
```

## Genome wide splice analysis

```{r}
extractSplicingGenomeWide(switchList,
    splicingToAnalyze = c('ES', 'MES', "IR", 'ATSS', "ATTS"),
    returnResult = FALSE)
```

## Splice consequences (impact on protein)

```{r}
n_tot <- extractSwitchSummary(switchList,
        filterForConsequences = FALSE,
        dIFcutoff = 0)$nrGenes

n_consequences <- extractSwitchSummary(switchList, 
        filterForConsequences = TRUE,
        dIFcutoff = 0)$nrGenes

sprintf("%s of %s genes with DTU show consequences for the predicted protein.",
    n_consequences, n_tot)

res_orf <- switchList$switchConsequence %>% 
    as_tibble() %>% 
    select(gene_id, switchConsequence) %>% 
    tidyr::drop_na() %>%
    group_by(gene_id) %>% 
    summarise(switchConsequence = paste(unique(switchConsequence), collapse = ';')) %>% 
    rename(gene_id = "ensembl_gene_id_version")
    
res_stageR %>%
    filter(gene < .05 | transcript < .05) %>% 
    select(ensembl_gene_id_version, mgi_symbol, description, gene_biotype, gene) %>% 
    dplyr::rename(padj = "gene") %>% 
    distinct() %>% 
    left_join(res_orf, by = "ensembl_gene_id_version") %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling() %>% 
    kableExtra::scroll_box(width = "100%", height = "600px")
```

```{r}
df <- extractConsequenceSummary(switchList,
        asFractionTotal = TRUE,
        returnResult = TRUE,
        plotGenes = TRUE,
        dIFcutoff = 0,
        plot = FALSE) %>% 
    select(featureCompared, switchConsequence, nrGenesWithConsequences, geneFraction)

df %>% 
    ggplot(aes(featureCompared, geneFraction)) +
    geom_bar(aes(fill = switchConsequence), stat = "identity") +
    theme_tufte() +
    geom_rangeframe(sides = "l", 
        data = df %>%
            group_by(featureCompared) %>%
            summarise (geneFraction = sum(geneFraction)) %>%
            add_row(featureCompared = "Intron retention", geneFraction = 0)) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    scale_fill_viridis_d() +
    scale_y_continuous(labels = scales::percent) +
    ylab("Fraction of genes") +
    xlab(NULL)


extractConsequenceEnrichment(switchList,
        returnResult = TRUE,
        returnSummary = TRUE, 
        dIFcutoff = 0,
        plot = FALSE) %>% 
    as_tibble() %>% 
    select(-feature, -condition_1, -condition_2) %>% 
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling()
```


# Plots

The proportions and counts are from drimseq and therefore only for transcripts that have passed the filter criteria.
The cross is the proportion estimated by DRIMSeq. The red transcripts are significantly changed.
Note that sometimes there is an indication of significant DTU at the genelevel, but no 
single transcript changes significantly.

```{r, results="asis"}
gene_ids <- res_stageR %>%
    filter(gene < .05 | transcript < .05) %>%
    pull(ensembl_gene_id_version) %>% 
    unique()

for (ens_gene_id in gene_ids) {
    
    # gene info for header
    gene_info <- res_stageR %>% 
        filter(ensembl_gene_id_version == ens_gene_id) %>% 
        select(mgi_symbol, ensembl_gene_id_version) %>% 
        distinct()
    
    cat('\n')  
    cat("## ", gene_info$mgi_symbol, gene_info$ensembl_gene_id_version, "\n") 

    # table with transcript level information
    res_stageR %>% 
        filter(ensembl_gene_id_version == ens_gene_id) %>% 
        left_join(biomart_tx, by = c("ensembl_transcript_id_version", "ensembl_gene_id_version")) %>% 
        dplyr::select(ensembl_transcript_id_version, transcript_biotype, transcript) %>% 
        left_join(targetP, by = c("ensembl_transcript_id_version" = "# ID")) %>% 
        left_join(CPAT, by = "ensembl_transcript_id_version") %>% 
        dplyr::rename(padj = "transcript", targetP = "Prediction",
            CPAT = "Coding_prob", ORF_length = "ORF") %>% 
        knitr::kable(format = "html") %>%
        kableExtra::kable_styling() %>% 
        print()
    
    # plot of isoforms with domains
    switchplot <- switchPlotTranscript(switchList, gene = ens_gene_id,
        localTheme = theme_bw(base_size = 10, base_family = "Helvetica")) +
        scale_fill_colorblind() +
        ggtitle(NULL)

    print(switchplot)
    
    df <- proportions(dmds) %>% 
        filter(gene_id == ens_gene_id) %>%
        dplyr::select(-gene_id) %>% 
        tidyr::gather("sample_id", "estimated_proportion", -feature_id) %>% 
        mutate(sample_id = substr(sample_id, 2, nchar(sample_id)))
    
    df2 <- counts(dmds) %>% 
        filter(gene_id == ens_gene_id) %>%
        select(-gene_id) %>% 
        mutate_if(is.numeric, function(x) x / sum(x)) %>% 
        tidyr::gather("sample_id", "proportion", -feature_id) %>% 
        mutate(sample_id = substr(sample_id, 2, nchar(sample_id))) %>% 
        left_join(df, by = c("feature_id", "sample_id")) %>% 
        left_join(DRIMSeq::samples(dmds), by = "sample_id")
    
    pl1 <- df2 %>% 
        mutate(condition_temp = relevel(condition_temp, "22")) %>% 
        ggplot(aes(feature_id, proportion)) +
        geom_point(aes(colour = condition_temp), position = position_dodge(width = .5)) +
        stat_summary(
            aes(colour = condition_temp),
            fun = mean, geom = "crossbar",
            position = position_dodge(width = .5),
            width = .5,
            fatten = 1.5, lwd = .25) +
        stat_summary(
            aes(colour = condition_temp),
            geom = "errorbar", fun.data = mean_se,
            width = .5 * .666,
            lwd = .25,
            position = position_dodge(width = .5)) +
        geom_point(aes(y = estimated_proportion, group = condition_temp),
            position = position_dodge(width = .5),
            shape = 3, size = 3) +
        scale_color_brewer(palette = "Dark2", name = NULL, labels = c("22" = "22°C", "4" = "4°C")) +
        theme_tufte(base_size = 8, base_family = "Helvetica") +
        geom_rangeframe(sides = "l") +
        xlab(NULL)

    if (snakemake@params[["axis"]] == "illumina") {
        pl1 <- pl1 +
            ylab("Proportion of scaled TPM")
    } else {
        pl1 <- pl1 +
            ylab("Proportion of counts")
    }
   
    pl2 <- counts(dmds) %>% 
        filter(gene_id == ens_gene_id) %>%
        select(-gene_id) %>% 
        tidyr::gather("sample_id", "count", -feature_id) %>%
        mutate(sample_id = substr(sample_id, 2, nchar(sample_id))) %>% 
        left_join(DRIMSeq::samples(dmds), by = "sample_id") %>% 
        mutate(condition_temp = relevel(condition_temp, "22")) %>% 
        ggplot(aes(feature_id, count)) +
        geom_point(aes(colour = condition_temp), position = position_dodge(width = .5)) +
        stat_summary(
            aes(colour = condition_temp),
            fun = mean, geom = "crossbar",
            position = position_dodge(width = .5),
            width = .5,
            fatten = 1.5, lwd = .25) +
        stat_summary(
            aes(colour = condition_temp),
            geom = "errorbar", fun.data = mean_se,
            width = .5 * .666,
            lwd = .25,
            position = position_dodge(width = .5)) +
        scale_color_brewer(palette = "Dark2", name = NULL, labels = c("22" = "22°C", "4" = "4°C")) +
        scale_y_log10() +
        theme_tufte(base_size = 8, base_family = "Helvetica") +
        geom_rangeframe(sides = "l") +
        xlab("T [°C]")

    
    if (snakemake@params[["axis"]] == "illumina") {
        pl2 <- pl2 +
            ylab("scaled TPM")
    } else {
        pl2 <- pl2 +
            ylab("Counts")
    }
    
    pl_gr <- pl1 + pl2 + plot_layout(guides = 'collect')
    
    print(pl_gr)
    
    cat('\n') 
}
```
