---
title: "qPCR dtu validation"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
    html_document:
        toc: true
        toc_float: true
        toc_depth: 2
        fig_caption: false
        dev: "svg"
editor_options:
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, cache = FALSE)
```

```{r library, include=FALSE}
source(".Rprofile")
library("dplyr")
library("readr")
library("purrr")
library("Gviz")
library("AnnotationDbi")
```


```{r browserPlotDefinition}
txdb_gencode <- loadDb(snakemake@input[["txdb"]])
txdb_gencode_dump <- as.list(txdb_gencode)

txdb_flair <- loadDb(snakemake@input[["txdb_flair"]])
txdb_flair_dump <- as.list(txdb_flair)

subset.txdb <- function(txdb, txdb_dump, gene_id) {

    all_tx <- AnnotationDbi::select(txdb, gene_id,
            c("GENEID", "TXNAME"), "GENEID") %>%
        pull(TXNAME)

    all_tx_ids <- txdb_dump[[1]] %>%
        as_tibble() %>%
        filter(tx_name %in% all_tx) %>%
        pull(tx_id)

    new_txdb_dump <- vector("list", 4)
    new_txdb_dump[1:3] <- txdb_dump[1:3] %>%
        purrr::map(filter, tx_id %in% all_tx_ids)
    new_txdb_dump[4] <- txdb_dump[4]

    do.call(makeTxDb, new_txdb_dump)
}

plot.transcripts <- function(gene_id, extend.right = 250, extend.left = 250) {
    
    gene_info <- AnnotationDbi::select(txdb_gencode, keys = gene_id,
            keytype = "GENEID",
            columns = c("TXCHROM", "TXSTART", "TXEND")) %>%
        as_tibble() %>%
        group_by(GENEID) %>%
        summarise(TXCHROM = unique(TXCHROM),
            TXSTART = min(TXSTART),
            TXEND = max(TXEND))
    
    # annotation track
    txdb_gencode_sub <- subset.txdb(txdb_gencode, txdb_gencode_dump, gene_id)
    
    atrack_gencode <- GeneRegionTrack(txdb_gencode_sub,
        chromosome = gene_info$TXCHROM,
        from = gene_info$TXSTART, to = gene_info$TXEND,
        name = "GENCODE")
    
    txdb_flair_sub <- subset.txdb(txdb_flair, txdb_flair_dump, gene_id)
    
    atrack_flair <- GeneRegionTrack(txdb_flair_sub,
        chromosome = gene_info$TXCHROM,
        from = gene_info$TXSTART, to = gene_info$TXEND,
        name = "flair cDNA")

    gtrack <- GenomeAxisTrack(from = gene_info$TXSTART, 
        to = gene_info$TXEND,
        scale = 0.25)

    # bam files
    cdna_warm <- AlignmentsTrack(snakemake@input[["cdna_warm"]],
        isPaired = FALSE,
        genome = "mm10",
        name = "cDNA 22째C",
        chromosome = gene_info$TXCHROM,
        start = gene_info$TXSTART,
        end = gene_info$TXEND)

    # cdna_cold <- AlignmentsTrack(snakemake@input[["cdna_cold"]],
    #     isPaired = FALSE,
    #     genome = "mm10",
    #     name = "cDNA 4째C",
    #     chromosome = gene_info$TXCHROM,
    #     start = gene_info$TXSTART,
    #     end = gene_info$TXEND)

    illumina_warm <- AlignmentsTrack(snakemake@input[["illumina_warm"]],
        isPaired = TRUE,
        genome = "mm10",
        name = "Illu 22째C",
        chromosome = gene_info$TXCHROM,
        start = gene_info$TXSTART,
        end = gene_info$TXEND)

    # illumina_cold <- AlignmentsTrack(snakemake@input[["illumina_cold"]],
    #     isPaired = TRUE,
    #     genome = "mm10",
    #     name = "Illu 4째C",
    #     chromosome = gene_info$TXCHROM,
    #     start = gene_info$TXSTART,
    #     end = gene_info$TXEND)
    
    options(ucscChromosomeNames = FALSE)

    plotTracks(list(gtrack,
            illumina_warm,
            cdna_warm,
            atrack_gencode,
            atrack_flair),
        from = gene_info$TXSTART,
        to = gene_info$TXEND,
#        type = c("pileup"),
        extend.left = extend.left,
        extend.right = extend.right,
        background.title = "darkblue",
        cex.axis = .4,
        min.height = 2, # controls min height of the single reads in the pileup, set to 0 to plot all reads
        sizes = c(.25, rep(1, 2), rep(.25, 2)))
}
```

```{r}
plot.transcripts("ENSMUSG00000024981.6")
plot.transcripts("ENSMUSG00000087382.7", extend.right = 5500)
```

