---
title: "qPCR dtu validation"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
    html_document:
        toc: true
        toc_float: true
        toc_depth: 2
        fig_caption: false
        dev: "svg"
editor_options:
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE,
    cache = FALSE, out.width = 800)
```

```{r library, include=FALSE}
source(".Rprofile")
suppressPackageStartupMessages({
    library("dplyr")
    library("readr")
    library("ggplot2")
    library("ggthemes")
    library("purrr")
    library("ggrepel")
    library("UpSetR")
    library("Gviz")
    library("org.Mm.eg.db")
    library("AnnotationDbi")
    source("R/subsetTxdb.R")
})
```


```{r importData}
df <- read_rds(snakemake@input$cq)
hkp <- c("Gtf2b", "Hprt")
```


```{r detectOutliers, warning=FALSE}
detect.outliers <- function(df, cutoff = 1) {
    out <- df %>%
        group_by_at(vars(matches("sample_id|gene"))) %>%
        mutate(error = max(cq, na.rm = TRUE) - min(cq, na.rm = TRUE)) %>%
        ungroup() %>%
        mutate(high_var = if_else(error > cutoff, TRUE, FALSE)) %>%
        dplyr::select(-error)

    if(any(out$high_var, na.rm = TRUE)) {
        cat(sprintf("%i samples set to NA due to high variance in qPCR
                replicates.\n",
            sum(out$high_var, na.rm = TRUE) / 2))
    } else {
        cat("No samples with high variance detected.\n")
    }

    return(out)
}

df <- df %>%
    map(detect.outliers, cutoff = 1)
```


# Cq values

```{r cq_loop, results="asis"}
for (plate in 1:3) {
    
    cat('\n') 
    cat("## ", plate, "\n") 
    
    
    pl <- df[[plate]] %>%
        ggplot(aes(paste(sample_id, condition_temp), -cq)) +
        geom_point(aes(colour = high_var)) +
        geom_text_repel(aes(label = pos),
            family = "serif", size = 2) +
        facet_wrap(~gene) +
        theme_tufte(base_family = "Helvetica") +
        geom_rangeframe() +
        ylab("cq") +
        xlab(NULL) +
        theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
        scale_color_colorblind()
    
    print(pl)
}
```


# \(\Delta\)Cq

Normalised to Gtf2b/Hprt

```{r average_cq}
get.averageCq <- function(df) {
    df %>%
        mutate(cq = if_else(high_var, NA_real_, cq)) %>%
        group_by_at(vars(matches("sample_id|condition|gene"))) %>%
        summarise(cq = mean(cq, na.rm = TRUE)) %>%
        ungroup()
}

df <- df %>%
    bind_rows() %>%
    get.averageCq()
```

```{r dcq}
get.dCq <- function(df, hkp) {
    ref <- df %>%
        filter(gene %in% hkp) %>%
        group_by_at(vars(matches("sample_id"))) %>%
        summarise(housekeeper = mean(cq)) %>%
        ungroup()

    df %>%
        filter(!gene %in% hkp) %>%
        left_join(ref, by = colnames(df)[grep("sample_id", colnames(df))]) %>%
        mutate(dcq = cq - housekeeper) %>%
        dplyr::select(-cq, -housekeeper)
}

df <- df %>%
    get.dCq(hkp = hkp)
```

```{r}
get.averageDCq <- function(df) {
    df %>%
        group_by_at(vars(matches("condition|gene|sample_id"))) %>%
        summarise(dcq = mean(dcq, na.rm = TRUE)) %>%
        ungroup()
}

df <- df %>%
    get.averageDCq()
```

## All genes

```{r}
df %>%
    mutate(condition_temp = relevel(condition_temp, "22")) %>%
    tidyr::separate(gene, c("gene", "isoform"), sep = '_') %>%
    tidyr::drop_na() %>%
    ggplot(aes(isoform, 2^-dcq)) +
    geom_point(aes(colour = condition_temp),
        size = 1,
        alpha = .5,
        position = position_dodge2(width = .75)) +
    stat_summary(aes(colour = condition_temp),
        fun = mean, geom = "crossbar",
        position = position_dodge(width = .75),
        width = 1,
        fatten = 1, lwd = .5) +
    stat_summary(aes(colour = condition_temp),
        geom = "errorbar", fun.data = mean_se,
        width = .75,
        lwd = .25,
        position = position_dodge(width = .75)) +
    facet_wrap(~gene, scales = "free") +
    scale_y_log10(labels = function(x) x) +
    theme_tufte(base_family = "Helvetica") +
    geom_rangeframe() +
    ylab("Rel. expr. vs. Gtf2b/Hprt") +
    xlab("Isoform") +
    scale_colour_brewer(palette = "Dark2", name = NULL)
```

## Only genes in DTU comparison

```{r}
df %>%
    mutate(condition_temp = relevel(condition_temp, "22")) %>%
    tidyr::separate(gene, c("gene", "isoform"), sep = '_') %>%
    filter(!gene %in% c("Smyd4", "Pex6", "Lipe", "Gnas")) %>% 
    tidyr::drop_na() %>%
    ggplot(aes(isoform, 2^-dcq)) +
    geom_point(aes(colour = condition_temp),
        size = 1,
        alpha = .5,
        position = position_dodge2(width = .75)) +
    stat_summary(aes(colour = condition_temp),
        fun = mean, geom = "crossbar",
        position = position_dodge(width = .75),
        width = 1,
        fatten = 1, lwd = .5) +
    stat_summary(aes(colour = condition_temp),
        geom = "errorbar", fun.data = mean_se,
        width = .75,
        lwd = .25,
        position = position_dodge(width = .75)) +
    facet_wrap(~gene, scales = "free") +
    scale_y_log10(labels = function(x) x) +
    theme_tufte(base_family = "Helvetica") +
    geom_rangeframe() +
    ylab("Rel. expr. vs. Gtf2b/Hprt") +
    xlab("Isoform") +
    scale_colour_brewer(palette = "Dark2", name = NULL)
```

# Statistics

## Gene level

This section contains the p value from the interaction terms from linear models.
It corresponds to the differential transcript usage.

```{r}
pvalues <- df %>%
    dplyr::select(gene, condition_temp, dcq) %>%
    tidyr::separate(gene, c("gene", "isoform"), sep = '_') %>%
    tidyr::drop_na() %>%
    split(.$gene) %>%
    map(lm, formula = dcq ~ isoform * condition_temp) %>%
    map(summary.lm) %>%
    map(function(x) x$coefficients[16]) %>%
    unlist() %>%
    tibble::enframe(name = "mgi_symbol", value = "qpcr_p") %>%
    mutate(qpcr_padj = p.adjust(qpcr_p)) %>%
    mutate(sign = symnum(qpcr_padj,
        cutpoints = c(0, .001, .01, .05, .1, 1),
        symbols = c("***", "**", "*", ".", " ")))

pvalues %>%
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling()
```

## Isoform level

This section contains t-tests on isoform level. It corresponds to differential
transcript expression in the sequencing datasets.

```{r}
df %>%
    dplyr::select(gene, condition_temp, dcq) %>%
    tidyr::separate(gene, c("gene", "isoform"), sep = '_') %>%
    tidyr::drop_na() %>%
    group_by(gene, isoform) %>%
    summarise(log2fc = t.test(dcq ~ condition_temp)$estimate["mean in group 22"] -
                  t.test(dcq ~ condition_temp)$estimate["mean in group 4"],
              p = t.test(dcq ~ condition_temp)$p.value) %>%
    mutate(padj = p.adjust(p)) %>%
    mutate(sign = symnum(padj,
        cutpoints = c(0, .001, .01, .05, .1, 1),
        symbols = c("***", "**", "*", ".", " "))) %>%
    knitr::kable(format = "html") %>%
    kableExtra::kable_styling()
```


# Individual genes

```{r browserPlotDefinition}
txdb_gencode <- loadDb(snakemake@input[["txdb"]])
txdb_gencode_dump <- as.list(txdb_gencode)

txdb_flair <- loadDb(snakemake@input[["txdb_flair"]])
txdb_flair_dump <- as.list(txdb_flair)

txdb_stringtie <- loadDb(snakemake@input[["txdb_stringtie"]])
txdb_stringtie_dump <- as.list(txdb_stringtie)

stringtie_refGene <- read_tsv(snakemake@input$tmap,
    col_types = "cccccidddici",
    na = "-") %>%
    dplyr::select(ref_gene_id, qry_gene_id) %>%
    tidyr::drop_na() %>%
    distinct()

txdb_stringtie_illumina <- loadDb(snakemake@input[["txdb_stringtie_illumina"]])
txdb_stringtie_illumina_dump <- as.list(txdb_stringtie_illumina)

stringtie_illumina_refGene <- read_tsv(snakemake@input$tmap_illumina,
                              col_types = "cccccidddici",
                              na = "-") %>%
    dplyr::select(ref_gene_id, qry_gene_id) %>%
    tidyr::drop_na() %>%
    distinct()

faidx <- read_tsv(snakemake@input$genome,
        col_types = "cidii",
        col_names = c("chrom", "length", "offset", "LINEBASES", "LINEWIDTH")) %>%
    dplyr::select(chrom, length) %>%
    mutate(is_circular = FALSE) %>%
    as.data.frame()

primer_stats <- read_csv(snakemake@input[["primer_stats"]]) %>%
    dplyr::select(mgi_symbol, ensembl_gene_id_version, primer_name, amplified_region) %>%
    tidyr::separate(amplified_region, c("chr", "start", "end")) %>%
    mutate_at(vars(matches("start|end")), as.integer) %>%
    dplyr::select(-chr)


# plot function
plot.transcripts <- function(gene_id, max_cov_illumina = 5, max_cov_cdna = 3, lwd_sashimi_max = 10) {

    gene_info <- AnnotationDbi::select(txdb_gencode, keys = gene_id,
            keytype = "GENEID",
            columns = c("TXCHROM", "TXSTART", "TXEND")) %>%
        as_tibble() %>%
        group_by(GENEID) %>%
        summarise(TXCHROM = unique(TXCHROM),
            TXSTART = min(TXSTART),
            TXEND = max(TXEND))

    primer_info <- primer_stats %>%
        filter(ensembl_gene_id_version == gene_id)

    # gencode
    txdb_gencode_subset <- subset.txdb(txdb_gencode, txdb_gencode_dump, gene_id)

    gencodetrack <- GeneRegionTrack(txdb_gencode_subset,
        chromosome = gene_info$TXCHROM,
        min.height = 3,
        from = gene_info$TXSTART, to = gene_info$TXEND,
        name = "M22")
    

    # flair
    txdb_flair_subset <- subset.txdb(txdb_flair, txdb_flair_dump, gene_id)

    flairtrack <- GeneRegionTrack(txdb_flair_subset,
        chromosome = gene_info$TXCHROM,
        min.height = 3,
        from = gene_info$TXSTART, to = gene_info$TXEND,
        name = "flair cDNA")


    # stringtie cdna
    txdb_stringtie_subset <- subset.txdb(txdb_stringtie, txdb_stringtie_dump,
        stringtie_refGene %>%  filter(ref_gene_id == gene_id) %>%  pull(qry_gene_id))

    stringtietrack <- GeneRegionTrack(txdb_stringtie_subset,
        chromosome = gene_info$TXCHROM,
        min.height = 3,
        from = gene_info$TXSTART, to = gene_info$TXEND,
        name = "stringtie TeloPrime")

    
    # stringtie illumina
    txdb_stringtie_illumina_subset <- subset.txdb(txdb_stringtie_illumina, txdb_stringtie_illumina_dump,
        stringtie_illumina_refGene %>%  filter(ref_gene_id == gene_id) %>%  pull(qry_gene_id))

    stringtietrack2 <- GeneRegionTrack(txdb_stringtie_illumina_subset,
                                      chromosome = gene_info$TXCHROM,
                                      min.height = 3,
                                      from = gene_info$TXSTART, to = gene_info$TXEND,
                                      name = "stringtie Illumina")    
   
    # scale
    gtrack <- GenomeAxisTrack(from = min(gene_info$TXSTART),
        to = max(gene_info$TXEND),
        scale = 0.25)
    
    
    # primer track
    primerTrack <- AnnotationTrack(
        start = c(primer_info[1, ]$start, primer_info[1, ]$end - 20,
            primer_info[2, ]$start, primer_info[2, ]$end - 20),
        width = rep(20, 4),
        group = rep(c(paste(primer_info[1, ]$mgi_symbol, primer_info[1, ]$primer_name),
                paste(primer_info[2, ]$mgi_symbol, primer_info[2, ]$primer_name)),
            each = 2),
        genome = "mm10",
        chromosome = gene_info$TXCHROM,
        name = "qPCR",
        shape = "box",
        groupAnnotation = "group",
        just.group = "above")
    
    # bam files
    illumina_w <- AlignmentsTrack(snakemake@input[["illumina_warm"]],
        isPaired = TRUE,
        genome = "mm10",
        name = "Illumina 22°C",
        type = c("coverage", "sashimi"),
        ylim = c(0, max_cov_illumina),
        transformation = function(x) log1p(x),
        sashimiTransformation = function(x) x,
        lwd.sashimiMax = lwd_sashimi_max,
        sashimiHeight = .33,
        chromosome = gene_info$TXCHROM,
        start = gene_info$TXSTART,
        end = gene_info$TXEND)

    illumina_c <- AlignmentsTrack(snakemake@input[["illumina_cold"]],
        isPaired = TRUE,
        genome = "mm10",
        name = "Illumina 4°C",
        type = c("coverage", "sashimi"),
        ylim = c(0, max_cov_illumina),
        transformation = function(x) log1p(x),
        sashimiTransformation = function(x) x,
        lwd.sashimiMax = lwd_sashimi_max,
        sashimiHeight = .33,
        chromosome = gene_info$TXCHROM,
        start = gene_info$TXSTART,
        end = gene_info$TXEND)

    cdna_w <- AlignmentsTrack(snakemake@input[["cdna_warm"]],
        isPaired = FALSE,
        genome = "mm10",
        name = "cDNA 22°C",
        type = c("coverage", "sashimi"),
        ylim = c(0, max_cov_cdna),
        transformation = function(x) log1p(x),
        sashimiTransformation = function(x) x,
        lwd.sashimiMax = lwd_sashimi_max,
        sashimiHeight = .33,        chromosome = gene_info$TXCHROM,
        start = gene_info$TXSTART,
        end = gene_info$TXEND)

    cdna_c <- AlignmentsTrack(snakemake@input[["cdna_cold"]],
        isPaired = FALSE,
        genome = "mm10",
        name = "cDNA 4°C",
        type = c("coverage", "sashimi"),
        ylim = c(0, max_cov_cdna),
        transformation = function(x) log1p(x),
        sashimiTransformation = function(x) x,
        lwd.sashimiMax = lwd_sashimi_max,
        sashimiHeight = .33,    
        chromosome = gene_info$TXCHROM,
        start = gene_info$TXSTART,
        end = gene_info$TXEND)

    # list of unique exons for sashimi plot
    unique_exons <- c(txdb_flair_subset, txdb_stringtie_subset, txdb_gencode_subset) %>%
        purrr::map(intronsByTranscript) %>% # extra intron information
        purrr::reduce(c) %>% # collapse the lists into one single list
        unlist() %>% # concatenate all exons of all transcripts into one granges object
        unique() # remove duplicate exons
    
    # actual plotting
    options(ucscChromosomeNames = FALSE)

    tracks <- list(
        gtrack,
        illumina_w, illumina_c,
        cdna_w, cdna_c,
        gencodetrack, stringtietrack2, stringtietrack, flairtrack,
        primerTrack)
    
    plotTracks(
        tracks,
        from = min(gene_info$TXSTART),
        to = max(gene_info$TXEND),
        extend.left = 250,
        extend.right = 250,
        background.title = "transparent",
        col.title = "grey10",
        cex = .75,
        cex.axis = .6,
        sashimiFilter = unique_exons,
        lwd = .25,
        sizes = c(
            .1, 
            rep(.5, 2),
            rep(.5, 2),
            rep(.25, 4), 
            .1))
}
```

```{r loop, results="asis", fig.width = 6.85, fig.asp = 1.2}
gene_ids <- tibble(
    symbol =
        c("Acsl5", "Adcy3", "Adtrp",
          "Aldoa", "Cars2", "Dipk1b",
          "Ergic1", "Gnas", "Lipe",
          "Mlxipl", "Pde4d", "Pex6",
          "Ppargc1a", "Scp2", "Smyd4"),
    ensembl_gene_id_version =
        c("ENSMUSG00000024981.6", "ENSMUSG00000020654.15", "ENSMUSG00000058022.14",
            "ENSMUSG00000030695.16", "ENSMUSG00000056228.10", "ENSMUSG00000036186.5", 
            "ENSMUSG00000001576.15", NA_character_, NA_character_,
            "ENSMUSG00000005373.13", "ENSMUSG00000021699.17", NA_character_,
            "ENSMUSG00000029167.13", "ENSMUSG00000028603.15", NA_character_))

gene_ids <- gene_ids %>% 
    tidyr::drop_na()

for (gene in gene_ids$symbol) {

    cat('\n')
    cat("## ", gene, gene_ids$ensembl_gene_id_version[gene_ids$symbol == gene], "\n")

    plot.transcripts(gene_ids$ensembl_gene_id_version[gene_ids$symbol == gene],
        max_cov_illumina = 5, max_cov_cdna = 5)

    cat('\n')
}

```
