---
title: "Comparison of reannotation by flair and stringtie"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
    html_document:
        toc: true
        toc_float: true
        toc_depth: 2
        fig_caption: false
        dev: "svg"
editor_options:
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE,
    cache = FALSE, out.width = 800)
```

```{r library, include=FALSE}
source(".Rprofile")
suppressPackageStartupMessages({
    library("dplyr")
    library("readr")
    library("purrr")
    library("ggplot2")
    library("ggthemes")
    library("UpSetR")
    library("scales")
})
```

```{r dataImport}
# load gffcompare
df <- snakemake@input[["gffcompare"]] %>% 
    set_names(., sub("^.+/gffcmp.", "", .) %>%
                     sub("\\.isoforms.gtf.tmap$", "", .) %>% 
                     sub("\\.gtf.+$", "", .))  %>% 
    map(read_tsv, show_col_types = FALSE) %>%  
    bind_rows(.id = "dataset") %>% 
    mutate(dataset = factor(dataset,
        levels = c("illumina_stringtie", "teloprime_stringtie", "cdna_stringtie",
            "rna_stringtie", "teloprime_flair", "cdna_flair", "rna_flair")))


# load sqanti
df2 <- snakemake@input[["sqanti"]] %>% 
    set_names(., sub("^.+stringtie/", "", .) %>% 
                     sub("^.+flair/", "", .) %>% 
                     sub("_noU.+$", "", .) %>% 
                     sub("\\.iso.+$", "", .) %>% 
                     sub("^.+/", "", .)) %>% 
    map(read_tsv, show_col_types = FALSE) %>% 
    bind_rows(.id = "dataset") %>% 
    mutate(
        dataset = factor(dataset,
            levels = c("illumina_stringtie", "teloprime_stringtie", "cdna_stringtie",
                "rna_stringtie", "teloprime_flair", "cdna_flair", "rna_flair")),
        structural_category = factor(structural_category,
            levels = c("full-splice_match", "incomplete-splice_match", "novel_in_catalog",
                "novel_not_in_catalog", "fusion", "antisense",
                "genic", "genic_intron", "intergenic")))

tracking <- read_tsv(
        snakemake@input$gffcmp_tracking,
        col_names = c("query_transfrag_id", "query_locus_id",
            "reference_gene_id", "class_code", 
            "flair_cdna", "flair_teloprime", "flair_rna",
            "stringtie_cdna", "stringtie_teloprime", "stringtie_rna", "stringtie_illumina"),
        show_col_types = FALSE,
        na = "-")
```


# Gffcompare

```{r}
knitr::include_graphics("http://ccb.jhu.edu/software/stringtie/gffcompare_codes.png")
```

http://ccb.jhu.edu/software/stringtie/gffcompare_codes.png

```{r plot_gffcompare, fig.width = 6.85, fig.asp = .65 / 2}
df %>%
    mutate(grouped_num_exons = cut(num_exons, 
        breaks = c(0, 1, 2, 5, max(num_exons)),
        labels = c("1", "2", "3-5", ">5"))) %>% 
    ggplot(aes(class_code)) +
    geom_bar(aes(fill = grouped_num_exons), position = "fill") +
    facet_wrap(~dataset, nrow = 1) +
    scale_fill_brewer(palette = "Dark2", name = "exons") +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    ylab(NULL) +
    scale_y_continuous(labels = scales::percent) +
    theme(legend.key.size = unit(.33, "cm"))


df %>% 
    ggplot(aes(class_code, len)) +
    geom_boxplot() +
    facet_wrap(~dataset) +
    scale_y_log10() +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    ylab("Transcript length [nt]")
```


```{r plot_gffcompare2, fig.width = 6.85 / 2, fig.asp = .65}
df %>% 
    ggplot(aes(dataset)) +
    geom_bar(aes(fill = class_code), position = "fill") +
    scale_fill_viridis_d() +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    ylab(NULL) +
    xlab(NULL) +
    scale_y_continuous(labels = scales::percent) +
    theme(legend.key.size = unit(.33, "cm")) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))

df %>% 
    ggplot(aes(dataset)) +
    geom_bar(aes(fill = class_code)) +
    scale_fill_viridis_d() +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    ylab(NULL) +
    xlab(NULL) +
    theme(legend.key.size = unit(.33, "cm")) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))
```

# Sqanti

```{r}
knitr::include_graphics("https://raw.githubusercontent.com/Magdoll/images_public/master/SQANTI2_figures/sqanti2_classification.png")
```

https://github.com/Magdoll/SQANTI2

```{r plot_sqanti, fig.width = 6.85, fig.asp = .65}
df2 %>% 
    mutate(grouped_num_exons = cut(exons, 
        breaks = c(0, 1, 2, 5, max(exons)),
        labels = c("1", "2", "3-5", ">5"))) %>% 
    ggplot(aes(structural_category)) +
    geom_bar(aes(fill = grouped_num_exons), position = "fill") +
    facet_wrap(~dataset, nrow = 2) +
    scale_fill_brewer(palette = "Dark2", name = "exons") +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    ylab(NULL) +
    xlab(NULL) +
    scale_y_continuous(labels = scales::percent) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5)) +
    theme(legend.key.size = unit(.33, "cm"))


df2 %>% 
    ggplot(aes(structural_category, length)) +
    geom_boxplot() +
    facet_wrap(~dataset, nrow = 2) +
    scale_y_log10() +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    ylab("Transcript length [nt]") +
    xlab(NULL) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))
```


```{r plot_sqanti2, fig.width = 6.85 / 2, fig.asp = .65}
df2 %>% 
    mutate(structural_category = factor(structural_category, rev(levels(structural_category)))) %>% 
    ggplot(aes(dataset)) +
    geom_bar(aes(fill = structural_category), position = "fill") +
    scale_fill_brewer(palette = "Paired", name = NULL, limits = levels(df2$structural_category)) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    ylab("Reassembled transcripts") +
    xlab(NULL) +
    scale_y_continuous(labels = scales::percent) +
    theme(legend.key.size = unit(.33, "cm")) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))

df2 %>% 
    mutate(structural_category = factor(structural_category, rev(levels(structural_category)))) %>% 
    ggplot(aes(dataset)) +
    geom_bar(aes(fill = structural_category)) +
    scale_fill_brewer(palette = "Paired", name = NULL, limits = levels(df2$structural_category)) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    ylab("Reassembled transcripts") +
    xlab(NULL) +
    theme(legend.key.size = unit(.33, "cm")) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))
```

```{r, fig.width = 6.85 / 2 / 3, fig.asp = .65 * 3}
df2 %>% 
    mutate(structural_category = factor(structural_category, rev(levels(structural_category)))) %>% 
    ggplot(aes(dataset)) +
    geom_bar(aes(fill = structural_category)) +
    scale_fill_brewer(palette = "Paired", name = NULL, limits = levels(df2$structural_category)) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    scale_y_continuous(
        "Reassembled transcripts",
        labels = label_number(accuracy = 1, scale = 1/1000, suffix = " k")) +
    xlab(NULL) +
    theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))
```


```{r plot_sqanti3, fig.width = 6.85 / 2, fig.asp = .65}
df2 %>% 
    filter(structural_category == "incomplete-splice_match") %>% 
    ggplot(aes(dataset)) +
    geom_bar(aes(fill = subcategory)) +
    scale_fill_brewer(palette = "Dark2", name = NULL) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    ylab("Incomplete splice matches") +
    xlab(NULL) +
    theme(legend.key.size = unit(.33, "cm")) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))
```


```{r, fig.width = 6.85 / 2 / 3, fig.asp = .65 * 3}
df2 %>% 
    filter(structural_category == "incomplete-splice_match") %>% 
    ggplot(aes(dataset)) +
    geom_bar(aes(fill = subcategory)) +
    scale_fill_brewer(palette = "Dark2", name = NULL) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    scale_y_continuous(
        "Incomplete splice matches",
        labels = label_number(accuracy = 1, scale = 1/1000, suffix = " k")) +
    xlab(NULL) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5)) +
    theme(legend.position = "none")
```


```{r}
knitr::include_graphics("https://github.com/Magdoll/images_public/raw/master/SQANTI2_figures/sqanti2_ISM_subtype.png")
```


```{r plot_sqanti4, fig.width = 6.85 / 2, fig.asp = .65}
df2 %>% 
    filter(structural_category %in% c("novel_in_catalog", "novel_not_in_catalog")) %>% 
    ggplot(aes(dataset)) +
    geom_bar(aes(fill = subcategory)) +
    scale_fill_brewer(palette = "Set1", name = NULL) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    ylab("Novel transcripts") +
    xlab(NULL) +
    theme(legend.key.size = unit(.33, "cm")) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))
```


```{r fig.width = 6.85 / 2 / 3, fig.asp = .65 * 3}
df2 %>% 
    filter(structural_category %in% c("novel_in_catalog", "novel_not_in_catalog")) %>% 
    ggplot(aes(dataset)) +
    geom_bar(aes(fill = subcategory)) +
    scale_fill_brewer(palette = "Set1", name = NULL) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    scale_y_continuous(
        "Novel transcripts",
        labels = label_number(accuracy = 1, scale = 1/1000, suffix = " k")) +
    xlab(NULL) +
    theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))
```


```{r}
knitr::include_graphics("https://github.com/Magdoll/images_public/raw/master/SQANTI2_figures/sqanti2_NIC_subtype.png")
```

# Overlap

## Squanti

Comparison between the different methods ability to identify 
transcripts from the reference annotation.

```{r upset, fig.asp = .65}
df2 %>% 
    filter(grepl("^ENS", associated_transcript)) %>% 
    group_by(dataset, associated_transcript) %>% 
    summarise(n = n()) %>% 
    ungroup() %>% 
    tidyr::pivot_wider(names_from = dataset, values_from = n, values_fill = 0) %>% 
    mutate_if(is.numeric, ~ as.integer(. > 0)) %>% 
    as.data.frame() %>%
    tibble::column_to_rownames("associated_transcript") %>% 
    upset(nsets = 7, keep.order = TRUE,
        nintersects = 20,
        order.by = "freq",
        sets = c("illumina_stringtie", "teloprime_stringtie", "cdna_stringtie",
            "rna_stringtie", "teloprime_flair", "cdna_flair", "rna_flair"),
        sets.x.label = "Annotated transcripts",
        mainbar.y.label = "Shared transcripts\n(based on reference annotation)",
        text.scale = 1.5)
```

## Gffcompare

Using the tracking file from gffcompare.

### Transcripts in the reference

Similar to the sqanti plot above, just using gffcompare.

The numbers are a little different, because, next to a different algorithm, 
gffcompare not only matches agains the reference, but matches every query against each other.
So there can be (and are) multiple matches for one reference transcript. 
In the sqanti analysis, the question was only "is there anything matching this reference transcript".
Interpretation should be the same as with sqanti.

Class code =, c or m.

c and m are equivalent to incomplete splice match in sqanti. So this should resemble
the old analysis more.

```{r, fig.asp = .65}
tracking %>% 
    filter(class_code %in% c("=", "c", "m")) %>% 
    dplyr::select(-query_locus_id, -reference_gene_id, -class_code) %>% 
    mutate(
        across(
            !matches("query_transfrag"),
            ~ if_else(is.na(.x), 0, 1))) %>% 
    as.data.frame() %>%
    tibble::column_to_rownames("query_transfrag_id") %>% 
    upset(nsets = 4, keep.order = TRUE,
        nintersects = 15,
        order.by = "freq",
        sets = c("stringtie_illumina", "stringtie_teloprime", "stringtie_cdna", 
            "stringtie_rna", "flair_teloprime", "flair_cdna", "flair_rna"),
        sets.x.label = "Annotated transcripts",
        mainbar.y.label = "Annotated transcript",
        text.scale = 1.5)
```



### Transcripts not in the reference

Class codes not in  =, m , c.

```{r, fig.asp = .65}
tracking %>% 
    filter(!class_code %in% c("=", "m", "c")) %>% 
    dplyr::select(-query_locus_id, -reference_gene_id, -class_code) %>% 
    mutate(
        across(
            !matches("query_transfrag"),
            ~ if_else(is.na(.x), 0, 1))) %>% 
    as.data.frame() %>%
    tibble::column_to_rownames("query_transfrag_id") %>% 
    upset(nsets = 4, keep.order = TRUE,
        nintersects = 15,
        order.by = "freq",
        sets = c("stringtie_illumina", "stringtie_teloprime", "stringtie_cdna", 
            "stringtie_rna", "flair_teloprime", "flair_cdna", "flair_rna"),
        sets.x.label = "Novel transcripts",
        mainbar.y.label = "Novel transcripts",
        text.scale = 1.5)
```


# Sensitivity and specificity

Based on gffcompare output. The numbers are entere by hand. Be carefull if something
changes in the analysis!

```{r, fig.width = 6.85 / 2, fig.asp = .65}
tibble(
    dataset = c("illumina_stringtie", "teloprime_stringtie", "cdna_stringtie",
        "rna_stringtie", "teloprime_flair", "cdna_flair", "rna_flair"),
    sensitivity = c(19.7, 23.1, 19.3, 18.1, 19.0, 15.1, 15.8) / 100,
    precision = c(72.0, 76.5, 70.0, 74.7, 43.9, 45.8, 65.2) / 100) %>% 
    ggplot(aes(precision, sensitivity)) +
    geom_point(aes(colour = grepl("flair", dataset)), size = 1, alpha = .5) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    scale_color_brewer(palette = "Dark2", guide = NULL) + 
    scale_x_continuous(name = "Precision", labels = label_percent(accuracy = 1)) +
    scale_y_continuous(name = "Sensitivity", labels = label_percent()) +
    ggrepel::geom_text_repel(aes(label = dataset), size = 2)
```


# Alternative splicing analysis

Compared to hypothetical longest pre-mRNA.

Get the tables from IsoformSwitchanalyser ($AlternativeSplicingAnalysis).

Had to run isoformswitchanalyser again, using no prefiltering, so even lowly 
expressed transcripts are analysed (which would never reach significance in 
DTU analysis and are therefore excluded there.)

```{r, fig.width = 6.85 / 2, fig.asp = .65}

# import switch analyzer lists and extract switching type info
switch_conseq <- snakemake@input$sal %>% 
    set_names(basename(.) %>% 
                  sub("_sal.rds$", "", .)) %>% 
    map(read_rds) %>% 
    map(`$`, "AlternativeSplicingAnalysis") %>% 
    map(as_tibble) %>% 
    map(arrange, isoform_id) %>%
    bind_rows(.id = "dataset") %>% 
    mutate(AS = if_else(ES | MEE | MES | IR | A5 | A3, 1, 0)) %>% 
    dplyr::select(dataset, isoform_id, ATTS, ATSS, AS)

# add class code
dataset_lookup <- c("illumina_stringtie", "cdna_stringtie", "cdna_flair",  
    "teloprime_flair", "rna_flair", "teloprime_stringtie", "rna_stringtie")

names(dataset_lookup) <- c("stringtie_illumina", "stringtie_cdna", "flair_cdna",
    "flair_teloprime", "flair_rna", "stringtie_teloprime", "stringtie_rna")

tracking2 <- tracking %>% 
    dplyr::select(-query_transfrag_id, -query_locus_id, -reference_gene_id) %>% 
    tidyr::pivot_longer(cols = !matches("class_code"), 
        names_to = "dataset", values_to = "isoform_id") %>% 
    tidyr::drop_na() %>% 
    tidyr::separate(isoform_id, c("gene", "isoform_id"), extra = "drop", sep = "\\|") %>% 
    dplyr::select(-gene) %>% 
    mutate(dataset = dataset_lookup[dataset])

# combine
df_splice <- switch_conseq %>%
    left_join(tracking2, c("dataset", "isoform_id")) %>%
    mutate(status = if_else(class_code %in% c("=", "m" , "c"), "known", "novel")) %>% 
    dplyr::select(-class_code) %>% 
    tidyr::pivot_longer(cols = c("ATTS", "ATSS", "AS"),
        names_to = "splice_type", values_to = "switching") %>% 
    group_by(dataset, status, splice_type) %>% 
    summarise(nr_isoforms = sum(switching)) %>% 
    ungroup() %>% 
    mutate(
        dataset = case_when(
            dataset == "illumina_stringtie"  ~ "illu_strin",
            dataset == "teloprime_stringtie" ~ "telo_strin",
            dataset == "cdna_stringtie"      ~ "cdna_strin",
            dataset == "rna_stringtie"       ~ "rna_strin",
            dataset == "teloprime_flair"     ~ "telo_flair",
            dataset == "cdna_flair"          ~ "cdna_flair",
            dataset == "rna_flair"           ~ "rna_flair")) %>% 
    mutate(dataset = factor(
            dataset,
            levels = c("illu_strin", "telo_strin",
                "cdna_strin", "rna_strin", "telo_flair", 
                 "cdna_flair", "rna_flair")))

df_splice %>% 
    ggplot(aes(dataset, nr_isoforms)) +
    geom_bar(aes(fill = splice_type), stat = "identity") +
    scale_fill_brewer(palette = "Dark2", name = NULL) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    ylab("Transcripts") +
    xlab(NULL) +
    theme(legend.key.size = unit(.33, "cm")) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5)) +
    facet_wrap(~status)
```

