---
title: "Nanopore iBAT, Feature detection"
author: "Christoph Kiefer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
    html_document:
        toc: true
        toc_float: true
        toc_depth: 2
        fig_caption: false
        dev: "svg"
editor_options:
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = FALSE,
    message = FALSE,
    warning = FALSE,
    cache = FALSE,
    fig.width = 6.85,
    fig.asp = 0.65,
    out.width = 800)
```

```{r loadPackages}
source(".Rprofile")
library("dplyr")
library("readr")
library("ggplot2")
library("ggthemes")
library("scattermore")
library("purrr")
library("UpSetR")
library("scales")
    rename <- dplyr::rename
    select <- dplyr::select
```

```{r importData}
# biomart
biomaRt_tx <- read_rds(snakemake@input[["biomaRt_tx"]])
biomaRt_gene <- read_rds(snakemake@input[["biomaRt_gene"]])
biotype_groups <- read_csv(snakemake@input[["biotype_groups"]])

# sample info
sample_info <- read_csv(snakemake@input[["sample_info"]],
        show_col_types = FALSE) %>%
    mutate_at(vars(matches("condition")), as.factor)


get_rowSums <- function(x) {
    x %>% 
        mutate(counts = rowSums(dplyr::select_if(x, is.numeric))) %>% 
        dplyr::select(id, counts)
}


# gene level
df_gene <- snakemake@input[["gene_counts"]] %>%
    set_names(basename(.) %>% sub("_gen.+$", "", .)) %>%
    map(read_csv, show_col_types = FALSE) %>% 
    map(get_rowSums) %>% 
    bind_rows(.id = "dataset") %>% 
    left_join(biomaRt_gene, by = c("id" = "ensembl_gene_id_version")) %>%
    left_join(biotype_groups, by = c("gene_biotype" = "transcript_biotype")) %>% 
    dplyr::select(-description, -mgi_symbol)


gene_tpm <- read_csv(snakemake@input[["gene_tpm"]], show_col_types = FALSE) %>% 
    mutate(average_tpm = rowMeans(dplyr::select_if(., is.numeric))) %>% 
    select(id, average_tpm)


# tx level
df_tx <- snakemake@input[["tx_counts"]] %>%
    set_names(basename(.) %>% sub("_tra.+$", "", .)) %>%
    map(read_csv, show_col_types = FALSE) %>% 
    map(get_rowSums) %>% 
    bind_rows(.id = "dataset") %>% 
    left_join(biomaRt_tx, by = c("id" = "ensembl_transcript_id_version")) %>%
    left_join(biotype_groups, by = "transcript_biotype") %>% 
    dplyr::select(-description, -mgi_symbol)


tx_tpm <- read_csv(snakemake@input[["tx_tpm"]], show_col_types = FALSE) %>%
    mutate(average_tpm = rowMeans(dplyr::select_if(., is.numeric))) %>% 
    select(id, average_tpm)
```


# Saturation curve

```{r}
get_numberDetectedFeatures <- function(cm, prop, cutoff = 1) {
    cm %>%
        DropletUtils::downsampleMatrix(prop = prop, bycol = TRUE) %>%
        Matrix::rowSums() %>%
        (function(x) x >= cutoff) %>%
        sum()
}
```

Cutoff is 1 read in any of the librarys per method. (Maybe it should be redone
taking replicates, condition etc. into account)

```{r saturationCurve, fig.width = 6.85 / 2, fig.asp = .65}
fractions <- 10^seq(-4, 0, length.out = 25)

cm_gene <- snakemake@input[["gene_counts"]] %>%
    set_names(basename(.) %>% sub("_gen.+$", "", .)) %>%
    map(read_csv, show_col_types = FALSE) %>%
    map(tibble::column_to_rownames, "id") %>%
    map(data.matrix)

cm_tx <- snakemake@input[["tx_counts"]] %>%
    set_names(basename(.) %>% sub("_tra.+$", "", .)) %>%
    map(read_csv, show_col_types = FALSE) %>%
    map(tibble::column_to_rownames, "id") %>%
    map(data.matrix)

df_features <- 
    tibble(fraction = fractions) %>% 
        mutate(teloprime_gene = map_int(fraction, get_numberDetectedFeatures, cm = cm_gene$teloprime),
               illumina_gene = map_int(fraction, get_numberDetectedFeatures, cm = cm_gene$illumina),
               cdna_gene = map_int(fraction, get_numberDetectedFeatures, cm = cm_gene$cdna),
               rna_gene = map_int(fraction, get_numberDetectedFeatures, cm = cm_gene$rna),
               teloprime_tx = map_int(fraction, get_numberDetectedFeatures, cm = cm_tx$teloprime),
               illumina_tx = map_int(fraction, get_numberDetectedFeatures, cm = cm_tx$illumina),
               cdna_tx = map_int(fraction, get_numberDetectedFeatures, cm = cm_tx$cdna),
               rna_tx = map_int(fraction, get_numberDetectedFeatures, cm = cm_tx$rna))
    
    
tot_counts <- list(df_gene, df_tx) %>%
    set_names("gene", "tx") %>%
    bind_rows(.id = "level") %>%
    group_by(level, dataset) %>%
    summarise(tot_counts = sum(counts)) %>%
    tidyr::unite(library, dataset, level)


df_features <-  df_features %>%
    tidyr::gather(key = "library", value = "features", -fraction) %>%
    left_join(tot_counts, by = "library") %>%
    mutate(reads = fraction * tot_counts) %>%
    tidyr::separate(library, c("method", "feature_type")) %>% 
    select(-tot_counts)


df_features %>%
    mutate(method = factor(method, levels = c("illumina", "teloprime", "cdna", "rna"))) %>% 
    mutate(feature_type = if_else(feature_type == "tx", "transcripts", "genes")) %>% 
    filter(reads >= 10^4) %>% 
    ggplot(aes(reads, features)) +
    geom_path(aes(colour = method), lwd = .25) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe() +
    facet_wrap(~feature_type, scale = "free_y") +
    scale_y_log10(labels = label_scientific()) +
    scale_x_log10(labels = trans_format("log10", math_format(10^.x))) +
    xlab("Counts") +
    ylab("Detected features") +
    scale_colour_brewer(palette = "Dark2", name = NULL) +
    theme(legend.key.size = unit(.33, "cm"))
```

While the illumina dataset has reached saturation, the ont datasets have not
yet. In addition, the teloprime curves seem to plateau out at smaller values.

# Transcript detection rate by transcript length

100 % is all features detected by any of the methods.

## Overview

```{r, fig.width = 6.85 / 2, fig.asp = .65}
df_tx %>%
    tidyr::drop_na(transcript_length) %>% 
    select(dataset, counts, transcript_length, id) %>%  
    tidyr::pivot_wider(values_from = counts, names_from = dataset, values_fill = 0) %>% 
    mutate(grouped_transcript_length = cut(transcript_length,
        breaks = c(0, 500, 1000, 1500, 2000, 2500, 3000,
            3500, 4000, 4500, 5000, 7500, 123179))) %>% 
    mutate(across(.cols = c("cdna", "teloprime", "illumina", "rna"),
        ~ . >= snakemake@params$cutoff)) %>% 
    tidyr::pivot_longer(cols = c("cdna", "teloprime", "illumina", "rna"),
        names_to = "dataset", values_to = "detected") %>% 
    group_by(dataset, grouped_transcript_length) %>%
    summarise(n = n(), n_detected = sum(detected)) %>%
    mutate(rate = n_detected / n) %>%
    mutate(dataset = factor(dataset, levels = c("illumina", "teloprime", "cdna", "rna"))) %>% 
    ggplot(aes(grouped_transcript_length, rate)) +
    geom_bar(stat = "identity") +
    facet_wrap(~dataset) +
    xlab("Transcript length") +
    ylab("Detected transcripts") +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(
        data = tibble(rate = c(0, .8), grouped_transcript_length = c(0,1)),
        sides = 'l') +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    scale_y_continuous(labels = label_percent())
```

## By transcript biotype

```{r, fig.width = 6.85 / 2, fig.asp = .65}
df_tx %>%
    tidyr::drop_na(transcript_length, grouped_biotype) %>% 
    select(dataset, counts, transcript_length, id, grouped_biotype) %>%  
    tidyr::pivot_wider(values_from = counts, names_from = dataset, values_fill = 0) %>% 
    mutate(grouped_transcript_length = cut(transcript_length,
        breaks = c(0, 1000, 2000, 3000, 4000, 5000, 7500, 123179),
        labels = c("1000", "2000", "3000", "4000",
            "5000", "7500", "123179"))) %>%
    mutate(across(.cols = c("cdna", "teloprime", "illumina", "rna"),
        ~ . >= snakemake@params$cutoff)) %>% 
    tidyr::pivot_longer(cols = c("cdna", "teloprime", "illumina", "rna"),
        names_to = "dataset", values_to = "detected") %>% 
    group_by(dataset, grouped_biotype, grouped_transcript_length) %>%
    summarise(n = n(), n_detected = sum(detected)) %>%
    mutate(rate = n_detected / n) %>%
    mutate(dataset = factor(dataset, levels = c("illumina", "teloprime", "cdna", "rna"))) %>% 
    filter(grouped_biotype != "small ncRNA") %>% 
    ggplot(aes(grouped_transcript_length, rate)) +
    geom_bar(aes(fill = dataset),
        stat = "identity",
        position = position_dodge()) +
    facet_wrap(~grouped_biotype) +
    xlab("Transcript length") +
    scale_y_continuous(labels = label_percent()) +
    ylab("Detected transcripts") +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(
        data = tibble(rate = c(0, 1), grouped_transcript_length = c(0,1)),
        sides = 'l') +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5)) +
    scale_fill_brewer(palette = "Dark2", name = NULL) +
    theme(legend.key.size = unit(.33, "cm")) +
    theme(legend.position = "top",
        legend.margin = margin(t = 0, r = 0, b = 0, l = 0, unit = "cm"))
```

# Features not detected by one but by the other method

## Genes

```{r, fig.asp = .65}
df_gene %>%
    select(dataset, counts, id) %>%  
    tidyr::pivot_wider(values_from = counts, names_from = dataset, values_fill = 0) %>% 
    mutate(across(.cols = c("cdna", "teloprime", "illumina", "rna"),
        ~ as.integer(. >= snakemake@params$cutoff))) %>% 
    as.data.frame() %>%
    tibble::column_to_rownames("id") %>%
    upset(nsets = 4, keep.order = TRUE,
        order.by = "freq",
        sets = c("illumina", "teloprime", "cdna", "rna"),
        sets.x.label = "Detected genes",
        mainbar.y.label = "Shared detected genes",
        text.scale = 1.75)
```

The violin plot only shows features detected by illumina.

```{r, fig.width = 6.85, fig.asp = .65 / 2}
df_gene %>%
    select(dataset, counts, id, grouped_biotype) %>%  
    tidyr::pivot_wider(values_from = counts, names_from = dataset, values_fill = 0) %>% 
    mutate(across(.cols = c("cdna", "teloprime", "illumina", "rna"),
        ~ . >= snakemake@params$cutoff)) %>% 
    filter(illumina) %>% 
    tidyr::pivot_longer(cols = c("cdna", "teloprime", "illumina", "rna"),
        names_to = "dataset", values_to = "detected") %>% 
    filter(dataset != "illumina") %>% 
    left_join(gene_tpm, by = "id") %>% 
    filter(!grouped_biotype == c("small ncRNA")) %>% 
    mutate(dataset = factor(dataset, levels = c("teloprime", "cdna", "rna"))) %>% 
    mutate(detected = relevel(as.factor(as.logical(detected)), "TRUE")) %>% 
    ggplot(aes(grouped_biotype, average_tpm)) +
    geom_violin(aes(fill = detected), lwd = .25) +
    facet_wrap(~dataset) +
    scale_y_continuous(
        breaks = 10^seq(0:5),
        name = "illumina [TPM]",
        labels = trans_format("log10", math_format(10^.x)),
        trans = "log1p") +
    xlab(NULL) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    scale_fill_brewer(palette = "Dark2", name = "detected") +
    theme(legend.key.size = unit(.33, "cm")) 
```


```{r, fig.width = 6.85, fig.asp = .65 / 2}
df_plot <-  df_gene %>%
    select(dataset, counts, id, grouped_biotype) %>% 
    mutate(detected = counts >= snakemake@params$cutoff) %>% 
    tidyr::pivot_wider(names_from = "dataset",
        values_from = c("counts", "detected"),
        values_fill = list("counts" = 0, "detected" = FALSE)) %>%
    mutate(
        detected_cdna = if_else(detected_cdna & detected_illumina, "both", 
            if_else(detected_cdna, "longRead", 
                if_else(detected_illumina, "illumina", "n.s."))),
        detected_teloprime = if_else(detected_teloprime & detected_illumina, "both", 
            if_else(detected_teloprime, "longRead", 
                if_else(detected_illumina, "illumina", "n.s."))),
        detected_rna = if_else(detected_rna & detected_illumina, "both", 
            if_else(detected_rna, "longRead", 
                if_else(detected_illumina, "illumina", "n.s."))))  %>% 
    select(-counts_illumina, -detected_illumina) %>% 
    mutate_if(is.numeric, as.character) %>% 
    tidyr::pivot_longer(cols = matches("counts|detected"),
        names_to = c("type", "dataset"),  names_sep = '_') %>% 
    tidyr::pivot_wider(names_from = "type", values_from = "value") %>% 
    mutate(counts = as.integer(counts)) %>% 
    left_join(gene_tpm, by = "id") %>%
    filter(!grouped_biotype == c("small ncRNA")) %>% 
    mutate(dataset = factor(dataset, levels = c("teloprime", "cdna", "rna"))) %>% 
    filter(detected != "n.s.") %>% 
    mutate(detected = factor(detected, levels = c("illumina", "longRead", "both"))) %>% 
    tidyr::drop_na(average_tpm)


df_plot %>%
    ggplot(aes(counts, average_tpm)) +
    geom_scattermore(
        aes(colour = detected, pch = grouped_biotype),
        pointsize = 4,
        alpha = .5,
        pixels = c(1024, 1024 * .65)) +
    facet_wrap(~dataset) +
    scale_x_continuous(
        name = "counts",
        breaks = 10^seq(0:5),
        labels = trans_format("log10", math_format(10^.x)),
        trans = "log1p") +
    scale_y_continuous(
        name = "illumina [TPM]",
        breaks = 10^seq(0:5),
        labels = trans_format("log10", math_format(10^.x)),
        trans = "log1p") +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    scale_colour_brewer(palette = "Dark2", name = "detected") +
    theme(legend.key.size = unit(.33, "cm")) 
```


```{r, fig.width = 6.85/2, fig.asp = .65}
df_plot %>% 
    group_by(dataset, detected, grouped_biotype) %>% 
    summarise(count = n()) %>% 
    mutate(fraction = count/sum(count)) %>% 
    mutate(grouped_biotype = factor(grouped_biotype, 
        levels = c("coding", "long ncRNA", "pseudogenes", "processed_others", "others"))) %>% 
    ungroup() %>% 
    ggplot(aes(detected, fraction)) +
    geom_bar(aes(fill = grouped_biotype), stat = "identity") + 
    facet_wrap(~dataset) +
    scale_fill_brewer(palette = "Dark2", name = NULL, drop = FALSE) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'lb', stat = "identity") +
    scale_y_continuous(labels = label_percent()) +
    xlab(NULL) +
    ylab("Detected genes") +
    theme(legend.key.size = unit(.33, "cm")) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))
```


## Transcripts

```{r, fig.asp = .65}
df_tx %>%
    select(dataset, counts, id) %>%  
    tidyr::pivot_wider(values_from = counts, names_from = dataset, values_fill = 0) %>% 
    mutate(across(.cols = c("cdna", "teloprime", "illumina", "rna"),
        ~ as.integer(. >= snakemake@params$cutoff))) %>% 
    as.data.frame() %>%
    tibble::column_to_rownames("id") %>%
    upset(nsets = 4, keep.order = TRUE,
        order.by = "freq",
        sets = c("illumina", "teloprime", "cdna", "rna"),
        sets.x.label = "Detected genes",
        mainbar.y.label = "Shared detected transcripts",
        text.scale = 1.75)
```

```{r, fig.width = 6.85, fig.asp = .65 / 2}
df_tx %>%
    select(dataset, counts, id, grouped_biotype) %>%  
    tidyr::pivot_wider(values_from = counts, names_from = dataset, values_fill = 0) %>% 
    mutate(across(.cols = c("cdna", "teloprime", "illumina", "rna"),
        ~ . >= snakemake@params$cutoff)) %>% 
    filter(illumina) %>% 
    tidyr::pivot_longer(cols = c("cdna", "teloprime", "illumina", "rna"),
        names_to = "dataset", values_to = "detected") %>% 
    filter(dataset != "illumina") %>% 
    left_join(tx_tpm, by = "id") %>% 
    filter(!grouped_biotype == c("small ncRNA")) %>% 
    mutate(dataset = factor(dataset, levels = c("teloprime", "cdna", "rna"))) %>% 
    mutate(detected = relevel(as.factor(as.logical(detected)), "TRUE")) %>% 
    ggplot(aes(grouped_biotype, average_tpm)) +
    geom_violin(aes(fill = detected), lwd = .25) +
    facet_wrap(~dataset) +
    scale_y_continuous(
        breaks = 10^seq(0:5),
        name = "illumina [TPM]",
        labels = trans_format("log10", math_format(10^.x)),
        trans = "log1p") +
    xlab(NULL) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    scale_fill_brewer(palette = "Dark2", name = "detected") +
    theme(legend.key.size = unit(.33, "cm")) 
```


```{r, fig.width = 6.85, fig.asp = .65 / 2}
df_plot <-  df_tx %>%
    select(dataset, counts, id, grouped_biotype) %>% 
    mutate(detected = counts >= snakemake@params$cutoff) %>% 
    tidyr::pivot_wider(names_from = "dataset",
        values_from = c("counts", "detected"),
        values_fill = list("counts" = 0, "detected" = FALSE)) %>%
    mutate(
        detected_cdna = if_else(detected_cdna & detected_illumina, "both", 
            if_else(detected_cdna, "longRead", 
                if_else(detected_illumina, "illumina", "n.s."))),
        detected_teloprime = if_else(detected_teloprime & detected_illumina, "both", 
            if_else(detected_teloprime, "longRead", 
                if_else(detected_illumina, "illumina", "n.s."))),
        detected_rna = if_else(detected_rna & detected_illumina, "both", 
            if_else(detected_rna, "longRead", 
                if_else(detected_illumina, "illumina", "n.s."))))  %>% 
    select(-counts_illumina, -detected_illumina) %>% 
    mutate_if(is.numeric, as.character) %>% 
    tidyr::pivot_longer(cols = matches("counts|detected"),
        names_to = c("type", "dataset"),  names_sep = '_') %>% 
    tidyr::pivot_wider(names_from = "type", values_from = "value") %>% 
    mutate(counts = as.integer(counts)) %>% 
    left_join(tx_tpm, by = "id") %>%
    filter(!grouped_biotype == c("small ncRNA")) %>% 
    mutate(dataset = factor(dataset, levels = c("teloprime", "cdna", "rna"))) %>% 
    filter(detected != "n.s.") %>% 
    mutate(detected = factor(detected, levels = c("illumina", "longRead", "both"))) %>% 
    tidyr::drop_na(average_tpm)


df_plot %>% 
    ggplot(aes(counts, average_tpm)) +
    geom_scattermore(
        aes(colour = detected, pch = grouped_biotype),
        pointsize = 4,
        alpha = .5,
        pixels = c(1024, 1024 * .65)) +
    facet_wrap(~dataset) +
    scale_x_continuous(
        name = "counts",
        breaks = 10^seq(0:5),
        labels = trans_format("log10", math_format(10^.x)),
        trans = "log1p") +
    scale_y_continuous(
        name = "illumina [TPM]",
        breaks = 10^seq(0:5),
        labels = trans_format("log10", math_format(10^.x)),
        trans = "log1p") +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'bl') +
    scale_colour_brewer(palette = "Dark2", name = "detected") +
    theme(legend.key.size = unit(.33, "cm")) 
```


```{r, fig.width = 6.85/2, fig.asp = .65}
df_plot %>% 
    group_by(dataset, detected, grouped_biotype) %>% 
    summarise(count = n()) %>% 
    mutate(fraction = count/sum(count)) %>% 
    mutate(grouped_biotype = factor(grouped_biotype, 
        levels = c("coding", "long ncRNA", "pseudogenes", "processed_others", "others"))) %>% 
    ungroup() %>% 
    ggplot(aes(detected, fraction)) +
    geom_bar(aes(fill = grouped_biotype), stat = "identity") + 
    facet_wrap(~dataset) +
    scale_fill_brewer(palette = "Dark2", name = NULL, drop = FALSE) +
    theme_tufte(base_size = 8, base_family = "Helvetica") +
    geom_rangeframe(sides = 'lb', stat = "identity") +
    scale_y_continuous(labels = label_percent()) +
    xlab(NULL) +
    ylab("Detected transcripts") +
    theme(legend.key.size = unit(.33, "cm")) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))
```
